{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "ff674e76",
   "metadata": {},
   "source": [
    "\n",
    "setwd('D:/ADRENAL_GLAND_PROJECT')\n",
    "only_STERO=readRDS(file='D:/ADRENAL_GLAND_PROJECT/Only_Stero')\n",
    "multiomics=readRDS(file='D:/ADRENAL_GLAND_PROJECT/multiomics')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5b96cfae",
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "# download package "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "43a4b0bb",
   "metadata": {},
   "outputs": [],
   "source": [
    "library(S4Vectors)\n",
    "library(Signac)\n",
    "library(Seurat)\n",
    "library(BSgenome.Mmusculus.UCSC.mm10)\n",
    "library(AnnotationHub)\n",
    "library(ggplot2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f53067b6",
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "# Read file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "23969204",
   "metadata": {
    "name": "Read file"
   },
   "outputs": [],
   "source": [
    "inputdata.Female_12 <- Read10X_h5(\"D:/Stage/RAW DATA MATRIX/raw_feature_bc_matrix_female12.h5\")     # REPLICAT 1\n",
    "fragpath.female_12 <- \"D:/Stage/RAW DATA MATRIX/atac_fragments_female12.tsv.gz\"\n",
    "\n",
    "inputdata.Female_13 <- Read10X_h5(\"D:/Stage/RAW DATA MATRIX/raw_feature_bc_matrix_female13.h5\")     # REPLICAT 2 \n",
    "fragpath.female_13 <- \"D:/Stage/RAW DATA MATRIX/atac_fragments_female13.tsv.gz\"\n",
    "\n",
    "inputdata.male_16 <- Read10X_h5(\"D:/Stage/RAW DATA MATRIX/raw_feature_bc_matrix_male16.h5\")         # REPLICAT 1 \n",
    "fragpath.male_16 <- \"D:/Stage/RAW DATA MATRIX/atac_fragments_male16.tsv.gz\"\n",
    "                \n",
    "inputdata.male_17 <- Read10X_h5(\"D:/Stage/RAW DATA MATRIX/raw_feature_bc_matrix_male17.h5\")         # REPLICAT 2\n",
    "fragpath.male_17 <- \"D:/Stage/RAW DATA MATRIX/atac_fragments_male17.tsv.gz\"                 \n",
    "                                                                             "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d08f3885",
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "# Get annotation file for peak mapping on reference genome "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4f17f01a",
   "metadata": {
    "name": "Get annotation file for peak mapping on reference genome"
   },
   "outputs": [],
   "source": [
    "annotation <- GetGRangesFromEnsDb(ensdb = EnsDb.Mmusculus.v79)\n",
    "seqlevelsStyle(annotation) <- \"UCSC\"\n",
    "genome(annotation) <- \"mm10\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "14b56d31",
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "# Create RNA and ATAC seurat object ; add quality control metrics for RNA and ATAC assay "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f9802239",
   "metadata": {
    "name": "Seurat Object Generation and Quality control"
   },
   "outputs": [],
   "source": [
    "\n",
    "# after running,  default assay will be set to ATAC \n",
    "\n",
    "male1 <- CreateSeuratObject(counts = inputdata.male_16$`Gene Expression`,project = 'male1',min.features = 50) # remove cell with less than      50genes\n",
    "male1$identity <- \"male\"\n",
    "male1$Replicat <- \"Replicat_2\"\n",
    "male1[[\"percent.mt\"]] <- PercentageFeatureSet(male1, pattern = \"^mt-\")\n",
    "male1[[\"percent.ribosomal\"]] <- PercentageFeatureSet(male1, pattern = \"^Rp[sl]\")\n",
    "male1[[\"ATAC\"]] <- CreateChromatinAssay( counts = inputdata.male_16$Peaks,sep = c(\":\", \"-\"),fragments = fragpath.male_16,annotation = annotation,min.cells = 3,genome =\"mm10\")  \n",
    "DefaultAssay(male1)='ATAC'\n",
    "male1 <- NucleosomeSignal(male1)\n",
    "male1 <- TSSEnrichment(male1)\n",
    "male1$blacklist_fraction <- FractionCountsInRegion(object = male1, assay = 'ATAC',regions = blacklist_mm10) \n",
    "saveRDS(male1,file = 'male1_preQC')\n",
    "  \n",
    "      \n",
    "male2 <- CreateSeuratObject(counts = inputdata.male_17$`Gene Expression`,project='male2',min.features = 50)\n",
    "male2$identity <- \"male\"\n",
    "male2$Replicat <- \"Replicat_1\"\n",
    "male2[[\"percent.mt\"]] <- PercentageFeatureSet(male2, pattern = \"^mt-\")\n",
    "male2[[\"percent.ribosomal\"]] <- PercentageFeatureSet(male2, pattern = \"^Rp[sl]\")\n",
    "male2[[\"ATAC\"]]<- CreateChromatinAssay( counts = inputdata.male_17$Peaks,sep = c(\":\", \"-\"),fragments = fragpath.male_17,annotation = annotation,min.cells = 3,genome =\"mm10\") \n",
    "DefaultAssay(male2)='ATAC'\n",
    "male2 <- NucleosomeSignal(male2)\n",
    "male2 <- TSSEnrichment(male2)\n",
    "male2$blacklist_fraction <- FractionCountsInRegion(object = male2, assay = 'ATAC',regions = blacklist_mm10)\n",
    "saveRDS(male2,file = 'male2_preQC')\n",
    "\n",
    "\n",
    "female1=CreateSeuratObject(counts = inputdata.Female_12$`Gene Expression`,project='female1',min.features = 50)\n",
    "female1$identity <- \"female\"\n",
    "female1$Replicat <- \"Replicat_1\"\n",
    "female1[[\"percent.mt\"]] <- PercentageFeatureSet(female1, pattern = \"^mt-\")\n",
    "female1[[\"percent.ribosomal\"]] <- PercentageFeatureSet(female1, pattern = \"^Rp[sl]\")\n",
    "female1[[\"ATAC\"]] <- CreateChromatinAssay( counts = inputdata.Female_12$Peaks,sep = c(\":\", \"-\"),fragments = fragpath.female_12,annotation = annotation,min.cells = 3,genome =\"mm10\") \n",
    "DefaultAssay(female1)='ATAC'\n",
    "female1 <- NucleosomeSignal(female1)\n",
    "female1 <- TSSEnrichment(female1)\n",
    "female1$blacklist_fraction <- FractionCountsInRegion(object = female1, assay = 'ATAC',regions = blacklist_mm10)\n",
    "saveRDS(female1,file = 'female1_preQC')\n",
    "\n",
    "\n",
    "female2=CreateSeuratObject(counts = inputdata.Female_13$`Gene Expression`,project = 'female2',min.features = 50)\n",
    "female2$identity <- \"female\"\n",
    "female2$Replicat <- \"Replicat_2\"\n",
    "female2[[\"percent.mt\"]] <- PercentageFeatureSet(female2, pattern = \"^mt-\")\n",
    "female2[[\"percent.ribosomal\"]] <- PercentageFeatureSet(female2, pattern = \"^Rp[sl]\")\n",
    "female2[[\"ATAC\"]] <- CreateChromatinAssay( counts = inputdata.Female_13$Peaks,sep = c(\":\", \"-\"),fragments = fragpath.female_13,annotation = annotation,min.cells = 3,genome =\"mm10\" ) \n",
    "DefaultAssay(female2)='ATAC'\n",
    "female2 <- NucleosomeSignal(female2)\n",
    "female2 <- TSSEnrichment(female2)\n",
    "female2$blacklist_fraction <- FractionCountsInRegion(object = female2, assay = 'ATAC',regions = blacklist_mm10)\n",
    "saveRDS(female2,file = 'female2_preQC')\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "83e86d8e",
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "# Metrics visualisation and subset\n",
    "## Male 1 "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "42efc089",
   "metadata": {
    "name": "Male1"
   },
   "outputs": [],
   "source": [
    "\n",
    "metadata_male1=male1@meta.data\n",
    "\n",
    "metadata_male1=subset(male1@meta.data, subset = nCount_RNA >50& nCount_RNA < 25000 & nCount_ATAC >5000 & nCount_ATAC <100000 &  TSS.enrichment > 4 & percent.mt< 0.50& nucleosome_signal<0.7)\n",
    "\n",
    "\n",
    "plotRank.sample <- subset(metadata_male1, cells = WhichCells(metadata_male1))\n",
    "plot(sort(plotRank.sample$nCount_RNA ), xlab='cell', log='y', main= paste('count per cell -mito(ordered) - '))\n",
    "\n",
    "metadata_male1 %>% \n",
    "  \tggplot(aes(color=identity, x=nFeature_RNA ,fill=identity)) + \n",
    "  \tgeom_density(alpha = 0.2) + \n",
    "  \tscale_x_log10() + \n",
    "  \ttheme_classic() +\n",
    "  \tylab(\"Cell density\")  # +\n",
    "   geom_vline(xintercept = 10)\n",
    "\n",
    "metadata_male1 %>% \n",
    "  \tggplot(aes(x=nCount_RNA, y=nFeature_RNA, color=nCount_RNA)) + \n",
    "  \tgeom_point() + \n",
    "\tscale_colour_gradient(low = \"gray90\", high = \"black\") +\n",
    "  \tscale_x_log10() + \n",
    "  \tscale_y_log10() + \n",
    "  \ttheme_classic() +\n",
    "  \tfacet_wrap(~identity)\n",
    "\n",
    "\n",
    "test1 <- subset(male1,subset = nFeature_RNA >500 & nCount_RNA < 25000 & nCount_ATAC >5000 & nCount_ATAC <100000 &  TSS.enrichment > 4 & percent.mt< 0.50 & nucleosome_signal<0.7)\n",
    "\n",
    "VlnPlot(object = test1,features =c( 'nCount_ATAC','nCount_RNA','nFeature_ATAC','nFeature_RNA','percent.mt','TSS.enrichment','nucleosome_signal'))\n",
    "plot1 <- FeatureScatter(test1, feature1 = \"percent.ribosomal\", feature2 = \"nCount_RNA\",group.by = \"identity\")\n",
    "plot2 <- FeatureScatter(test1, feature1 = \"nCount_RNA\", feature2 = \"nFeature_RNA\",group.by = \"Replicat\",raster = F)\n",
    "plot3 <- FeatureScatter(test1, feature1= 'nCount_ATAC' , feature2 = 'nFeature_ATAC' , group.by = 'Replicat')\n",
    "plot4 <- FeatureScatter(test1, feature1= 'TSS.enrichment' , feature2 = 'TSS.percentile' , group.by = 'Replicat')\n",
    "plot1 + plot2 + plot3  +plot4\n",
    "\n",
    "male1 <- subset(male1,subset = nFeature_RNA >500 & nCount_RNA < 25000 & nCount_ATAC >5000 & nCount_ATAC <100000 &  TSS.enrichment > 4 & percent.mt< 0.50 & nucleosome_signal<0.7)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "84e92688",
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "## Male 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cfac8633",
   "metadata": {
    "name": "Male2"
   },
   "outputs": [],
   "source": [
    "\n",
    "metadata_male2=subset(male2@meta.data , subset = nFeature_RNA > 500 & nCount_RNA <35000& percent.mt< 0.75 &nCount_ATAC>5000 &nCount_ATAC <100000 & TSS.enrichment>3)\n",
    "                      \n",
    "plotRank.sample <- subset(metadata_male2, cells = WhichCells(metadata_male2))\n",
    "plot(sort(plotRank.sample$nCount_RNA), xlab='cell', log='y', main= paste('count per cell -mito(ordered) - '))\n",
    "\n",
    "metadata_male2 %>% \n",
    "  \tggplot(aes(color=identity, x=nucleosome_signal, fill=identity)) + \n",
    "  \tgeom_density(alpha = 0.2) + \n",
    "  \tscale_x_log10() + \n",
    "  \ttheme_classic() +\n",
    "  \tylab(\"Cell density\") #+\n",
    " # geom_vline(xintercept = 5000)\n",
    "\n",
    "metadata_male2 %>% \n",
    "  \tggplot(aes(x=nCount_ATAC, y=nFeature_ATAC, color=percent.mt)) + \n",
    "  \tgeom_point() + \n",
    "\tscale_colour_gradient(low = \"gray90\", high = \"black\") +\n",
    "  \tstat_smooth(method=lm) +\n",
    "  \tscale_x_log10() + \n",
    "  \tscale_y_log10() + \n",
    "  \ttheme_classic() +\n",
    "  \tgeom_vline(xintercept = 500) +\n",
    "  \tgeom_hline(yintercept = 250) +\n",
    "  \tfacet_wrap(~identity)\n",
    "\n",
    "\n",
    "test2 <- subset(male2, subset = nFeature_RNA > 500 & nCount_RNA <35000 & percent.mt< 0.75 &nCount_ATAC>5000 &nCount_ATAC <100000 & TSS.enrichment>3)\n",
    "\n",
    "VlnPlot(object = test2,features =c( 'nCount_ATAC','nCount_RNA','nFeature_ATAC','nFeature_RNA','percent.mt','TSS.enrichment','nucleosome_signal'))\n",
    "plot1 <- FeatureScatter(test2, feature1 = \"percent.ribosomal\", feature2 = \"nCount_RNA\",group.by = \"identity\")\n",
    "plot2 <- FeatureScatter(test2, feature1 = \"nCount_RNA\", feature2 = \"nFeature_RNA\",group.by = \"Replicat\",raster = F)\n",
    "plot3 <- FeatureScatter(test2, feature1= 'nCount_ATAC' , feature2 = 'nFeature_ATAC' , group.by = 'Replicat')\n",
    "plot4 <- FeatureScatter(test2, feature1= 'TSS.enrichment' , feature2 = 'TSS.percentile' , group.by = 'Replicat')\n",
    "plot1 + plot2 + plot3  +plot4\n",
    "\n",
    "\n",
    "male2 <- subset(male2, subset = nFeature_RNA > 500 & nCount_RNA <35000 & percent.mt< 0.75 &nCount_ATAC>5000 &nCount_ATAC <100000 & TSS.enrichment>3)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "db6ff0da",
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "## Female 1 "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1cbf3daf",
   "metadata": {
    "name": "Female1"
   },
   "outputs": [],
   "source": [
    "                                                     \n",
    "\n",
    "metadata_female1=subset(female1@meta.data, subset = nCount_RNA > 700 &nCount_RNA < 50000 & nCount_ATAC> 7000 & nCount_ATAC< 300000 &TSS.enrichment>3 & percent.mt< 0.75&nucleosome_signal<0.7)\n",
    "\n",
    "plotRank.sample <- subset(metadata_female1, cells = WhichCells(metadata_female1))\n",
    "plot(sort(plotRank.sample$nCount_RNA), xlab='cell', log='y', main= paste('count per cell -mito(ordered) - '))\n",
    "\n",
    "                        \n",
    "\n",
    "metadata_female1 %>% \n",
    "  \tggplot(aes(color=identity, x=nCount_RNA, fill=identity)) + \n",
    "  \tgeom_density(alpha = 0.2) + \n",
    "  \tscale_x_log10() + \n",
    "  \ttheme_classic() +\n",
    "  \tylab(\"Cell density\") #+\n",
    " # geom_vline(xintercept = 5000)\n",
    "\n",
    "metadata_female1 %>% \n",
    "  \tggplot(aes(x=nCount_ATAC, y=nFeature_ATAC, color=percent.mt)) + \n",
    "  \tgeom_point() + \n",
    "\tscale_colour_gradient(low = \"gray90\", high = \"black\") +\n",
    "  \tstat_smooth(method=lm) +\n",
    "  \tscale_x_log10() + \n",
    "  \tscale_y_log10() + \n",
    "  \ttheme_classic() +\n",
    "  \tgeom_vline(xintercept = 500) +\n",
    "  \tgeom_hline(yintercept = 250) +\n",
    "  \tfacet_wrap(~identity)\n",
    "\n",
    "\n",
    "test3 <- subset(female1, subset = nCount_RNA > 700 &nCount_RNA < 50000 & nCount_ATAC> 7000 & nCount_ATAC< 300000 &TSS.enrichment>3 & percent.mt< 0.75&nucleosome_signal<0.7)\n",
    "\n",
    "VlnPlot(object = test3,features =c( 'nCount_ATAC','nCount_RNA','nFeature_ATAC','nFeature_RNA','percent.mt','TSS.enrichment','nucleosome_signal'))\n",
    "plot1 <- FeatureScatter(test3, feature1 = \"percent.ribosomal\", feature2 = \"nCount_RNA\",group.by = \"identity\")\n",
    "plot2 <- FeatureScatter(test3, feature1 = \"nCount_RNA\", feature2 = \"nFeature_RNA\",group.by = \"Replicat\",raster = F)\n",
    "plot3 <- FeatureScatter(test3, feature1= 'nCount_ATAC' , feature2 = 'nFeature_ATAC' , group.by = 'Replicat')\n",
    "plot4 <- FeatureScatter(test3, feature1= 'TSS.enrichment' , feature2 = 'TSS.percentile' , group.by = 'Replicat')\n",
    "plot1 + plot2 + plot3  +plot4\n",
    "\n",
    "female1 <- subset(female1, subset = nCount_RNA > 700 &nCount_RNA < 50000 & nCount_ATAC> 7000 & nCount_ATAC< 300000 &TSS.enrichment>3 & percent.mt< 0.75&nucleosome_signal<0.7)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "405e84e9",
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "## Female 2 "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b51d6576",
   "metadata": {
    "name": "Female2"
   },
   "outputs": [],
   "source": [
    "\n",
    "metadata_female2=subset(female2@meta.data,subset = nFeature_RNA > 500 & nCount_RNA<100000 & nCount_ATAC>5000 & nCount_ATAC <500000& TSS.enrichment>5 &percent.mt<0.5&nucleosome_signal<0.7)\n",
    "\n",
    "\n",
    "plotRank.sample <- subset(metadata_female2, cells = WhichCells(metadata_female2))\n",
    "plot(sort(plotRank.sample$nCount_RNA), xlab='cell', log='y', main= paste('count per cell -mito(ordered) - '))\n",
    "\n",
    "metadata_female2 %>% \n",
    "  \tggplot(aes(color=identity, x=nCount_RNA, fill=identity)) + \n",
    "  \tgeom_density(alpha = 0.2) + \n",
    "  \tscale_x_log10() + \n",
    "  \ttheme_classic() +\n",
    "  \tylab(\"Cell density\") #+\n",
    " # geom_vline(xintercept = 5000)\n",
    "\n",
    "metadata_female2 %>% \n",
    "  \tggplot(aes(x=nCount_ATAC, y=nFeature_ATAC, color=percent.mt)) + \n",
    "  \tgeom_point() + \n",
    "\tscale_colour_gradient(low = \"gray90\", high = \"black\") +\n",
    "  \tstat_smooth(method=lm) +\n",
    "  \tscale_x_log10() + \n",
    "  \tscale_y_log10() + \n",
    "  \ttheme_classic() +\n",
    "  \tgeom_vline(xintercept = 500) +\n",
    "  \tgeom_hline(yintercept = 250) +\n",
    "  \tfacet_wrap(~identity)\n",
    "\n",
    "\n",
    "# total fragment\n",
    "totals <- CountFragments(fragments = path.to.fragfile)\n",
    "hist(log10(totals$frequency_count), breaks = 100)\n",
    "cells.use <- totals[totals$frequency_count > 1000, \"CB\"]\n",
    "counts <- FeatureMatrix(fragments = fragment.object, features = peaks, cells = cells.use)\n",
    "\n",
    "\n",
    "\n",
    "test4 <- subset(female2, subset = nFeature_RNA > 500 & nCount_RNA < 15000&nCount_ATAC >5000& nCount_ATAC <100000 & TSS.enrichment > 3&nucleosome_signal<0.7)\n",
    "\n",
    "VlnPlot(object = test4,features =c( 'nCount_ATAC','nCount_RNA','nFeature_ATAC','nFeature_RNA','percent.mt','TSS.enrichment','nucleosome_signal'))\n",
    "plot1 <- FeatureScatter(test4, feature1 = \"percent.ribosomal\", feature2 = \"nCount_RNA\",group.by = \"identity\")\n",
    "plot2 <- FeatureScatter(test4, feature1 = \"nCount_RNA\", feature2 = \"nFeature_RNA\",group.by = \"Replicat\",raster = F)\n",
    "plot3 <- FeatureScatter(test4, feature1= 'nCount_ATAC' , feature2 = 'nFeature_ATAC' , group.by = 'Replicat')\n",
    "plot4 <- FeatureScatter(test4, feature1= 'TSS.enrichment' , feature2 = 'TSS.percentile' , group.by = 'Replicat')\n",
    "plot1 + plot2 + plot3  +plot4\n",
    "\n",
    "female2 <- subset(female2, subset = nFeature_RNA > 500 & nCount_RNA < 15000&nCount_ATAC >5000& nCount_ATAC <100000 & TSS.enrichment > 3&nucleosome_signal<0.7)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e2a4226d",
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "# merge all 4 files "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "de774219",
   "metadata": {
    "name": "Merge"
   },
   "outputs": [],
   "source": [
    "\n",
    "multiomics=merge(x = male1,y = c(male2,female1,female2),add.cell.ids =c( 'male1','male2','female1','female2'),project = \"multiomics\",merge.data = T)\n",
    "saveRDS(data_filt_merged)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "55d6b6ed",
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "# Quick overview"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7c2037db",
   "metadata": {
    "name": "Metrics visualisation and subset"
   },
   "outputs": [],
   "source": [
    "multiomics=readRDS(file = '/home/absta/Documents/DATA_FINAL_STAGE/multiomics_int')\n",
    "DefaultAssay(multiomics)='RNA'\n",
    "multiomics[['SCT']]=NULL\n",
    "multiomics[['peaks']]=NULL\n",
    "#multiomics <- multiomics[,!colnames(multiomics) %in% subset_endo] \n",
    "#multiomics <- multiomics[,!colnames(multiomics) %in% subset_medulla]\n",
    "#multiomics <- multiomics[,!colnames(multiomics) %in% ENDO2]\n",
    "multiomics <- multiomics[,!colnames(multiomics) %in% subset] # process by first running to the clustering and highlinghting this cells\n",
    "\n",
    "\n",
    "multiomics@meta.data[15:27]=NULL\n",
    "multiomics=subset(multiomics,subset  = nCount_RNA<20000)\n",
    "multiomics@reductions[[\"umap.rna.pca\"]]=NULL\n",
    "multiomics@reductions[[\"umap.rna_ica\"]]=NULL\n",
    "multiomics@reductions[[\"umap.atac\"]]=NULL\n",
    "multiomics@reductions[[\"wnn.umap\"]]=NULL\n",
    "\n",
    "VlnPlot(multiomics,features = 'nCount_RNA')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f9f3747e",
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "# FULL dataset analysis ( Multiomics )\n",
    "\n",
    "## RNA PROCESSING "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "df8b3dfb",
   "metadata": {},
   "outputs": [],
   "source": [
    "#  Normalisation , scaling, PCA, ICA , UMAP and clustering\n",
    "\n",
    "\n",
    "# 0) Replicat integration  \n",
    "      \n",
    "      DefaultAssay(multiomics)='RNA'  # take RNA assay \n",
    "      multiomics <- SplitObject(multiomics, split.by = \"Replicat\") # list \n",
    "      for (i in 1:length(multiomics)) {\n",
    "          multiomics[[i]] <- SCTransform(multiomics[[i]],variable.features.n =NULL,variable.features.rv.th =1.1\n",
    "                                         ,return.only.var.genes = FALSE,assay = 'RNA',\n",
    "                                         new.assay.name = 'SCT_INT',vars.to.regress = 'percent.mt')\n",
    "          }\n",
    "      \n",
    "      \n",
    "      # Select the most variable features to use for integration\n",
    "      integ_features <- SelectIntegrationFeatures(object.list = multiomics, nfeatures = 3000) \n",
    "      \n",
    "      # Prepare the SCT list object for integration\n",
    "      multiomics  <- PrepSCTIntegration(object.list = multiomics, \n",
    "                                         anchor.features =integ_features)\n",
    "      \n",
    "      \n",
    "      integ_anchors <- FindIntegrationAnchors(object.list = multiomics, \n",
    "                                              normalization.method = \"SCT\", \n",
    "                                              anchor.features = integ_features,reduction=\"cca\")\n",
    "      \n",
    "      # Integrate across conditions\n",
    "      multiomics <- IntegrateData(anchorset = integ_anchors,normalization.method = \"SCT\")    \n",
    "      \n",
    "     \n",
    "\n",
    "      \n",
    "### 1)  Normalisation / PC / IC / UMAP /CLUSTERING on integrated and non integrated assay\n",
    "      \n",
    "    # 1.1) NORMALISATION AND PC ON NON ONTEGRATED assay       \n",
    "          multiomics=SCTransform(multiomics,variable.features.n =NULL,variable.features.rv.th =1.1,return.only.var.genes = FALSE,assay = 'RNA',new.assay.name = 'SCT_NO_INT',vars.to.regress = 'percent.mt')\n",
    "          multiomics=RunPCA(multiomics, verbose = FALSE,assay=\"SCT_NO_INT\",reduction.name='PCA_NO_INT') # Default 50 \n",
    "          multiomics=RunICA(multiomics, verbose = FALSE,assay=\"SCT_NO_INT\",reduction.name='ICA_NO_INT')\n",
    "          ElbowPlot(multiomics,ndims = 50,reduction = 'PCA_NO_INT') # Variance plot \n",
    "      \n",
    "    # 1.2) ON INTEGRATED  assay \n",
    "          DefaultAssay(multiomics)='integrated'\n",
    "          multiomics=RunPCA(multiomics, verbose = FALSE,assay=\"integrated\",reduction.name='PCA_INT',npcs = 50)\n",
    "          multiomics=RunICA(multiomics, verbose = FALSE,assay=\"integrated\",reduction.name='ICA_INT')\n",
    "\n",
    "          ElbowPlot(multiomics,reduction ='PCA_INT',ndims = 50 ) # choose 15\n",
    "      \n",
    "    # 1.3)  Get the total variance explained for each PC:\n",
    "        \n",
    "        total.variance   <- function(x,y) { \n",
    "              mat <- Seurat::GetAssayData(multiomics, assay = x, slot = \"scale.data\")\n",
    "              pca <- multiomics[[y]]\n",
    "              \n",
    "              # Get the total variance:\n",
    "              total_variance <- sum(matrixStats::rowVars(mat))\n",
    "              eigValues = (pca@stdev)^2  ## EigenValues\n",
    "              varExplained = eigValues / total_variance\n",
    "              barplot(varExplained)\n",
    "              print(varExplained)\n",
    "        }\n",
    "        \n",
    "        total.variance('integrated','PCA_INT')\n",
    "        \n",
    "        \n",
    "\n",
    "        \n",
    "    # 1.4) Tcheck Variabilty explained by MITOCHONDRIAL genes \n",
    "        summary(multiomics@meta.data$percent.mt)\n",
    "        # Turn mitoRatio into categorical factor vector based on quartile values\n",
    "        percent.mt_pbmc_QC=multiomics # create new pbmc object for inserting  labels \n",
    "        percent.mt_pbmc_QC@meta.data$percent.mt<- cut(multiomics@meta.data$percent.mt, \n",
    "                           breaks=c(-Inf, 0, 0.03814   , 0.08238   , Inf), \n",
    "                           labels=c(\"Low\",\"Medium\",\"Medium high\", \"High\"))\n",
    "        # Plot the PCA colored by mitoFr\n",
    "        DimPlot(percent.mt_pbmc_QC,reduction = \"PCA_NO_INT\",group.by= \"percent.mt\", split.by = \"percent.mt\")  \n",
    "        # View all PC \n",
    "        runfunction <- function(z,x,y){DimPlot(object = z, dims = c(x, y), reduction = \"PCA_NO_INT\",group.by   = \"percent.mt\") \n",
    "        }\n",
    "        for ( i in (1:15)){x=runfunction(percent.mt_pbmc_QC,i+1,i)  # plot 15 dims\n",
    "                                     print(x)}\n",
    "  \n",
    "\n",
    "        \n",
    "### 2 ) UMAP REPRESENTATION and clustering       \n",
    "              \n",
    "    # 2.1) INTEGRATED ASSAY\n",
    "      multiomics=RunUMAP(multiomics,reduction='PCA_INT',dims = 1:20,reduction.name = 'UMAP_PCA_INT')  # UMAP on PC \n",
    "      multiomics=RunUMAP(multiomics,dims = 1:20, local.connectivity  = 6 ,reduction = 'ICA_INT',umap.method = \"umap-learn\",metric = \"correlation\",n.neighbors =200,reduction.name = 'umap.rna_ica',reduction.key='icaUMAP_') # UMAP on IC\n",
    "      \n",
    "    # 2.2) NON INTEGRATED ASSAY\n",
    "      multiomics=RunUMAP(multiomics,reduction='PCA_NO_INT' ,dims = 1:20,reduction.name = 'UMAP_PCA_NO_INT')  # UMAP on PC \n",
    "      multiomics=RunUMAP(multiomics,dims = 1:20, local.connectivity  = 6 ,reduction = 'ICA_NO_INT',umap.method = \"umap-learn\",metric = \"correlation\",n.neighbors =200,reduction.name = 'UMAP_ICA_NO_INT',reduction.key='icaUMAP_') # UMAP on IC\n",
    "\n",
    "  \n",
    "    # 2.3) Neighbors and Clustering on non integrated pc\n",
    "      multiomics=FindNeighbors(multiomics,reduction = 'UMAP_PCA_NO_INT',dims = 1:30) # take the right reduc \n",
    "      multiomics=FindClusters(multiomics, resolution = seq(from = 0.1, to = 1, by = 0.1),algorithm = 'leiden')# store cluster resolution in metadata slot\n",
    "\n",
    "          # Test a range of cluster resolution \n",
    "                ResolutionList <- grep(\"_snn_res\", colnames(multiomics@meta.data), value = TRUE)\n",
    "                for (Resolution in ResolutionList){\n",
    "                    #pdf(paste0(Resolution, \"_UMAP.pdf\"), width=7, height=7)\n",
    "                    g <- DimPlot(object = multiomics, reduction = \"umap\",label=T,group.by = Resolution  )\n",
    "                    print(g)\n",
    "                    # dev.off()\n",
    "                }\n",
    "          \n",
    "       \n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6119634f",
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "## Found variable genes and name clusters "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "730effbc",
   "metadata": {
    "name": "Annotation  and rename Idents"
   },
   "outputs": [],
   "source": [
    "\n",
    "# Find markers genes of clusters\n",
    "all_markers=FindAllMarkers(object = multiomics,assay ='SCT',logfc.threshold = 0.1)\n",
    "\n",
    "\n",
    "# Featureplot known markers genes \n",
    "FeaturePlot(multiomics, features=c('Wnt4','Dab2'),reduction = 'UMAP_PCA_INT')                                   # glomerulosa markers gene \n",
    "FeaturePlot(multiomics, features=c('Akr1b7','Cyp11b1'),reduction = 'UMAP_PCA_INT')                              # Fasciculata markers gene \n",
    "FeaturePlot(multiomics, features=c('Th','Slc18a1','Slc18a2','Ddc','Dbh','Lgr5'),reduction = 'UMAP_PCA_INT')     # medulla markers gene \n",
    "FeaturePlot(multiomics, features=c('Star','Cyp11a1', 'Hsd3b1', 'Scarb1'),reduction = 'UMAP_PCA_INT')                # cortex markers gene \n",
    "FeaturePlot(multiomics, features=c(\"Rspo3\",'Gli1'),reduction = 'UMAP_PCA_INT')                                  # capsular markers gene \n",
    "FeaturePlot(multiomics, features=c('Akr1c18'),reduction = 'UMAP_PCA_INT')                                       # x zone\n",
    "FeaturePlot(multiomics, features=c('Cdh5','Pecam1','Flt1'),reduction = 'UMAP_PCA_INT')                          # endothelial  markers gene \n",
    "FeaturePlot(multiomics, features=c('Star','Cyp11a1', 'Hsd3b1', 'Scarb1'),reduction = 'UMAP_PCA_INT')            # cortex markers gene \n",
    "FeaturePlot(multiomics, features=c('Ptprc'),reduction = 'UMAP_PCA_INT')                                         # C immutnitaire  markers gene\n",
    "FeaturePlot(multiomics, features=c(\"Ano1\",\"Col3a1\",\"Cdh11\",\"Dcn\",\"Rspo3\"),reduction = 'UMAP_PCA_INT')               # stromal cells \n",
    "\n",
    "\n",
    "\n",
    "Idents(multiomics)=multiomics@meta.data$Initial_clustering      # Save Old idents \n",
    "multiomics <- RenameIdents(multiomics,`Endothelial Cells`=\"Endothelial Cells\",\n",
    "                     `Male specific Medulla`='Male Medulla',\n",
    "                     `Male Fasciculata Replicate 1`='Male Fasciculata',\n",
    "                     `Male Fasciculata Replicate 2`='Male Fasciculata',\n",
    "                     `Female Fasiculata Replicate 2`='Female Fasciculata',\n",
    "                     `Female Glomerulosa`='Female Glomerulosa',\n",
    "                     `Female Fasciculate Replicate 1`='Female Fasciculata',\n",
    "                     `Male Glomerulosa`='Male Glomerulosa',\n",
    "                     `Female specific Medulla`=\"Male Fasciculata\",\n",
    "                     `Medulla progenitor Cells`=\"Female Medulla\",\n",
    "                     `Capsular Cells`='Medulla progenitor Cells',\n",
    "                     `Immune Cells`='Capsular Cells',\n",
    "                     `Fetal adrenocortical Cells`='Immune Cells',\n",
    "                     `X zone`='Fetal adrenocortical Cells')\n",
    "\n",
    "Idents(multiomics)=multiomics$Cell_types # Store clustering in metadata slot\n",
    "\n",
    "\n",
    "# check cell number\n",
    "table(Idents(multiomics))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f3bc2a53",
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "## Variable genes remove non interesing genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "00d8fcf2",
   "metadata": {
    "name": "Variable genes remove non interesing genes"
   },
   "outputs": [],
   "source": [
    "  DefaultAssay(multiomics)='SCT' \n",
    "  remove_multiomics=multiomics\n",
    "  remove_multiomics[['peaks']] <- NULL\n",
    "  counts <- GetAssayData(remove_multiomics, assay = \"SCT\")\n",
    "  counts <- counts[-(which(rownames(counts) %in% c('Xist','Tsix','Uty','Ddx3y'))),]\n",
    "  remove_multiomics <- subset(remove_multiomics, features = rownames(counts))\n",
    "  \n",
    "# male vs Female | fasciculata replicate\n",
    "multiomics_replicat=FindMarkers(remove_multiomics,assay = 'SCT',logfc.threshold = 0.1,ident.1 =c(\"Male Fasciculata Replicate 2\",\"Male Fasciculata Replicate 1\"),ident.2 = c( \"Female Fasciculata Replicate 1\", \"Female Fasiculata Replicate 2\"))\n",
    "multiomics_replicat=multiomics_replicat[multiomics_replicat$p_val_adj<0.05,]\n",
    "write.csv(multiomics_replicat,file = 'D:/ADRENAL_GLAND_PROJECT/FASCICULATA/replicate/diff_male_pos_female_neg_fasc.csv')\n",
    "\n",
    "\n",
    "# male vs male | fasciculata replicate\n",
    "diff_MALE_fasc=FindMarkers(remove_multiomics,assay = 'SCT',logfc.threshold = 0.1,ident.1 =c(\"Male Fasciculata Replicate 2\"),ident.2 = c(\"Male Fasciculata Replicate 1\"))\n",
    "diff_MALE_fasc=diff_MALE_fasc[diff_MALE_fasc$p_val_adj<0.05,]\n",
    "write.csv(diff_MALE_fasc,file = 'D:/ADRENAL_GLAND_PROJECT/FASCICULATA/replicate/diff_male_2_male_1_fasc.csv')\n",
    "\n",
    "\n",
    "# Female vs male | fasciculata replicate\n",
    "diff_FEMALE_fasc=FindMarkers(remove_multiomics,assay = 'SCT',logfc.threshold = 0.1,ident.1 =c(\"Female Fasiculata Replicate 2\"),ident.2 = c(\"Female Fasciculata Replicate 1\"))\n",
    "diff_FEMALE_fasc=diff_FEMALE_fasc[diff_FEMALE_fasc$p_val_adj<0.05,]\n",
    "write.csv(diff_FEMALE_fasc,file = 'D:/ADRENAL_GLAND_PROJECT/FASCICULATA/Replicate/diff_female_2_female_1_fasc.csv')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3090c399",
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "## ATAC peaks calling; normalisation; UMAP representation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ff6a1a8e",
   "metadata": {
    "name": "ATAC processing"
   },
   "outputs": [],
   "source": [
    "# From https://satijalab.org/signac/articles/pbmc_multiomic.html\n",
    "library(EnsDb.Mmusculus.v79)\n",
    "library(BSgenome.Mmusculus.UCSC.mm10)\n",
    "  \n",
    "#1) Peak calling \n",
    "      DefaultAssay(multiomics)='ATAC'\n",
    "      run_macs<- CallPeaks(object = multiomics,macs2.path = '/home/absta/.conda/envs/envpython/bin/macs2',effective.genome.size = 1.87e9 ,group.by = \"Cell_types\",assay = 'ATAC') \n",
    "\n",
    "      #get gene annotations for hg38\n",
    "      annotation <- GetGRangesFromEnsDb(ensdb = EnsDb.Mmusculus.v79)\n",
    "      seqlevelsStyle(annotation) <- \"UCSC\"\n",
    "\n",
    "\n",
    "      #remove peaks on nonstandard chromosomes and in genomic blacklist regions\n",
    "      peaks <- keepStandardChromosomes(run_macs, pruning.mode = \"coarse\",species ='Mus_musculus' )\n",
    "      peaks <- subsetByOverlaps(x = peaks, ranges = blacklist_mm10, invert = TRUE)\n",
    "\n",
    "      #quantify counts in each peak\n",
    "      macs2_counts <- FeatureMatrix(\n",
    "          fragments = Fragments(multiomics),\n",
    "          features = peaks,\n",
    "          cells = colnames(multiomics)\n",
    "        )\n",
    "\n",
    "\n",
    "#2) create a new assay (peaks) using the MACS2 peak set and add it to the Seurat object\n",
    "      multiomics[[\"peaks\"]] <- CreateChromatinAssay(\n",
    "        counts = macs2_counts,\n",
    "        fragments = Fragments(multiomics),\n",
    "        annotation = annotation\n",
    "      )\n",
    "      \n",
    "\n",
    "#3) ATAC Normalisation UMAP representation \n",
    "    # Normalisation\n",
    "    DefaultAssay(multiomics) <- \"peaks\"\n",
    "    multiomics <- FindTopFeatures(multiomics, min.cutoff = 5) # Find top features based on total number of counts for the feature\n",
    "    multiomics <- RunTFIDF(multiomics) # Run term frequency inverse document frequency (TF-IDF) normalization on a matrix.\n",
    "    multiomics <- RunSVD(multiomics) # Run partial singular value decomposition using irlba\n",
    "    DepthCor(multiomics)   # correlation between sequencing depth and first component \n",
    "\n",
    "multiomics <- RunUMAP(object = multiomics,reduction = 'lsi',dims = 2:50,reduction.name = 'lsi_umap') # UMAP visualization with RNA based idents as clusters\n",
    "#multiomics <- FindNeighbors(object = multiomics,reduction = 'lsi',dims = 2:50) # use SNN for findclusters \n",
    "#multiomics <- FindClusters(object = multiomics,algorithm = 1,resolution = 0.7,verbose = FALSE)\n",
    "\n",
    "DimPlot(object = multiomics, label = TRUE)+DimPlot(multiomics,reduction = 'UMAP_PCA_NO_INT')\n",
    "\n",
    "\n",
    "#4) INTEGRATION ON REPLICAT  for atac \n",
    "\n",
    "      Replicat=multiomics # create new variable\n",
    "      Replicat <- SplitObject(Replicat, split.by = \"Replicat\")  # Split Object by replicat (create list with 2 variables )\n",
    "\n",
    "      # Find top 3000 conserved peaks features between both variable \n",
    "      x=SelectIntegrationFeatures(               \n",
    "        Replicat,\n",
    "        nfeatures = 3000,\n",
    "        assay = c('peaks','peaks'),\n",
    "        verbose = TRUE\n",
    "      )\n",
    "\n",
    "      # find integration anchors ( common features)\n",
    "      integration.anchors <- FindIntegrationAnchors(\n",
    "        object.list = Replicat,\n",
    "        anchor.features =x,\n",
    "        reduction = \"rlsi\",\n",
    "        dims = 2:30\n",
    "      )\n",
    "      y=multiomics \n",
    "\n",
    "      # integrate LSI embeddings\n",
    "      integrated <- IntegrateEmbeddings(\n",
    "        anchorset = integration.anchors,\n",
    "        reductions = y[[\"lsi\"]],\n",
    "        new.reduction.name = \"integrated_lsi\",\n",
    "        dims.to.integrate = 2:30\n",
    "      )\n",
    "\n",
    "    integrated <- RunUMAP(integrated, reduction = \"integrated_lsi\", dims = 2:30)\n",
    "    DimPlot(integrated ,group.by = 'Replicat')\n",
    "       \n",
    " \n",
    "    # ADD INT_LSI into mainn object OBJECT \n",
    "\n",
    "    multiomics@reductions$UMAP_LSI_ATAC_INT=integrated@reductions[[\"umap\"]] # add UMAP LSI int into main object \n",
    "    multiomics@reductions$integrated_lsi=integrated@reductions[[\"integrated_lsi\"]] # Add LSI int into main object \n",
    "\n",
    "    # Visualisation \n",
    "    DimPlot(object = multiomics, label = TRUE,group.by = 'Replicat',reduction = 'lsi_umap')+DimPlot(object = multiomics, label = TRUE,group.by = 'Replicat',reduction = 'UMAP_LSI_ATAC_INT')\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9b17057b",
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "## WNN UMAP representation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "066820fe",
   "metadata": {
    "name": "wnn.umap CLUSTERING"
   },
   "outputs": [],
   "source": [
    "\n",
    "# 1) Joint UMAP visualization\n",
    "ElbowPlot(object, ndims = 20, reduction = \"integrated_lsi\")\n",
    "ElbowPlot(object, ndims = 20, reduction = \"PCA_INT\")\n",
    "\n",
    "\n",
    "multiomics <- FindMultiModalNeighbors(object = multiomics,reduction.list = list(\"PCA_INT\", \"integrated_lsi\"), dims.list = list(1:20,2:40),modality.weight.name = \"weighted.nn\",verbose = TRUE) # Create weighted nearest neighbors \n",
    "multiomics <- RunUMAP(object = multiomics,nn.name = \"weighted.nn\",assay = \"SCT\",verbose = TRUE,reduction.name = 'wnn.umap')\n",
    "\n",
    "DimPlot(object = multiomics, label = TRUE,reduction='wnn.umap')\n",
    "DimPlot(object = multiomics, label = TRUE,reduction='UMAP_PCA_NO_INT')\n",
    "DimPlot(object = multiomics, label = TRUE,reduction='UMAP_LSI_ATAC_INT')\n",
    "\n",
    "\n",
    "# 2) Findclusters on new representation \n",
    "multiomics=FindClusters(only_STERO, resolution = seq(from = 0.1, to = 2, by = 0.1),algorithm = 'leiden',graph.name = 'wsnn') # store cluster resolution in metadata slot\n",
    "  \n",
    "  # Test a range of cluster resolution \n",
    "ResolutionList <- grep(\"wsnn_res\", colnames(multiomics@meta.data), value = TRUE)\n",
    "for (Resolution in ResolutionList){\n",
    "    #pdf(paste0(Resolution, \"_UMAP.pdf\"), width=7, height=7)\n",
    "    g <- DimPlot(object = multiomics, reduction = \"wnn.umap\",label=T,group.by = Resolution  )\n",
    "    DefaultAssay(multiomics)='SCT_RUN_STERO'\n",
    "    assign(paste0('ALLMARKERS_',ResolutionList[Resolution]),FindAllMarkers(multiomics,assay ='SCT_RUN_STERO',logfc.threshold = 0.5))\n",
    "    print(g)\n",
    "    # dev.off()\n",
    "}\n",
    "\n",
    "\n",
    "DefaultAssay(multiomics)='SCT'\n",
    "FeaturePlot(multiomics, features = c(\"Ptprc\",\"Rspo3\",\"Th\",\"Pecam1\",\"Wnt4\",\"Akr1b7\",\"Akr1c18\",'Pnmt'),reduction = 'wnn.umap')  # gene markers by yasmine \n",
    "\n",
    "# Save \n",
    " # saveRDS(multiomics,file = 'MULTIOMICS_FINAL')\n",
    " # multiomics=readRDS(file='D:/Stage/aposimz/multiomics_final_object')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "83100514",
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "## CELL number"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3be98b4c",
   "metadata": {
    "name": "CELL number"
   },
   "outputs": [],
   "source": [
    "library(dplyr)\n",
    "\n",
    "\n",
    "# plot cell number \n",
    "df_idents <- as.data.frame(table(Idents(multiomics)))\n",
    "\n",
    "ggplot(df_idents, aes(reorder(Var1, -Freq), Freq))+\n",
    "geom_bar(stat = 'identity', aes(fill = Var1))+\n",
    "coord_flip()+\n",
    "labs(x=\"Sample\",y=\"Cells number\")\n",
    "\n",
    "\n",
    "# plot cell proportion \n",
    "pbmc_mix<-data.frame(sample=multiomics$orig.ident,population=Idents(multiomics))\n",
    "pbmc_df <- as.data.frame(table(pbmc_mix))\n",
    "pbmc_df <- pbmc_df %>% group_by(population) %>% dplyr::mutate(proportion = Freq/(sum(Freq)))\n",
    "\n",
    "ggplot(pbmc_df, aes(population, proportion))+\n",
    "geom_bar(stat = 'identity', aes(fill = sample), position = 'dodge')+theme(axis.text.x = element_text(angle = 45, hjust = 1))\n",
    "\n",
    "\n",
    "# other way \n",
    "pbmc_mix<-data.frame(sample=multiomics$orig.ident,population=Idents(multiomics))\n",
    "ggplot(pbmc_mix, aes(population))+coord_flip()+\n",
    "geom_bar(aes(fill = sample))+theme(axis.text.x = element_text(angle = 45, hjust = 1))\n",
    "\n",
    "\n",
    "\n",
    "# cell types sexual identity proportion\n",
    "DF_Count <- multiomics@meta.data  %>%group_by(cell_types,identity) %>%\n",
    "  count() %>%\n",
    "  ungroup() %>%\n",
    "  group_by(cell_types) %>%\n",
    "  mutate(Freq = n/sum(n)*100)\n",
    "\n",
    "ggplot(DF_Count, aes(x = cell_types, y = Freq, fill = identity))+\n",
    "  geom_col()+ geom_text(aes(label = paste(round(Freq, 2),\"%\")),position = position_stack(vjust = 0.5))+coord_flip()\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a829544d",
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "## Cell Cycle"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "48b42c32",
   "metadata": {
    "name": "CELL CYCLE"
   },
   "outputs": [],
   "source": [
    "DefaultAssay(multiomics)='SCT'\n",
    "m.g2m.genes=readRDS(file ='D:/Stage/aposimz/m.g2m.genes' )\n",
    "m.s.genes=readRDS(file ='D:/Stage/aposimz/m.s.genes' )\n",
    "\n",
    "#  add to the object meta data: S.Score, G2M.Score, and Phase\n",
    "multiomics <- CellCycleScoring(multiomics, s.features = m.s.genes, g2m.features = m.g2m.genes, set.ident = TRUE)\n",
    "DimPlot(multiomics,reduction = 'wnn.umap')\n",
    "\n",
    "\n",
    "DF_Count <- multiomics@meta.data  %>%group_by(cell_types,Phase) %>%\n",
    "  count() %>%\n",
    "  ungroup() %>%\n",
    "  group_by(cell_types) %>%\n",
    "  mutate(Freq = n/sum(n)*100)\n",
    "\n",
    "ggplot(DF_Count, aes(x = cell_types, y = Freq, fill = Phase))+\n",
    "  geom_col()+ geom_text(aes(label = paste(round(Freq, 2),\"%\")),position = position_stack(vjust = 0.5))+coord_flip()\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f64b26c0",
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "## Plot umap "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "19cbcfaa",
   "metadata": {
    "name": "PLOT umap"
   },
   "outputs": [],
   "source": [
    "library(ggplot2)\n",
    "\n",
    "Idents(multiomics)=multiomics$cell_types\n",
    "# PLOT UMAP\n",
    "DimPlot(multiomics,reduction = 'wnn.umap',pt.size = 1.66,group.by='Replicat') + guides(color = guide_legend(override.aes = list(size=5), ncol=1) ) + theme_classic(base_size = 15)| DimPlot(multiomics,reduction = 'wnn.umap',pt.size = 1.66,label = T) + guides(color = guide_legend(override.aes = list(size=5), ncol=1) ) + theme_classic(base_size = 15)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bd6ef895",
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "## plot genes markers| FeaturePlot "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4ea5369f",
   "metadata": {
    "name": "plot genes markers| FeaturePlot"
   },
   "outputs": [],
   "source": [
    "FeaturePlot(multiomics, features=c('Wnt4','Dab2'),reduction = 'pca_umap')                                   # glomerulosa markers gene \n",
    "FeaturePlot(multiomics, features=c('Akr1b7','Cyp11b1'),reduction = 'pca_umap')                              # Fasciculata markers gene \n",
    "FeaturePlot(multiomics, features=c('Th','Slc18a1','Slc18a2','Ddc','Dbh','Lgr5'),reduction = 'pca_umap')     # medulla markers gene \n",
    "FeaturePlot(multiomics, features=c(\"Rspo3\",'Gli1'),reduction = 'pca_umap')                                  # capsular ( progennitor compartment) markers gene \n",
    "FeaturePlot(multiomics, features=c('Akr1c18'),reduction = 'pca_umap')                                       # x zone\n",
    "FeaturePlot(multiomics, features=c('Cdh5','Pecam1','Flt1'),reduction = 'pca_umap')                          # endothelial  markers gene \n",
    "FeaturePlot(multiomics, features=c('Star','Cyp11a1', 'Hsd3b1', 'Scarb1'),reduction = 'wnn.umap')            # cortex markers gene \n",
    "FeaturePlot(multiomics, features=c('Csmd1'),reduction = 'pca_umap')                                         # capsular marker genes \n",
    "FeaturePlot(multiomics, features=c(\"Csmd1\",\"Ano1\",\"Col3a1\",\"Cdh11\",\"Dcn\",\"Rspo3\"),reduction = 'pca_umap')   # stromal cells"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "31572b08",
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "##  plot genes markers| DOTPLOT"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0d2a084b",
   "metadata": {
    "lines_to_next_cell": 2,
    "name": "plot genes markers| DOTPLOT"
   },
   "outputs": [],
   "source": [
    "# markers genes \n",
    "genes=c('Wnt4','Dab2','Akr1b7','Cyp11b1',\"Csmd1\",\"Rspo3\",\"Tcf21\", 'Flt1','Pecam1',\"Lsamp\",\"Lrrtm4\",'Th','Slc18a1',\"Cd74\",\"Cd84\",\"Akr1c18\",\"Pik3c2g\")\n",
    "\n",
    "dotx= DotPlot(multiomics,assay= 'SCT',features = genes,dot.scale = 6)+ RotatedAxis() +  scale_colour_gradient2(low = \"darkred\", mid = \"white\",high = \"darkgreen\",midpoint = 0,)\n",
    "\n",
    " # Markers genes plot \n",
    " # Main plot\n",
    "  minmaxthis = function(x) { \n",
    "    (x - min(x))/(max(x) - min(x))\n",
    "  }\n",
    "  \n",
    "  data.dot = dotx$data\n",
    "  \n",
    "  # scaled value\n",
    "  scaled_data <- \n",
    "    data.dot %>%\n",
    "    group_by(features.plot) %>%\n",
    "    mutate(minmaxscale = minmaxthis(avg.exp))\n",
    "  \n",
    "  scaled_data$avg.exp2 = ( scaled_data$avg.exp - min(scaled_data$avg.exp) )    / \n",
    "    (max(scaled_data$avg.exp)- min(scaled_data$avg.exp))\n",
    "  \n",
    "  \n",
    "  pmain <- ggplot(scaled_data)  + \n",
    "    geom_point(aes(x = features.plot, y = id, color = minmaxscale, size = pct.exp), stat = \"identity\") + \n",
    "    scale_size(range = c(0, 10), breaks  = seq(0,100,20), labels  = paste0(seq(0,100,20),\"%\"))  + \n",
    "    scale_color_gradient(low = \"white\",high = \"#0f4d91\",breaks = seq(0,1,0.2))  +\n",
    "    guides(color = guide_colourbar(title = \"Expression\", barwidth = 1, barheight = 10),\n",
    "           size = guide_legend(title = \"Cells (%)\",reverse=TRUE)) + \n",
    "    xlab(\"\") + ylab(\"\") + \n",
    "    \n",
    "    theme(panel.border=element_blank(), \n",
    "          axis.line = element_line(colour = 'black', size = 1), \n",
    "          axis.ticks = element_blank(),\n",
    "          axis.text = element_text(size=12, colour = \"black\"), \n",
    "          axis.text.x = element_text(angle = 90, hjust = 1),\n",
    "          panel.grid.major = element_blank(),\n",
    "          panel.grid.minor = element_blank(),\n",
    "          panel.background = element_blank(),\n",
    "          legend.key=element_blank(),\n",
    "          # legend.title=element_blank(),\n",
    "          legend.text=element_text(size=12))\n",
    "  \n",
    "pmain\n",
    "  \n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6a947ebd",
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "# only steroidogenic cell types dataset analysis ( only_stero )\n",
    "\n",
    "## ONLY STERO CELLS RNA/ATAC processing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b1bcb71c",
   "metadata": {
    "name": "RNA-ATAC computing"
   },
   "outputs": [],
   "source": [
    "table(Idents(only_STERO))\n",
    "# Subset\n",
    "DefaultAssay(multiomics)='RNA'\n",
    "#only_STERO <- subset(x=multiomics,idents=c(\"Male Fasciculata\",\"Female Fasciculata\",\"Female Glomerulosa\",\"Capsular Cells\",\"Male Glomerulosa\"))\n",
    "only_STERO <- subset(x=multiomics,idents=c(\"Male Fasciculata Replicate 1\",\"Male Fasciculata Replicate 2\",\"Female Fasiculata Replicate 2\",\"Female Glomerulosa\",\"Female Fasciculate Replicate 1\" ,\"Male Glomerulosa\",\"Capsular Cells\"))\n",
    "only_STERO=subset(only_STERO,subset  = nCount_ATAC<100000)\n",
    "only_STERO=subset(only_STERO,subset  = nCount_RNA<15000)\n",
    "#only_STERO <- only_STERO[,!colnames(only_STERO) %in% mixt_clusters] # process by first running to the clustering and highlinghting this cells\n",
    "#only_STERO <- only_STERO[,!colnames(only_STERO) %in% mixt_clusters_2] # process by first running to the clustering and highlinghting this cells\n",
    "\n",
    "       ####RNA PROCESSING### \n",
    "\n",
    "DefaultAssay(only_STERO)='RNA'  # take RNA assay \n",
    "\n",
    "only_STERO <- SplitObject(only_STERO, split.by = \"Replicat\") # list \n",
    "\n",
    "for (i in 1:length(only_STERO)) {\n",
    "    only_STERO[[i]] <- SCTransform(only_STERO[[i]],variable.features.n =NULL,variable.features.rv.th =1.1,return.only.var.genes = FALSE,assay = 'RNA',new.assay.name = 'SCT_INT',vars.to.regress = 'percent.mt')\n",
    "    }\n",
    "\n",
    "\n",
    "# Select the most variable features to use for integration\n",
    "integ_features <- SelectIntegrationFeatures(object.list = only_STERO, nfeatures = 3000) \n",
    "\n",
    "# Prepare the SCT list object for integration\n",
    "only_STERO  <- PrepSCTIntegration(object.list = only_STERO, \n",
    "                                   anchor.features =integ_features)\n",
    "\n",
    "\n",
    "integ_anchors <- FindIntegrationAnchors(object.list = only_STERO, \n",
    "                                        normalization.method = \"SCT\", \n",
    "                                        anchor.features = integ_features,reduction=\"cca\")\n",
    "\n",
    "# Integrate across conditions\n",
    "only_STERO <- IntegrateData(anchorset = integ_anchors,normalization.method = \"SCT\")                 # replace seurat list by object\n",
    "\n",
    "\n",
    "\n",
    "# only_STERO[['SCT']]=NULL\n",
    "# only_STERO[['peaks']]=NULL\n",
    "# only_STERO@meta.data[15:21]=NULL\n",
    "\n",
    "\n",
    "# CLustering \n",
    "DefaultAssay(only_STERO)='integrated'\n",
    "only_STERO=RunPCA(only_STERO, verbose = FALSE,assay=\"integrated\",reduction.name='PCA_INT',npcs = 50)\n",
    "ElbowPlot(only_STERO,reduction ='PCA_INT',ndims = 50 ) # choose 15\n",
    "only_STERO <- RunICA(only_STERO,verbose = TRUE,nics = 30,reduction.name='ICA_STERO',assay = \"integrated\")\n",
    "only_STERO=FindNeighbors(only_STERO,assay ='integrated',reduction ='PCA_INT', dims = 1:25)\n",
    "only_STERO=FindClusters(only_STERO, resolution = 0.5,algorithm = 4,graph.name = 'integrated_nn')\n",
    "only_STERO=RunUMAP(only_STERO,dims = 1:20, reduction.name = 'UMAP_PCA_RNA_INT',reduction = 'PCA_INT')\n",
    "\n",
    "# compute new assay based on RNA , it will be the new default assay for rna expression \n",
    "\n",
    "only_STERO=SCTransform(only_STERO,variable.features.n =NULL,variable.features.rv.th =1.1,return.only.var.genes = FALSE,assay = 'RNA',new.assay.name = 'SCT_RUN_STERO',vars.to.regress = 'percent.mt')\n",
    "DefaultAssay(only_STERO)='SCT_RUN_STERO'\n",
    "\n",
    "#subset_endo=WhichCells(object = only_STERO,idents = 5)\n",
    "#subset_medulla=WhichCells(object = only_STERO,idents = 9)\n",
    "\n",
    "\n",
    "FeaturePlot(only_STERO, features=c('Wnt4','Dab2'),reduction = 'UMAP_PCA_RNA_INT')                                   # glomerulosa markers gene \n",
    "FeaturePlot(only_STERO, features=c('Akr1b7','Cyp11b1'),reduction = 'UMAP_PCA_RNA_INT')                              # Fasciculata markers gene \n",
    "FeaturePlot(only_STERO, features=c('Th','Slc18a1','Slc18a2','Ddc','Dbh','Lgr5'),reduction = 'UMAP_PCA_RNA_INT')     # medulla markers gene \n",
    "FeaturePlot(only_STERO, features=c(\"Rspo3\",'Gli1'),reduction = 'UMAP_PCA_RNA_INT')                                  # capsular ( progennitor compartment) markers gene \n",
    "FeaturePlot(only_STERO, features=c('Akr1c18'),reduction = 'UMAP_PCA_RNA_INT')                                       # x zone\n",
    "FeaturePlot(only_STERO, features=c('Cdh5','Pecam1','Flt1'),reduction = 'UMAP_PCA_RNA_INT')                          # endothelial  markers gene \n",
    "FeaturePlot(only_STERO, features=c('Star','Cyp11a1', 'Hsd3b1', 'Scarb1'),reduction = 'UMAP_PCA_RNA_INT')            # cortex markers gene \n",
    "FeaturePlot(only_STERO, features=c('Csmd1'),reduction = 'UMAP_PCA_RNA_INT')                                         # capsular marker genes \n",
    "FeaturePlot(only_STERO, features=c('Ptprc'),reduction = 'UMAP_PCA_RNA_INT')                                         # C immutnitaire  markers gene \n",
    "\n",
    "\n",
    "DimPlot(only_STERO,reduction = 'UMAP_PCA_RNA_INT',label = T) | DimPlot(object = only_STERO, label = TRUE,group.by = 'identity',reduction = 'UMAP_PCA_RNA_INT') |DimPlot(object = only_STERO, label = TRUE,group.by = 'Replicat',reduction = 'UMAP_PCA_RNA_INT')\n",
    "\n",
    "################## NOT RUN\n",
    "#there is two clusters with endothelial markers genes , and no cortex markers genes , likely doublet , i remove them for the rest of the analysis \n",
    "# take ident of bad cluusters and subset ident 7 and 9\n",
    " #mixt_clusters=WhichCells(object = only_STERO,idents = c('7','9'))\n",
    "#DimPlot(only_STERO,cells.highlight = mixt_clusters_2)\n",
    "#FeaturePlot(only_STERO, features=c('Cdh5','Pecam1','Flt1'),reduction = 'UMAP_PCA_RNA_INT') # endothelial  markers gene \n",
    "# remove clusters 16 ==> medulla \n",
    "# mixt_clusters_2=WhichCells(object = only_STERO,idents = c('16'))\n",
    "###################NOT RUN\n",
    "\n",
    "\n",
    "findmarkers=FindAllMarkers(only_STERO,assay = 'SCT_RUN_STERO')\n",
    " # annotation cluster \n",
    "only_STERO@meta.data$old_ident=Idents(only_STERO)   # save idents number \n",
    "only_STERO <- RenameIdents(only_STERO,`1`='Male_Fasciculata_1',\n",
    "                     `2`='Male_Glomerulosa',\n",
    "                     `3`='Female_Fasciculata_1',\n",
    "                     `4`='Female_Glomerulosa',\n",
    "                     `5`='Female_Fasciculata_2',\n",
    "                     `6`='Male_Fasciculata_2',\n",
    "                     `7`='Capsular_Cells'\n",
    "                     )\n",
    "only_STERO@meta.data$new_ident=Idents(only_STERO)\n",
    "# only_STERO$cell_types =NULL        # remove from previous multiomics object\n",
    "DimPlot(only_STERO,reduction  = 'UMAP_PCA_RNA_INT',label = T)\n",
    "\n",
    "\n",
    "##### ############################################################################################ATAC\n",
    "       #### ATAC processing  AND PEAKS  generating ## \n",
    "##Peak calling \n",
    "library(EnsDb.Mmusculus.v79)\n",
    "library(BSgenome.Mmusculus.UCSC.mm10)\n",
    "DefaultAssay(only_STERO)='ATAC'\n",
    "\n",
    "run_macs_ONLY_STERO<- CallPeaks(object = only_STERO,macs2.path = '/home/absta/.conda/envs/envpython/bin/macs2',effective.genome.size = 1.87e9 ,group.by = \"new_ident\",assay = 'ATAC') \n",
    "\n",
    "# get gene annotations for hg38\n",
    "annotation <- GetGRangesFromEnsDb(ensdb = EnsDb.Mmusculus.v79)\n",
    "seqlevelsStyle(annotation) <- \"UCSC\"\n",
    "\n",
    "\n",
    "# remove peaks on nonstandard chromosomes and in genomic blacklist regions\n",
    "peaks_only_stero <- keepStandardChromosomes(run_macs_ONLY_STERO, pruning.mode = \"coarse\",species ='Mus_musculus' )\n",
    "peaks_only_stero <- subsetByOverlaps(x = peaks_only_stero, ranges = blacklist_mm10, invert = TRUE)\n",
    "\n",
    "# quantify counts in each peak\n",
    "macs2_counts <- FeatureMatrix(\n",
    "  fragments = Fragments(only_STERO),\n",
    "  features = peaks_only_stero,\n",
    "  cells = colnames(only_STERO)\n",
    ")\n",
    "\n",
    "\n",
    "# create a new assay using the MACS2 peak set and add it to the Seurat object\n",
    "only_STERO[[\"peaks\"]] <- CreateChromatinAssay(\n",
    "  counts = macs2_counts,\n",
    "  fragments = Fragments(only_STERO),\n",
    "  annotation = annotation\n",
    ")\n",
    "\n",
    "DefaultAssay(only_STERO) <- \"peaks\"\n",
    "only_STERO <- FindTopFeatures(only_STERO, min.cutoff = 5)\n",
    "only_STERO <- RunTFIDF(only_STERO)\n",
    "only_STERO <- RunSVD(only_STERO)\n",
    "DepthCor(only_STERO,n = 50)\n",
    "\n",
    "only_STERO <- RunUMAP(object = only_STERO,reduction = 'lsi',dims = 2:50,reduction.name = 'UMAP_LSI_ATAC')\n",
    "only_STERO <- FindNeighbors(object = only_STERO,reduction = 'lsi',dims = 2:50) # use SNN for findclusters \n",
    "#only_STERO <- FindClusters(object = only_STERO,algorithm = 1,resolution = 0.7,verbose = FALSE)\n",
    "DimPlot(object = only_STERO, label = TRUE,group.by = 'Replicat',reduction = 'UMAP_LSI_ATAC')+DimPlot(only_STERO,reduction = 'UMAP_PCA_RNA_INT',group.by = 'Replicat')\n",
    "\n",
    "##################################INTEGRATION ON REPLICAT  for atac \n",
    "\n",
    "\n",
    "crazy=only_STERO\n",
    "\n",
    "crazy <- SplitObject(crazy, split.by = \"Replicat\") # list \n",
    "\n",
    "\n",
    "x=SelectIntegrationFeatures(\n",
    "  crazy,\n",
    "  nfeatures = 3000,\n",
    "  assay = c('peaks','peaks'),\n",
    "  verbose = TRUE\n",
    ")\n",
    "\n",
    "\n",
    "# find integration anchors\n",
    "integration.anchors <- FindIntegrationAnchors(\n",
    "  object.list = crazy,\n",
    "  anchor.features =x,\n",
    "  reduction = \"rlsi\",\n",
    "  dims = 2:30\n",
    ")\n",
    "\n",
    "# integrate LSI embeddings\n",
    "integrated <- IntegrateEmbeddings(\n",
    "  anchorset = integration.anchors,\n",
    "  reductions = X1[[\"lsi\"]],\n",
    "  new.reduction.name = \"integrated_lsi\",\n",
    "  dims.to.integrate = 2:30\n",
    ")\n",
    "\n",
    "integrated <- RunUMAP(integrated, reduction = \"integrated_lsi\", dims = 2:30)\n",
    "\n",
    "DimPlot(integrated ,group.by = 'Replicat')\n",
    " \n",
    " \n",
    " #######################ADD INT LSI ON OBJECT \n",
    "\n",
    "only_STERO@reductions$UMAP_LSI_ATAC_INT=integrated@reductions[[\"umap\"]]\n",
    "only_STERO@reductions$integrated_lsi=integrated@reductions[[\"integrated_lsi\"]]\n",
    "\n",
    "\n",
    "\n",
    "DimPlot(object = only_STERO, label = TRUE,group.by = 'Replicat',reduction = 'UMAP_LSI_ATAC')+DimPlot(only_STERO,reduction = 'UMAP_PCA_RNA_INT',group.by = 'Replicat')+DimPlot(object = only_STERO, label = TRUE,group.by = 'Replicat',reduction = 'UMAP_LSI_ATAC_INT')\n",
    "\n",
    " \n",
    " "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a97e7d2f",
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "## ONLY STERO WNN UMAP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d3cae938",
   "metadata": {
    "name": "wnn.umap"
   },
   "outputs": [],
   "source": [
    "\n",
    "ElbowPlot(only_STERO,reduction = 'PCA_INT',ndims = 50)\n",
    "\n",
    "\n",
    "only_STERO <- FindMultiModalNeighbors(\n",
    "  only_STERO, reduction.list = list(\"PCA_INT\", \"integrated_lsi\"), \n",
    "  dims.list = list(1:20, 2:30),weighted.nn.name = 'X1'\n",
    ")\n",
    "only_STERO <- RunUMAP(only_STERO, nn.name = \"X1\", reduction.name = \"wnn.umap\", reduction.key = \"wnnUMAP_\")\n",
    "\n",
    "\n",
    "FeaturePlot(only_STERO, features=c('Wnt4','Dab2'),reduction = 'wnn.umap')                                   # glomerulosa markers gene \n",
    "FeaturePlot(only_STERO, features=c('Akr1b7','Cyp11b1'),reduction = 'wnn.umap')                              # Fasciculata markers gene \n",
    "FeaturePlot(only_STERO, features=c('Cdh5','Pecam1','Flt1'),reduction = 'wnn.umap')                          # endothelial  markers gene \n",
    "FeaturePlot(only_STERO, features=c('Star','Cyp11a1', 'Hsd3b1', 'Scarb1'),reduction = 'wnn.umap')            # cortex markers gene \n",
    "FeaturePlot(only_STERO, features=c('Csmd1'),reduction = 'wnn.umap')                                         # capsular marker genes \n",
    "DimPlot(only_STERO)# capsular marker genes \n",
    "\n",
    "\n",
    "only_STERO@meta.data$old_ident=Idents(only_STERO)\n",
    "only_STERO <- RenameIdents(only_STERO,`1`='Male_Fasciculata_1',\n",
    "                     `2`='Female_Fasciculata_2',\n",
    "                     `3`='Male_Glomerulosa',\n",
    "                     `4`='Female_Glomerulosa',\n",
    "                     `5`='Male_Fasciculata_2',\n",
    "                     `6`='Female_Fasciculata_1',\n",
    "                     `7`='Capsular_Cells'\n",
    "                        )\n",
    "only_STERO@meta.data$new_ident=Idents(only_STERO)\n",
    "\n",
    "\n",
    "DimPlot(only_STERO,reduction  = 'wnn.umap',label = T)\n",
    "\n",
    "\n",
    "\n",
    "Idents(only_STERO)=only_STERO@meta.data$old_ident\n",
    "only_STERO <- RenameIdents(only_STERO,`1`='Male_Fasciculata',\n",
    "                     `2`='Female_Fasciculata',\n",
    "                     `3`='Male_Glomerulosa',\n",
    "                     `4`='Female_Glomerulosa',\n",
    "                     `5`='Male_Fasciculata',\n",
    "                     `6`='Female_Fasciculata',\n",
    "                     `7`='Capsular_Cells'\n",
    "                        )\n",
    "only_STERO@meta.data$merging_ident=Idents(only_STERO)\n",
    "\n",
    "DimPlot(only_STERO,reduction  = 'wnn.umap',label = T,pt.size = 2)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "saveRDS(only_STERO,file = 'only_STERO_final')\n",
    "only_STERO=readRDS(file = 'D:/Users/Documents/aposimz/only_STERO_final')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5fed8a21",
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "## Remove sex chromosome ( only_stero )\n",
    "\n",
    "### ONLY STERO Remove SEX"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ea68e0c1",
   "metadata": {
    "name": "multimodal umap remove sex chromosome"
   },
   "outputs": [],
   "source": [
    "\n",
    "# 1) REMOVE SEX GENES \n",
    "      # need to remove previous Peaks assay\n",
    "\n",
    "      DefaultAssay(only_STERO)='RNA'\n",
    "      counts <- GetAssayData(only_STERO, assay = \"RNA\") # get genes from RNA \n",
    "      counts <- counts[-(which(rownames(counts) %in% chrx_y$chrX_and_Y)),] # remove sexual genes \n",
    "      remove_sex=only_STERO  # create other object \n",
    "      remove_sex[['peaks']] <- NULL\n",
    "      remove_sex@reductions[[\"UMAP_LSI_ATAC\"]] <- NULL\n",
    "      remove_sex@reductions[[\"UMAP_LSI_ATAC_INT\"]] <- NULL\n",
    "      remove_sex@assays$SCT_RUN_STERO<-NULL\n",
    "      remove_sex <- subset(remove_sex, features = rownames(counts))  # remove_sex= seurat object RNA ASSAY without sex genes \n",
    "\n",
    "\n",
    "# 2) remove peaks of X and Y \n",
    "      # need to remove previous RNA assay \n",
    "      \n",
    "      DefaultAssay(only_STERO)='peaks'\n",
    "      \n",
    "      peaks=only_STERO\n",
    "      peaks[['RNA']] <- NULL\n",
    "      counts_peaks=as.data.frame(rownames(peaks)) #  take all the peaks\n",
    "      counts_peak_grepXY=counts_peaks$`rownames(peaks)`[!grepl(\"chrX|Y.*\\\\d\",counts_peaks$`rownames(peaks)`)] # subset  peaks non beginning by chrX | chrY\n",
    "      peaks <- peaks[rownames(peaks) %in% counts_peak_grepXY, ]\n",
    "      peaks <- GetAssayData(object =  peaks[['peaks']])\n",
    " \n",
    "           \n",
    "# 3) ADD peaks assay with no sex peaks in remove_sex\n",
    "      remove_sex[[\"peaks_no_sex\"]] <- peaks@assays[[\"peaks\"]]  # peaks_no_sex\n",
    "      remove_sex=RenameAssays(remove_sex,RNA='RNA_no_sex')     # RNA_no_sex\n",
    "\n",
    "\n",
    "### RNA process \n",
    "  # 1) Replicat integration\n",
    "      \n",
    "      DefaultAssay(remove_sex)='RNA_no_sex'  # take RNA assay \n",
    "      remove_sex <- SplitObject(remove_sex, split.by = \"Replicat\") # list \n",
    "      for (i in 1:length(remove_sex)) {\n",
    "          remove_sex[[i]] <- SCTransform(remove_sex[[i]],variable.features.n =NULL,variable.features.rv.th =1.1\n",
    "                                         ,return.only.var.genes = FALSE,assay = 'RNA_no_sex',\n",
    "                                         new.assay.name = 'SCT_INT',vars.to.regress = 'percent.mt')\n",
    "          }\n",
    "      \n",
    "      \n",
    "      # Select the most variable features to use for integration\n",
    "      integ_features <- SelectIntegrationFeatures(object.list = remove_sex, nfeatures = 3000) \n",
    "      \n",
    "      # Prepare the SCT list object for integration\n",
    "      remove_sex  <- PrepSCTIntegration(object.list = remove_sex, \n",
    "                                         anchor.features =integ_features)\n",
    "      \n",
    "      \n",
    "      integ_anchors <- FindIntegrationAnchors(object.list = remove_sex, \n",
    "                                              normalization.method = \"SCT\", \n",
    "                                              anchor.features = integ_features,reduction=\"cca\")\n",
    "      \n",
    "      # Integrate across conditions\n",
    "      remove_sex <- IntegrateData(anchorset = integ_anchors,normalization.method = \"SCT\")                 \n",
    "      \n",
    "  \n",
    "\n",
    "   # 2)  CLustering on integrated assay \n",
    "      \n",
    "      DefaultAssay(remove_sex)='integrated'\n",
    "      remove_sex=RunPCA(remove_sex, verbose = FALSE,assay=\"integrated\",reduction.name='PCA_INT',npcs = 50)\n",
    "      ElbowPlot(remove_sex,reduction ='PCA_INT',ndims = 50 ) # choose 15\n",
    "      remove_sex <- RunICA(remove_sex,verbose = TRUE,nics = 30,reduction.name='ICA_STERO',assay = \"integrated\")\n",
    "      remove_sex=FindNeighbors(remove_sex,assay ='integrated',reduction ='PCA_INT', dims = 1:25)\n",
    "      remove_sex=FindClusters(remove_sex, resolution = 0.5,algorithm = 4,graph.name = 'integrated_nn')\n",
    "      remove_sex=RunUMAP(remove_sex,dims = 1:20, reduction.name = 'UMAP_PCA_RNA_INT',reduction = 'PCA_INT')\n",
    "\n",
    "    # compute new assay based on RNA_no_sex , it will be the new default assay for rna expression \n",
    "        \n",
    "        remove_sex=SCTransform(remove_sex,variable.features.n =NULL,variable.features.rv.th =1.1,return.only.var.genes = FALSE,assay = 'RNA_no_sex',new.assay.name = 'SCT_RUN_STERO',vars.to.regress = 'percent.mt')\n",
    "        DefaultAssay(remove_sex)='SCT_RUN_STERO'\n",
    "\n",
    "\n",
    "# ATAC process   \n",
    "    # Normalisation and reduction\n",
    "      DefaultAssay(remove_sex) <- \"peaks_no_sex\"\n",
    "      remove_sex <- FindTopFeatures(remove_sex, min.cutoff = 5)\n",
    "      remove_sex <- RunTFIDF(remove_sex)\n",
    "      remove_sex <- RunSVD(remove_sex)\n",
    "      DepthCor(remove_sex,n = 50)\n",
    "\n",
    "      remove_sex <- RunUMAP(object = remove_sex,reduction = 'lsi',dims = 2:50,reduction.name = 'UMAP_LSI_ATAC')\n",
    "      remove_sex <- FindNeighbors(object = remove_sex,reduction = 'lsi',dims = 2:50) # use SNN for findclusters \n",
    "      #only_STERO <- FindClusters(object = only_STERO,algorithm = 1,resolution = 0.7,verbose = FALSE)\n",
    "      DimPlot(object = remove_sex, label = TRUE,group.by = 'Replicat',reduction = 'UMAP_LSI_ATAC')+DimPlot(remove_sex,reduction = 'UMAP_PCA_RNA_INT',group.by = 'Replicat')\n",
    "\n",
    "    #INTEGRATION ON REPLICAT  for atac \n",
    "\n",
    "      Replicat=remove_sex\n",
    "       # Split Object \n",
    "      Replicat <- SplitObject(Replicat, split.by = \"Replicat\") \n",
    "      \n",
    "      # Find top 3000 conserved features \n",
    "      x=SelectIntegrationFeatures(               \n",
    "        Replicat,\n",
    "        nfeatures = 3000,\n",
    "        assay = c('peaks','peaks'),\n",
    "        verbose = TRUE\n",
    "      )\n",
    "\n",
    "\n",
    "      # find integration anchors\n",
    "      integration.anchors <- FindIntegrationAnchors(\n",
    "        object.list = Replicat,\n",
    "        anchor.features =x,\n",
    "        reduction = \"rlsi\",\n",
    "        dims = 2:30\n",
    "      )\n",
    "      X1=remove_sex \n",
    "\n",
    "      # integrate LSI embeddings\n",
    "      integrated <- IntegrateEmbeddings(\n",
    "        anchorset = integration.anchors,\n",
    "        reductions = X1[[\"lsi\"]],\n",
    "        new.reduction.name = \"integrated_lsi\",\n",
    "        dims.to.integrate = 2:30\n",
    "      )\n",
    "\n",
    "    integrated <- RunUMAP(integrated, reduction = \"integrated_lsi\", dims = 2:30)\n",
    "    DimPlot(integrated ,group.by = 'Replicat')\n",
    "       \n",
    " \n",
    "    # ADD INT_LSI into OBJECT \n",
    "\n",
    "    remove_sex@reductions$UMAP_LSI_ATAC_INT=integrated@reductions[[\"umap\"]] # add UMAP int\n",
    "    remove_sex@reductions$integrated_lsi=integrated@reductions[[\"integrated_lsi\"]] # Add LSI int  \n",
    "\n",
    "    # Visualisation \n",
    "    DimPlot(object = remove_sex, label = TRUE,group.by = 'Replicat',reduction = 'UMAP_LSI_ATAC')+DimPlot(remove_sex,reduction = 'UMAP_PCA_RNA_INT',group.by = 'Replicat')+DimPlot(object = remove_sex, label = TRUE,group.by = 'Replicat',reduction = 'UMAP_LSI_ATAC_INT')\n",
    "\n",
    "\n",
    "    # Joint UMAP visualization\n",
    "   \n",
    "    # non integrated multimodal umap    \n",
    "    remove_sex <- FindMultiModalNeighbors(object = remove_sex,reduction.list = list(\"pca\", \"lsi\"), dims.list = list(1:20,2:40),modality.weight.name = \"weighted.nn\",verbose = TRUE)\n",
    "    ElbowPlot(remove_sex,ndims = 50)\n",
    "  remove_sex <- RunUMAP(object = remove_sex,nn.name = \"weighted.nn\",assay = \"SCT\",verbose = TRUE,reduction.name = 'wnn.umap')\n",
    "  \n",
    "  # integrated multimodal umap \n",
    "\n",
    "    remove_sex <- FindMultiModalNeighbors(\n",
    "      remove_sex, reduction.list = list(\"PCA_INT\", \"integrated_lsi\"), \n",
    "      dims.list = list(1:20, 2:30),weighted.nn.name = 'X1'\n",
    "    )\n",
    "    remove_sex <- RunUMAP(remove_sex, nn.name = \"X1\", reduction.name = \"wnn.umap\", reduction.key = \"wnnUMAP_\")\n",
    "\n",
    "    DimPlot(remove_sex,reduction = 'wnn.umap')\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5714a6c9",
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "## not run "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3a5445b9",
   "metadata": {
    "name": "not run"
   },
   "outputs": [],
   "source": [
    "# markers genes \n",
    " # 0) all_markers\n",
    "      all_markers_replicat=FindAllMarkers(multiomics,assay = 'SCT_RUN_STERO',logfc.threshold = 0.1)\n",
    "    # Write as csv \n",
    "      write.csv(all_markers$gene[all_markers$p_val_adj<0.05&all_markers$avg_log2FC>1&all_markers$cluster==\"Immune Cells\"],file = 'all_markers.txt',quote = F,row.names = F,col.names = F)\n",
    "      \n",
    "      \n",
    "      \n",
    "      \n",
    "      \n",
    "      \n",
    "      \n",
    "      multiomics <- RenameIdents(multiomics,`Endothelial Cells`='Endothelial Cells',\n",
    "                     `Male specific Medulla`='Male Medulla',\n",
    "                     `Male Fasciculata Replicate 1`='Male Fasciculata Replicate 1',\n",
    "                     `Male Fasciculata Replicate 2`='Male Fasciculata Replicate 2',\n",
    "                     `Female Fasiculata Replicate 2`='Female Fasciculata Replicate 2',\n",
    "                     `Female Glomerulosa`='Female Glomerulosa',\n",
    "                     `Female Fasciculate Replicate 1`='Female Fasciculata Replicate 1',\n",
    "                     `Male Glomerulosa`='Male Glomerulosa',\n",
    "                     `Female specific Medulla`='Female Medulla',\n",
    "                     `Medulla progenitor Cells`='Medulla progenitor Cells',\n",
    "                     `Capsular Cells`='Capsular Cells',`Immune Cells`='Immune Cells',\n",
    "                     `Fetal adrenocortical Cells `= 'Stromal cells',\n",
    "                     `X zone `='X zone')\n",
    "      \n",
    "      multiomics@meta.data$cell_types=Idents(multiomics)\n",
    "      \n",
    "      \n",
    "      \n",
    "    # 0.bis) conserved markers\n",
    "\n",
    "      FindConservedMarkers(pbmc_small, ident.1 = 0, ident.2 = 1, grouping.var = \"groups\")\n",
    "\n",
    "      all_conserved=FindMarkers(object = multiomics,ident.1 = c(\"Male Glomerulosa\",\n",
    "                                      \"Male Fasciculata Replicate 1\",\n",
    "                                      \"Male Fasciculata Replicate 2\",\n",
    "                                      \"Female Fasciculata Replicate 1\",\n",
    "                                      \"Female Fasiculata Replicate 2\"))\n",
    "      \n",
    "      all_conserved$gene=rownames(all_conserved)\n",
    " # 1) reorder clusters :\n",
    "        # reorder clusters\n",
    "      # full dataset\n",
    "        multiomics@active.ident <- factor(multiomics@active.ident,   \n",
    "                                    levels=c(\"Female Glomerulosa\",\n",
    "                                              \"Male Glomerulosa\",\n",
    "                                              \"Male Fasciculata Replicate 1\",\n",
    "                                              \"Male Fasciculata Replicate 2\",\n",
    "                                              \"Female Fasciculata Replicate 1\",\n",
    "                                              \"Female Fasiculata Replicate 2\",\n",
    "                                              \"Capsular Cells\",\n",
    "                                              \"Stromal cells\",\n",
    "                                              \"Endothelial Cells\",\n",
    "                                              \"Medulla progenitor Cells\",\n",
    "                                              \"Male Medulla\",\n",
    "                                              \"Female Medulla\",\n",
    "                                              \"Immune Cells\",\n",
    "                                               \"X zone\"))\n",
    "\n",
    "\n",
    "    # only_STERO\n",
    "          only_STERO@active.ident <- factor(only_STERO@active.ident, \n",
    "                            levels=c(\"Male_Glomerulosa\",\n",
    "                                      \"Female_Glomerulosa\",\n",
    "                                      \"Female_tZone\",\n",
    "                                      \"Male_tZone\",\n",
    "                                      \"Male_Fasciculata\",\n",
    "                                      \"Female_Fasciculata\",\n",
    "                                      \"Capsular_cells\"))\n",
    "\n",
    "\n",
    "# FIND MARKERS REPLICATE DIFFERENCES : \n",
    "  \n",
    "Replicat=FindMarkers(multiomics,ident.1 =\"Female Fasciculata Replicate 1\",ident.2 = \"Female Fasiculata Replicate 2\",assay = 'SCT' )\n",
    "Replicat$genes=rownames(Replicat)\n",
    "write.table(Replicat$genes,file = 'ReplicatFEMALE.txt',quote = F,col.names = F,row.names = F)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "722da8f7",
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "## Cell cycle"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "26a1f02b",
   "metadata": {
    "name": "CELL CYCLE"
   },
   "outputs": [],
   "source": [
    "DefaultAssay(only_STERO)='SCT_RUN_STERO'\n",
    "m.g2m.genes=readRDS(file ='D:/Stage/aposimz/m.g2m.genes' )\n",
    "m.s.genes=readRDS(file ='D:/Stage/aposimz/m.s.genes' )\n",
    "\n",
    "#  add to the object meta data: S.Score, G2M.Score, and Phase\n",
    "multiomics <- CellCycleScoring(multiomics, s.features = m.s.genes, g2m.features = m.g2m.genes, set.ident = TRUE)\n",
    "DimPlot(multiomics,reduction = 'wnn.umap')\n",
    "\n",
    "\n",
    "DF_Count <- multiomics@meta.data  %>%group_by(cell_types,Phase) %>%\n",
    "  count() %>%\n",
    "  ungroup() %>%\n",
    "  group_by(cell_types) %>%\n",
    "  mutate(Freq = n/sum(n)*100)\n",
    "\n",
    "ggplot(DF_Count, aes(x = cell_types, y = Freq, fill = Phase))+\n",
    "  geom_col()+ geom_text(aes(label = paste(round(Freq, 2),\"%\")),position = position_stack(vjust = 0.5))+coord_flip()\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5745865c",
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "## Cell number "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4df9ab91",
   "metadata": {
    "name": "CELL number"
   },
   "outputs": [],
   "source": [
    "\n",
    "Idents(only_STERO)=only_STERO@meta.data$leiden_annotated\n",
    "\n",
    "# plot cell number \n",
    "df_idents <- as.data.frame(table(Idents(only_STERO)))\n",
    "\n",
    "ggplot(df_idents, aes(reorder(Var1, -Freq), Freq))+\n",
    "geom_bar(stat = 'identity', aes(fill = Var1))+\n",
    "coord_flip()+\n",
    "labs(x=\"Sample\",y=\"Cells number\")\n",
    "\n",
    "\n",
    "# plot cell proportion \n",
    "pbmc_mix<-data.frame(sample=only_STERO$orig.ident,population=Idents(only_STERO))\n",
    "pbmc_df <- as.data.frame(table(pbmc_mix))\n",
    "pbmc_df <- pbmc_df %>% group_by(population) %>% dplyr::mutate(proportion = Freq/(sum(Freq)))\n",
    "\n",
    "ggplot(pbmc_df, aes(population, proportion))+\n",
    "geom_bar(stat = 'identity', aes(fill = sample), position = 'dodge')+theme(axis.text.x = element_text(angle = 45, hjust = 1))\n",
    "\n",
    "\n",
    "# other way \n",
    "pbmc_mix<-data.frame(sample=only_STERO$orig.ident,population=Idents(only_STERO))\n",
    "ggplot(pbmc_mix, aes(population))+coord_flip()+\n",
    "geom_bar(aes(fill = sample))+theme(axis.text.x = element_text(angle = 45, hjust = 1))\n",
    "\n",
    "\n",
    "\n",
    "# cell types sexual identity proportion\n",
    "DF_Count <- only_STERO@meta.data  %>%group_by(leiden_annotated,identity) %>%\n",
    "  count() %>%\n",
    "  ungroup() %>%\n",
    "  group_by(leiden_annotated) %>%\n",
    "  mutate(Freq = n/sum(n)*100)\n",
    "\n",
    "ggplot(DF_Count, aes(x = leiden_annotated, y = Freq, fill = identity))+\n",
    "  geom_col()+ geom_text(aes(label = paste(round(Freq, 2),\"%\")),position = position_stack(vjust = 0.5))+coord_flip()\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b0cb2a0b",
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "## Scanpy clustering | Paga | FDG ( PYTHON )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9b58c738",
   "metadata": {
    "name": "Scanpy clustering | Paga | FDG"
   },
   "outputs": [],
   "source": [
    "%%python\n",
    "\n",
    "\n",
    "\n",
    "import scanpy  # pip install in  R terminal\n",
    "import scanpy as sc\n",
    "import numpy as np\n",
    "\n",
    "print(np.__version__)\n",
    "import pandas as pd\n",
    "from matplotlib import rcParams\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import networkx as nwx\n",
    "import os\n",
    "\n",
    "anndata = sc.read_h5ad('C:/Users/mehdi/Downloads/multiomics_convert(1).h5ad')\n",
    "\n",
    "male = sc.read_h5ad('D:/TRUTH/male_STERO.h5ad')\n",
    "del male.uns['neighbors']\n",
    "del male.obsp['distances']\n",
    "del male.varm['ICA_STERO']\n",
    "del male.obsm['X_ICA_STERO']\n",
    "del male.obsm['X_UMAP_LSI_ATAC']\n",
    "del male.obsm['X_UMAP_PCA_RNA_INT']\n",
    "del male.obsm['X_UMAP_LSI_ATAC_INT']\n",
    "# del anndata.obsm['X_PCA_INT']\n",
    "del male.varm['PCA_INT']\n",
    "del male.obsm['X_wnn.umap']\n",
    "\n",
    "sc.pp.neighbors(male, n_neighbors=50, use_rep='X_PCA_INT')\n",
    "# sc.tl.leiden(anndata, resolution = 1)\n",
    "sc.tl.paga(male, groups='leiden_merging')\n",
    "sc.pl.paga(male)\n",
    "male.uns['iroot'] = np.flatnonzero(male.obs['leiden_merging'] == 'Male_Glomerulosa')[0]\n",
    "sc.tl.draw_graph(male, init_pos='paga')\n",
    "sc.tl.dpt(male)\n",
    "sc.pl.draw_graph(male, color=['leiden_merging', 'dpt_pseudotime', 'leiden', 'identity'], legend_loc='on data')\n",
    "\n",
    "dpt_pseudotime = pd.DataFrame(index=male.obs.index.tolist(),\n",
    "                              columns=[\"Pseudotime\"],\n",
    "                              data=male.obs[\"dpt_pseudotime\"].tolist())\n",
    "\n",
    "dpt_pseudotime.to_csv(\"D:/TRUTH/Pseudotime_male.csv\")\n",
    "\n",
    "female = sc.read_h5ad('D:/TRUTH/female_STERO.h5ad')\n",
    "del female.uns['neighbors']\n",
    "del female.obsp['distances']\n",
    "del female.varm['ICA_STERO']\n",
    "del female.obsm['X_ICA_STERO']\n",
    "del female.obsm['X_UMAP_LSI_ATAC']\n",
    "del female.obsm['X_UMAP_PCA_RNA_INT']\n",
    "del female.obsm['X_UMAP_LSI_ATAC_INT']\n",
    "# del anndata.obsm['X_PCA_INT']\n",
    "del female.varm['PCA_INT']\n",
    "del female.obsm['X_wnn.umap']\n",
    "\n",
    "sc.pp.neighbors(female, n_neighbors=50, use_rep='X_PCA_INT')\n",
    "# sc.tl.leiden(anndata, resolution = 1)\n",
    "sc.tl.paga(female, groups='leiden_merging')\n",
    "sc.pl.paga(female)\n",
    "female.uns['iroot'] = np.flatnonzero(female.obs['leiden_merging'] == 'Female_Glomerulosa')[0]\n",
    "sc.tl.draw_graph(female, init_pos='paga')\n",
    "sc.tl.dpt(female)\n",
    "sc.pl.draw_graph(female, color=['leiden_merging', 'dpt_pseudotime', 'leiden', 'identity'], legend_loc='on data')\n",
    "\n",
    "dpt_pseudotime = pd.DataFrame(index=female.obs.index.tolist(),\n",
    "                              columns=[\"Pseudotime\"],\n",
    "                              data=female.obs[\"dpt_pseudotime\"].tolist())\n",
    "\n",
    "dpt_pseudotime.to_csv(\"D:/TRUTH/Pseudotime_female.csv\")\n",
    "\n",
    "del anndata.uns['neighbors']\n",
    "del anndata.obsp['distances']\n",
    "del anndata.varm['ICA_STERO']\n",
    "del anndata.obsm['X_ICA_STERO']\n",
    "del anndata.obsm['X_UMAP_LSI_ATAC']\n",
    "del anndata.obsm['X_UMAP_PCA_RNA_INT']\n",
    "del anndata.obsm['X_UMAP_LSI_ATAC_INT']\n",
    "# del anndata.obsm['X_PCA_INT']\n",
    "del anndata.varm['PCA_INT']\n",
    "\n",
    "sc.pp.neighbors(anndata, n_neighbors=50, use_rep='X_PCA_INT')\n",
    "# sc.tl.leiden(anndata, resolution = 1)\n",
    "\n",
    "\n",
    "sc.tl.paga(anndata, groups='leiden_annotated')\n",
    "sc.pl.paga(anndata, threshold=0.09)\n",
    "anndata.uns['iroot'] = np.flatnonzero(anndata.obs['leiden_merging'] == 'Capsular_cells')[0]\n",
    "sc.tl.draw_graph(anndata,\n",
    "                 init_pos='paga')\n",
    "\n",
    "# root =np.flatnonzero(anndata.obs['leiden_merging']=='Capsular_cells')[0])\n",
    "\n",
    "sc.tl.dpt(anndata)\n",
    "sc.pl.draw_graph(anndata, color=['merging_ident', 'dpt_pseudotime', 'leiden', 'identity'], legend_loc='on data')\n",
    "\n",
    "# plot\n",
    "\n",
    "myColors = ['#e6194b', '#3cb44b', '#ffe119', '#4363d8', '#f58231',\n",
    "            '#911eb4', '#46f0f0', '#f032e6', '#bcf60c', '#fabebe',\n",
    "            '#008080', '#e6beff', '#9a6324', '#000000', '#800000',\n",
    "            '#aaffc3', '#808000', '#ffd8b1', '#000075', '#808080',\n",
    "            '#ffffff', '#fffac8']\n",
    "\n",
    "f, axs = plt.subplots(1, 2, figsize=(36, 12))\n",
    "sns.set(font_scale=1.5)\n",
    "sns.set_style(\"white\")\n",
    "\n",
    "sc.pl.draw_graph(anndata,\n",
    "                 color='merging_ident',\n",
    "                 palette=myColors,\n",
    "                 legend_loc='on data',\n",
    "                 show=True,\n",
    "                 size=100,\n",
    "                 legend_fontsize='large')\n",
    "\n",
    "sc.pl.draw_graph(anndata,\n",
    "                 color='dpt_pseudotime',\n",
    "                 legend_loc='on data',\n",
    "                 show=True,\n",
    "                 size=400,\n",
    "                 color_map=\"viridis\",\n",
    "                 legend_fontsize='large')\n",
    "\n",
    "sns.despine(offset=10, trim=False)\n",
    "plt.tight_layout()\n",
    "plt.show(block=False)\n",
    "plt.close(\"all\")\n",
    "\n",
    "# save value\n",
    "\n",
    "dpt_pseudotime = pd.DataFrame(index=anndata.obs.index.tolist(),\n",
    "                              columns=[\"Pseudotime\"],\n",
    "                              data=anndata.obs[\"dpt_pseudotime\"].tolist())\n",
    "\n",
    "dpt_pseudotime.to_csv(\"D:/Stage/Pseudotime_RNA.csv\")\n",
    "\n",
    "FA2_comp = pd.DataFrame(index=anndata.obs.index.tolist(),\n",
    "                        columns=[\"FA1\", \"FA2\"],\n",
    "                        data=anndata.obs[\"X_draw_graph_fa\"])\n",
    "\n",
    "FA2_comp.to_csv(\"D:/Stage/Pseudotime_RNA.csvFA2_components_ATAC.csv\")\n",
    "\n",
    "# save paga embending\n",
    "clust_embedding = pd.DataFrame(anndata.uns['paga']['pos'], columns=['X', 'Y'])\n",
    "clust_embedding['Cluster'] = range(clust_embedding.shape[0])\n",
    "clust_embedding = clust_embedding[['Cluster', 'X', 'Y']]\n",
    "\n",
    "clust_embedding.to_csv('D:/Stage/TRUTH/PAGA/cluster_embedding.csv'),\n",
    "index = False)\n",
    "\n",
    "# save object\n",
    "\n",
    "\n",
    "out_dir = 'C:/Users/mehdi/Downloads'\n",
    "if not os.path.exists(out_dir):\n",
    "    os.makedirs(out_dir)\n",
    "\n",
    "\n",
    "def con2edges(con, names=None, sparse=True):\n",
    "    print('Converting connectivity matrix to edges...')\n",
    "    n = con.shape[0]\n",
    "    edges = pd.DataFrame(columns=['From', 'To', 'Connectivity'])\n",
    "\n",
    "    for i in range(n):\n",
    "        for j in range(i + 1, n):\n",
    "            if names is not None:\n",
    "                fr = names[i]\n",
    "                to = names[j]\n",
    "            else:\n",
    "                fr = str(i)\n",
    "                to = str(j)\n",
    "\n",
    "            connectivity = con[i, j]\n",
    "            if sparse and connectivity == 0:\n",
    "                continue\n",
    "\n",
    "            entry = {'From': fr, 'To': to,\n",
    "                     'Connectivity': con[i, j]}\n",
    "            edges = edges.append(entry, ignore_index=True)\n",
    "\n",
    "    return edges\n",
    "\n",
    "\n",
    "clust_con = anndata.uns['paga']['connectivities'].toarray()\n",
    "clust_edges = con2edges(clust_con)\n",
    "clust_edges.to_csv(os.path.join(out_dir, 'cluster_edges.csv'), index=False)\n",
    "\n",
    "clust_tree_con = anndata.uns['paga']['connectivities_tree'].toarray()\n",
    "clust_tree_edges = con2edges(clust_tree_con)\n",
    "clust_tree_edges.to_csv(os.path.join(out_dir, 'cluster_tree_edges.csv'), index=False)\n",
    "\n",
    "clust_embedding = pd.DataFrame(anndata.uns['paga']['pos'], columns=['X', 'Y'])\n",
    "clust_embedding['Cluster'] = range(clust_embedding.shape[0])\n",
    "clust_embedding = clust_embedding[['Cluster', 'X', 'Y']]\n",
    "\n",
    "clust_embedding.to_csv(os.path.join(out_dir, 'cluster_embedding.csv'), index=False)\n",
    "\n",
    "# cells edges\n",
    "cells = anndata.obs['index']\n",
    "cell_con = anndata.uns['neighbors']['connectivities']\n",
    "cell_edges = pd.DataFrame(columns=['From', 'To', 'Connectivity'])\n",
    "n_rows = len(cell_con.indptr)\n",
    "\n",
    "for i in range(len(cell_con.indptr) - 1):\n",
    "    row_ind = cell_con.indices[cell_con.indptr[i]:cell_con.indptr[i + 1]]\n",
    "    print(f'\\r\\tRow {i} of {n_rows}', end='')\n",
    "    for k, j in enumerate(row_ind):\n",
    "        if j > i:\n",
    "            con = cell_con.data[cell_con.indptr[i] + k]\n",
    "            fr = cells[i]\n",
    "            to = cells[j]\n",
    "            entry = {'From': fr, 'To': to, 'Connectivity': con}\n",
    "            cell_edges = cell_edges.append(entry, ignore_index=True)\n",
    "print('\\n')\n",
    "\n",
    "cell_edges.to_csv(os.path.join(out_dir, 'cell_edges.csv'), index=False)\n",
    "\n",
    "print('outputting cell embedding...')\n",
    "x = anndata.obsm['X_draw_graph_fr'][:, 0]\n",
    "y = anndata.obsm['X_draw_graph_fr'][:, 1]\n",
    "cell_embedding = pd.DataFrame({'Cell': cells, 'X': x, 'Y': y})\n",
    "cell_embedding.to_csv(os.path.join(out_dir, 'cell_embedding.csv'), index=False)\n",
    "\n",
    "print('Done!')\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b50ff77f",
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "## FIND ALL MARKERS on LEIDEN clustering"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "17f3daa2",
   "metadata": {
    "name": "FIND ALL MARKERS on LEIDEN clustering"
   },
   "outputs": [],
   "source": [
    "# use scanpy clustering on FA graph \n",
    "Idents(only_STERO)=only_STERO$leiden\n",
    "DimPlot(only_STERO,reduction = 'FA',pt.size = 3,label = T)+DimPlot(only_STERO,reduction = 'FA',pt.size = 3,group.by = 'identity')\n",
    "all_markers=FindAllMarkers(only_STERO,assay = 'SCT_RUN_STERO',logfc.threshold = 0.1,min.pct = 0.1)\n",
    "markers_foldchange=all_markers[all_markers$avg_log2FC>0.5,]\n",
    " # annotation cluster highlinghting sub zone \n",
    "only_STERO <- RenameIdents(only_STERO,\n",
    "                     `0`='Male_Fasciculata_1',  \n",
    "                     `1`='Female_Glomerulosa',\n",
    "                     `2`='Male_Glomerulosa',\n",
    "                     `3`='Female_Fasciculata_1',\n",
    "                     `4`='Male_Fasciculata_2',\n",
    "                     `5`='Female_Fasciculata_2',\n",
    "                     `6`='Male_tZone',\n",
    "                     `7`='Female_tZone',\n",
    "                     `8`='Capsular_cells'\n",
    "                     )\n",
    "only_STERO@meta.data$leiden_annotated=Idents(only_STERO)\n",
    "\n",
    "# Create new ident merging \n",
    "Idents(only_STERO)=only_STERO$leiden\n",
    "only_STERO <- RenameIdents(only_STERO,\n",
    "                     `0`='Male_Fasciculata',  \n",
    "                     `1`='Female_Glomerulosa',\n",
    "                     `2`='Male_Glomerulosa',\n",
    "                     `3`='Female_Fasciculata',\n",
    "                     `4`='Male_Fasciculata',\n",
    "                     `5`='Female_Fasciculata',\n",
    "                     `6`='Male_tZone',\n",
    "                     `7`='Female_tZone',\n",
    "                     `8`='Capsular_cells'\n",
    "                     )\n",
    "\n",
    "only_STERO$leiden_merging=Idents(only_STERO)\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ca84f25e",
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "## Get sexual related genes and TF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4bdb4f57",
   "metadata": {
    "lines_to_next_cell": 2,
    "name": "Get sexual related genes and TF"
   },
   "outputs": [],
   "source": [
    "\n",
    "library(biomaRt)\n",
    "library(BSgenome.Mmusculus.UCSC.mm10)\n",
    "\n",
    "DefaultAssay(only_STERO)='SCT_RUN_STERO'\n",
    "mart <- useMart(\"ENSEMBL_MART_ENSEMBL\", dataset = \"mmusculus_gene_ensembl\")\n",
    "\n",
    "genes.table <- try(biomaRt::getBM(attributes = c(\"ensembl_gene_id\", \"external_gene_name\",\n",
    "\"description\", \"gene_biotype\", \"chromosome_name\", \"start_position\"), mart = mart, useCache = F))   # fetch chromosome info plus some other annotations\n",
    "\n",
    "chrY.gene = genes.table$external_gene_name[genes.table$chromosome_name == \"Y\"] \n",
    "chrX.gene = genes.table$external_gene_name[genes.table$chromosome_name == \"X\"]\n",
    "chrX_and_Y=append(chrY.gene,chrX.gene)\n",
    "mouse_TF=read.table('D:/ADRENAL_GLAND_PROJECT/mouseTF.txt')\n",
    "chrx_y=as.data.frame(chrX_and_Y)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c1a40b5e",
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "## DE ANALYSIS , for merging  and sub clusters and sub clusters intra comparison "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4772b0f4",
   "metadata": {
    "lines_to_next_cell": 2,
    "name": "DE ANALYSIS"
   },
   "outputs": [],
   "source": [
    "\n",
    "# 0) optional ) remove genes ==> create an assay object without genes xyz\n",
    "  remove_only_stero=only_STERO\n",
    "  remove_only_stero[['peaks']] <- NULL\n",
    "  counts <- GetAssayData(only_STERO, assay = \"SCT_RUN_STERO\")\n",
    "  counts <- counts[-(which(rownames(counts) %in% c('Xist','Tsix','Uty','Ddx3y'))),]\n",
    "  remove_only_stero <- subset(remove_only_stero, features = rownames(counts))\n",
    "  \n",
    "  \n",
    "# 1) take the right ident - old / new /subset \n",
    "  # COMPARING EACH MERGING CLUSTERS \n",
    "Idents(remove_only_stero)=remove_only_stero@meta.data$leiden_merging\n",
    "DefaultAssay(remove_only_stero)='SCT_RUN_STERO' #take good assay \n",
    "levels(remove_only_stero)\n",
    "vector_clusters_F=c( \"Female_Fasciculata\",\"Female_Glomerulosa\",\"Female_tZone\")\n",
    "vector_clusters_M=c( \"Male_Fasciculata\", \"Male_Glomerulosa\",\"Male_tZone\")\n",
    "\n",
    "for( i in 1:length(vector_clusters_F)){\n",
    "  assign(paste0('DE_STERO_',vector_clusters_F[i]),FindMarkers(remove_only_stero,ident.1 =vector_clusters_F[i],ident.2 = vector_clusters_M[i],assay = 'SCT_RUN_STERO',min.pct = 0.1))\n",
    "\n",
    "}\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# FIND all \n",
    "ALL_diffmarkers=FindMarkers(remove_only_stero,ident.1 =c(\"Female_Fasciculata\",\"Female_tZone\",\"Female_Glomerulosa\"),ident.2=c(\"Male_Fasciculata\",\"Male_tZone\",\"Male_Glomerulosa\"),assay = 'SCT_RUN_STERO' ,logfc.threshold = 0.1)\n",
    "ALL_diffmarkers<- ALL_diffmarkers[ALL_diffmarkers$p_val_adj < 0.05,]\n",
    "ALL_diffmarkers$gene_names <- rownames(ALL_diffmarkers)\n",
    "ALL_diffmarkers$TF <- mouse_TF$V1[match(ALL_diffmarkers$gene_names, mouse_TF$V1)]\n",
    "ALL_diffmarkers$sex <- chrx_y$chrX_and_Y[match(ALL_diffmarkers$gene_names, chrx_y$chrX_and_Y)]\n",
    "write.csv(ALL_diffmarkers,\"D:/ADRENAL_GLAND_PROJECT/ALL_diffmarkers.csv\", row.names = FALSE)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# FEMALE ZF \n",
    "DE_STERO_Female_ZF<- DE_STERO_Female_Fasciculata[DE_STERO_Female_Fasciculata$p_val_adj < 0.05&DE_STERO_Female_Fasciculata$avg_log2FC>0.1,]\n",
    "DE_STERO_Female_ZF$gene_names <- rownames(DE_STERO_Female_ZF)\n",
    "DE_STERO_Female_ZF$TF <- mouse_TF$V1[match(DE_STERO_Female_ZF$gene_names, mouse_TF$V1)]\n",
    "DE_STERO_Female_ZF$sex <- chrx_y$chrX_and_Y[match(DE_STERO_Female_ZF$gene_names, chrx_y$chrX_and_Y)]\n",
    "write.csv(DE_STERO_Female_ZF,\"D:/ADRENAL_GLAND_PROJECT/FASCICULATA/DE_STERO_Female_ZF.csv\", row.names = FALSE)\n",
    "\n",
    "\n",
    "\n",
    "# Male ZF \n",
    "DE_STERO_Male_ZF <- DE_STERO_Female_Fasciculata[DE_STERO_Female_Fasciculata$p_val_adj < 0.05&DE_STERO_Female_Fasciculata$avg_log2FC< - 0.1,]\n",
    "DE_STERO_Male_ZF$gene_names <- rownames(DE_STERO_Male_ZF)\n",
    "DE_STERO_Male_ZF$TF <- mouse_TF$V1[match(DE_STERO_Male_ZF$gene_names, mouse_TF$V1)]\n",
    "DE_STERO_Male_ZF$sex <- chrx_y$chrX_and_Y[match(DE_STERO_Male_ZF$gene_names, chrx_y$chrX_and_Y)]\n",
    "write.csv(DE_STERO_Male_ZF,\"D:/ADRENAL_GLAND_PROJECT/FASCICULATA/DE_STERO_Male_ZF.csv\", row.names = FALSE)\n",
    "\n",
    "\n",
    "# Female ZG \n",
    "DE_STERO_Female_ZG<- DE_STERO_Female_Glomerulosa[DE_STERO_Female_Glomerulosa$p_val_adj < 0.05 & DE_STERO_Female_Glomerulosa$avg_log2FC>0.1,]\n",
    "DE_STERO_Female_ZG$gene_names <- rownames(DE_STERO_Female_ZG)\n",
    "DE_STERO_Female_ZG$TF <- mouse_TF$V1[match(DE_STERO_Female_ZG$gene_names, mouse_TF$V1)]\n",
    "DE_STERO_Female_ZG$sex <- chrx_y$chrX_and_Y[match(DE_STERO_Female_ZG$gene_names, chrx_y$chrX_and_Y)]\n",
    "write.csv(DE_STERO_Female_ZG,\"D:/ADRENAL_GLAND_PROJECT/GLOMERULOSA/DE_STERO_Female_ZG.csv\", row.names = FALSE)\n",
    "\n",
    "\n",
    "# Male ZG \n",
    "DE_STERO_Male_ZG<- DE_STERO_Female_Glomerulosa[DE_STERO_Female_Glomerulosa$p_val_adj < 0.05&DE_STERO_Female_Glomerulosa$avg_log2FC< - 0.1,]\n",
    "DE_STERO_Male_ZG$gene_names <- rownames(DE_STERO_Male_ZG)\n",
    "DE_STERO_Male_ZG$TF <- mouse_TF$V1[match(DE_STERO_Male_ZG$gene_names, mouse_TF$V1)]\n",
    "DE_STERO_Male_ZG$sex <- chrx_y$chrX_and_Y[match(DE_STERO_Male_ZG$gene_names, chrx_y$chrX_and_Y)]\n",
    "write.csv(DE_STERO_Male_ZG,\"D:/ADRENAL_GLAND_PROJECT/GLOMERULOSA/DE_STERO_Male_ZG.csv\", row.names = FALSE)\n",
    "\n",
    "\n",
    "# Female tZ \n",
    "DE_STERO_Female_tZ<- DE_STERO_Female_tZone[DE_STERO_Female_tZone$p_val_adj < 0.05 & DE_STERO_Female_tZone$avg_log2FC>0.1,]\n",
    "DE_STERO_Female_tZ$gene_names <- rownames(DE_STERO_Female_tZ)\n",
    "DE_STERO_Female_tZ$TF <- mouse_TF$V1[match(DE_STERO_Female_tZ$gene_names, mouse_TF$V1)]\n",
    "DE_STERO_Female_tZ$sex <- chrx_y$chrX_and_Y[match(DE_STERO_Female_tZ$gene_names, chrx_y$chrX_and_Y)]\n",
    "write.csv(DE_STERO_Female_tZ,\"D:/ADRENAL_GLAND_PROJECT/TZONE/DE_STERO_Female_tZ.csv\", row.names = FALSE)\n",
    "\n",
    "\n",
    "# Male tZ\n",
    "DE_STERO_Male_tZ<- DE_STERO_Female_tZone[DE_STERO_Female_tZone$p_val_adj < 0.05&DE_STERO_Female_tZone$avg_log2FC< - 0.1,]\n",
    "DE_STERO_Male_tZ$gene_names <- rownames(DE_STERO_Male_tZ)\n",
    "DE_STERO_Male_tZ$TF <- mouse_TF$V1[match(DE_STERO_Male_tZ$gene_names, mouse_TF$V1)]\n",
    "DE_STERO_Male_tZ$sex <- chrx_y$chrX_and_Y[match(DE_STERO_Male_tZ$gene_names, chrx_y$chrX_and_Y)]\n",
    "write.csv(DE_STERO_Male_tZ,\"D:/ADRENAL_GLAND_PROJECT/TZONE/DE_STERO_Male_tZ.csv\", row.names = FALSE)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "rm(DE_STERO_SUBFemale_Fasciculata_1,DE_STERO_SUBFemale_Fasciculata_2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fda8ff3d",
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "## differential peaks accesibility AND motif accessibility"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e5ae3589",
   "metadata": {
    "lines_to_next_cell": 0,
    "name": "differential peaks accesibility AND motif accessibility"
   },
   "outputs": [],
   "source": [
    "\n",
    "Idents(only_STERO)=only_STERO@meta.data$leiden_merging\n",
    "DefaultAssay(only_STERO)='peaks'\n",
    "vector_clusters_F=c( \"Female_Fasciculata\",\"Female_Glomerulosa\",\"Female_tZone\")\n",
    "vector_clusters_M=c( \"Male_Fasciculata\", \"Male_Glomerulosa\",\"Male_tZone\")\n",
    "\n",
    "\n",
    "for( i in 1:length(vector_clusters_F)){ \n",
    "  assign(paste0('DA_STERO_',vector_clusters_F[i]),FindMarkers(only_STERO,\n",
    "                                                        ident.1 =vector_clusters_F[i],\n",
    "                                                        ident.2 = vector_clusters_M[i],\n",
    "                                                        assay = 'peaks',\n",
    "                                                        only.pos = F, test.use = 'LR',\n",
    "                                                        min.pct = 0.05,\n",
    "                                                        latent.vars = 'nCount_peaks'))\n",
    "}\n",
    "\n",
    "\n",
    "\n",
    "# FASCICULATA \n",
    "DA_STERO_Female_Fasciculata$peaks=rownames(DA_STERO_Female_Fasciculata)\n",
    "DA_STERO_Female_Fasciculata <- DA_STERO_Female_Fasciculata[DA_STERO_Female_Fasciculata$p_val_adj < 0.05,]\n",
    "\n",
    "peaks_Fem_FASC=DA_STERO_Female_Fasciculata$peaks[DA_STERO_Female_Fasciculata$p_val_adj < 0.05&DA_STERO_Female_Fasciculata$avg_log2FC>0.1]\n",
    "peaks_Male_FASC=DA_STERO_Female_Fasciculata$peaks[DA_STERO_Female_Fasciculata$p_val_adj < 0.05&DA_STERO_Female_Fasciculata$avg_log2FC< -0.1]\n",
    "\n",
    "\n",
    "# GLOMERULOSA \n",
    "DA_STERO_Female_Glomerulosa$peaks=rownames(DA_STERO_Female_Glomerulosa)\n",
    "DA_STERO_Female_Glomerulosa <- DA_STERO_Female_Glomerulosa[DA_STERO_Female_Glomerulosa$p_val_adj < 0.05,]\n",
    "\n",
    "peaks_Fem_GLO=DA_STERO_Female_Glomerulosa$peaks[DA_STERO_Female_Glomerulosa$p_val_adj < 0.05&DA_STERO_Female_Glomerulosa$avg_log2FC>0.1]\n",
    "peaks_Male_GLO=DA_STERO_Female_Glomerulosa$peaks[DA_STERO_Female_Glomerulosa$p_val_adj < 0.05&DA_STERO_Female_Glomerulosa$avg_log2FC< -0.1]\n",
    "\n",
    "\n",
    "#  Tzone\n",
    "DA_STERO_Female_tZone$peaks=rownames(DA_STERO_Female_tZone)\n",
    "DA_STERO_Female_tZone <- DA_STERO_Female_tZone[DA_STERO_Female_tZone$p_val_adj < 0.05,]\n",
    "\n",
    "peaks_Fem_tzone=DA_STERO_Female_tZone$peaks[DA_STERO_Female_tZone$p_val_adj < 0.05&DA_STERO_Female_tZone$avg_log2FC>0.1]\n",
    "peaks_Male_tzone=DA_STERO_Female_tZone$peaks[DA_STERO_Female_tZone$p_val_adj < 0.05&DA_STERO_Female_tZone$avg_log2FC< -0.1]\n",
    "\n",
    "\n",
    "# Background and motif enrichement :\n",
    "              \n",
    "              # Fasciculata : \n",
    "                      open.peaks <- AccessiblePeaks(only_STERO, idents = c(\"Female_Fasciculata\",\"Male_Fasciculata\"),assay = 'peaks') \n",
    "                      \n",
    "                      meta.feature <- GetAssayData(only_STERO, assay = \"peaks\", slot = \"meta.features\")\n",
    "                      # Female \n",
    "                            peaks.matched <- MatchRegionStats(\n",
    "                            meta.feature = meta.feature[open.peaks, ],\n",
    "                            query.feature = meta.feature[peaks_Fem_FASC, ],\n",
    "                            n = 50000\n",
    "                            )\n",
    "                              \n",
    "                            enrich_motif_Fem_FASC <- FindMotifs(object = only_STERO,features = peaks_Fem_FASC,background = peaks.matched)\n",
    "                            enrich_motif_Fem_FASC <- enrich_motif_Fem_FASC[enrich_motif_Fem_FASC$pvalue<0.05,]\n",
    "                                          \n",
    "                      # Male \n",
    "                            peaks.matched <- MatchRegionStats(\n",
    "                            meta.feature = meta.feature[open.peaks, ],\n",
    "                            query.feature = meta.feature[peaks_Male_FASC, ],\n",
    "                            n = 50000\n",
    "                            )\n",
    "                            enrich_motif_Male_FASC <- FindMotifs(object = only_STERO,features = peaks_Male_FASC,background = peaks.matched)\n",
    "                            enrich_motif_Male_FASC <- enrich_motif_Male_FASC[enrich_motif_Male_FASC$pvalue<0.05,]\n",
    "\n",
    "              # Glomerulosa : \n",
    "                      open.peaks <- AccessiblePeaks(only_STERO, idents = c(\"Male_Glomerulosa\",\"Female_Glomerulosa\"),assay = 'peaks') \n",
    "                      \n",
    "                      meta.feature <- GetAssayData(only_STERO, assay = \"peaks\", slot = \"meta.features\")\n",
    "                      # Female \n",
    "                            peaks.matched <- MatchRegionStats(\n",
    "                            meta.feature = meta.feature[open.peaks, ],\n",
    "                            query.feature = meta.feature[peaks_Fem_GLO, ],\n",
    "                            n = 50000\n",
    "                            )\n",
    "                              \n",
    "                            enrich_motif_Fem_GLO<- FindMotifs(object = only_STERO,features = peaks_Fem_GLO,background = peaks.matched)\n",
    "                            enrich_motif_Fem_GLO <- enrich_motif_Fem_GLO[enrich_motif_Fem_GLO$pvalue<0.05,]\n",
    "\n",
    "                                          \n",
    "                      # Male \n",
    "                            peaks.matched <- MatchRegionStats(\n",
    "                            meta.feature = meta.feature[open.peaks, ],\n",
    "                            query.feature = meta.feature[peaks_Male_GLO, ],\n",
    "                            n = 50000\n",
    "                            )\n",
    "                            enrich_motif_Male_GLO <- FindMotifs(object = only_STERO,features = peaks_Male_GLO,background = peaks.matched)\n",
    "                            enrich_motif_Male_GLO <- enrich_motif_Male_GLO[enrich_motif_Male_GLO$pvalue<0.05,]\n",
    "\n",
    "                            \n",
    "                # TZONE  : \n",
    "                      open.peaks <- AccessiblePeaks(only_STERO, idents = c(\"Female_tZone\",\"Male_tZone\"),assay = 'peaks') \n",
    "                      \n",
    "                      meta.feature <- GetAssayData(only_STERO, assay = \"peaks\", slot = \"meta.features\")\n",
    "                      # Female \n",
    "                            peaks.matched <- MatchRegionStats(\n",
    "                            meta.feature = meta.feature[open.peaks, ],\n",
    "                            query.feature = meta.feature[peaks_Fem_tzone, ],\n",
    "                            n = 50000\n",
    "                            )\n",
    "                              \n",
    "                            enrich_motif_Fem_TZONE<- FindMotifs(object = only_STERO,features = peaks_Fem_tzone,background = peaks.matched)\n",
    "                            enrich_motif_Fem_TZONE <- enrich_motif_Fem_TZONE[enrich_motif_Fem_TZONE$pvalue<0.05,]\n",
    "                                          \n",
    "                      # Male \n",
    "                            peaks.matched <- MatchRegionStats(\n",
    "                            meta.feature = meta.feature[open.peaks, ],\n",
    "                            query.feature = meta.feature[peaks_Male_tzone, ],\n",
    "                            n = 50000\n",
    "                            )\n",
    "                            enrich_motif_Male_TZONE <- FindMotifs(object = only_STERO,features = peaks_Male_tzone,background = peaks.matched)\n",
    "                            enrich_motif_Male_TZONE <- enrich_motif_Male_TZONE[enrich_motif_Male_TZONE$pvalue<0.05,]\n",
    "                            \n",
    "                            \n",
    "\n",
    "                          \n",
    "                            \n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d4240b14",
   "metadata": {},
   "source": [
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2a56161e",
   "metadata": {
    "lines_to_next_cell": 0,
    "name": "PLOT umap"
   },
   "outputs": [],
   "source": [
    "library(ggplot2)\n",
    "# PLOT UMAP\n",
    "DimPlot(only_STERO,reduction = 'FA',pt.size = 1.66) + guides(color = guide_legend(override.aes = list(size=5), ncol=1) ) + theme_classic(base_size = 15)|\n",
    "  DimPlot(multiomics,reduction = 'wnn.umap',pt.size = 1.66,group.by='Replicat') + guides(color = guide_legend(override.aes = list(size=5), ncol=1) ) + theme_classic(base_size = 15)| DimPlot(multiomics,reduction = 'wnn.umap',pt.size = 1.66) + guides(color = guide_legend(override.aes = list(size=5), ncol=1) ) + theme_classic(base_size = 15)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a151fa7c",
   "metadata": {},
   "source": [
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "07974a8d",
   "metadata": {
    "lines_to_next_cell": 2
   },
   "outputs": [],
   "source": [
    "FeaturePlot(multiomics, features=c('Wnt4','Dab2'),reduction = 'pca_umap')                                   # glomerulosa markers gene \n",
    "FeaturePlot(multiomics, features=c('Akr1b7','Cyp11b1'),reduction = 'pca_umap')                              # Fasciculata markers gene \n",
    "FeaturePlot(multiomics, features=c('Th','Slc18a1','Slc18a2','Ddc','Dbh','Lgr5'),reduction = 'pca_umap')     # medulla markers gene \n",
    "FeaturePlot(multiomics, features=c(\"Rspo3\",'Gli1'),reduction = 'pca_umap')                                  # capsular ( progennitor compartment) markers gene \n",
    "FeaturePlot(multiomics, features=c('Akr1c18'),reduction = 'pca_umap')                                       # x zone\n",
    "FeaturePlot(multiomics, features=c('Cdh5','Pecam1','Flt1'),reduction = 'pca_umap')                          # endothelial  markers gene \n",
    "FeaturePlot(multiomics, features=c('Star','Cyp11a1', 'Hsd3b1', 'Scarb1'),reduction = 'wnn.umap')            # cortex markers gene \n",
    "FeaturePlot(multiomics, features=c('Csmd1'),reduction = 'pca_umap')                                         # capsular marker genes \n",
    "FeaturePlot(multiomics, features=c(\"Csmd1\",\"Ano1\",\"Col3a1\",\"Cdh11\",\"Dcn\",\"Rspo3\"),reduction = 'pca_umap')   # stromal cells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "86725de1",
   "metadata": {
    "lines_to_next_cell": 2
   },
   "outputs": [],
   "source": [
    "# markers genes \n",
    "genes=c('Wnt4','Dab2','Akr1b7','Cyp11b1',\"Csmd1\",\"Rspo3\",\"Tcf21\", 'Flt1','Pecam1',\"Lsamp\",\"Lrrtm4\",'Th','Slc18a1',\"Cd74\",\"Cd84\",\"Akr1c18\",\"Pik3c2g\")\n",
    "\n",
    "dotx= DotPlot(multiomics,assay= 'SCT',features = genes,dot.scale = 6)+ RotatedAxis() +  scale_colour_gradient2(low = \"darkred\", mid = \"white\",high = \"darkgreen\",midpoint = 0,)\n",
    "\n",
    " # Markers genes plot \n",
    " # Main plot\n",
    "  minmaxthis = function(x) { \n",
    "    (x - min(x))/(max(x) - min(x))\n",
    "  }\n",
    "  \n",
    "  data.dot = dotx$data\n",
    "  \n",
    "  # scaled value\n",
    "  scaled_data <- \n",
    "    data.dot %>%\n",
    "    group_by(features.plot) %>%\n",
    "    mutate(minmaxscale = minmaxthis(avg.exp))\n",
    "  \n",
    "  scaled_data$avg.exp2 = ( scaled_data$avg.exp - min(scaled_data$avg.exp) )    / \n",
    "    (max(scaled_data$avg.exp)- min(scaled_data$avg.exp))\n",
    "  \n",
    "  \n",
    "  pmain <- ggplot(scaled_data)  + \n",
    "    geom_point(aes(x = features.plot, y = id, color = minmaxscale, size = pct.exp), stat = \"identity\") + \n",
    "    scale_size(range = c(0, 10), breaks  = seq(0,100,20), labels  = paste0(seq(0,100,20),\"%\"))  + \n",
    "    scale_color_gradient(low = \"white\",high = \"#0f4d91\",breaks = seq(0,1,0.2))  +\n",
    "    guides(color = guide_colourbar(title = \"Expression\", barwidth = 1, barheight = 10),\n",
    "           size = guide_legend(title = \"Cells (%)\",reverse=TRUE)) + \n",
    "    xlab(\"\") + ylab(\"\") + \n",
    "    \n",
    "    theme(panel.border=element_blank(), \n",
    "          axis.line = element_line(colour = 'black', size = 1), \n",
    "          axis.ticks = element_blank(),\n",
    "          axis.text = element_text(size=12, colour = \"black\"), \n",
    "          axis.text.x = element_text(angle = 90, hjust = 1),\n",
    "          panel.grid.major = element_blank(),\n",
    "          panel.grid.minor = element_blank(),\n",
    "          panel.background = element_blank(),\n",
    "          legend.key=element_blank(),\n",
    "          # legend.title=element_blank(),\n",
    "          legend.text=element_text(size=12))\n",
    "  \n",
    "pmain\n",
    "  \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "37fefa88",
   "metadata": {
    "name": "Variable genes remove non interesing genes"
   },
   "outputs": [],
   "source": [
    "  DefaultAssay(multiomics)='SCT'\n",
    "  remove_multiomics=multiomics\n",
    "  remove_multiomics[['peaks']] <- NULL\n",
    "  counts <- GetAssayData(remove_multiomics, assay = \"SCT\")\n",
    "  counts <- counts[-(which(rownames(counts) %in% c('Xist','Tsix','Uty','Ddx3y'))),]\n",
    "  remove_multiomics <- subset(remove_multiomics, features = rownames(counts))\n",
    "  \n",
    "# male vs Female | fasciculata replicate\n",
    "multiomics_replicat=FindMarkers(remove_multiomics,assay = 'SCT',logfc.threshold = 0.1,ident.1 =c(\"Male Fasciculata Replicate 2\",\"Male Fasciculata Replicate 1\"),ident.2 = c( \"Female Fasciculata Replicate 1\", \"Female Fasiculata Replicate 2\"))\n",
    "multiomics_replicat=multiomics_replicat[multiomics_replicat$p_val_adj<0.05,]\n",
    "\n",
    "# male vs male | fasciculata replicate\n",
    "diff_MALE_fasc=FindMarkers(remove_multiomics,assay = 'SCT',logfc.threshold = 0.1,ident.1 =c(\"Male Fasciculata Replicate 2\"),ident.2 = c(\"Male Fasciculata Replicate 1\"))\n",
    "diff_MALE_fasc=diff_MALE_fasc[diff_MALE_fasc$p_val_adj<0.05,]\n",
    "\n",
    "# Female vs male | fasciculata replicate\n",
    "diff_FEMALE_fasc=FindMarkers(remove_multiomics,assay = 'SCT',logfc.threshold = 0.1,ident.1 =c(\"Female Fasiculata Replicate 2\"),ident.2 = c(\"Female Fasciculata Replicate 1\"))\n",
    "diff_FEMALE_fasc=diff_FEMALE_fasc[diff_FEMALE_fasc$p_val_adj<0.05,]\n",
    "\n",
    "write.csv(diff_FEMALE_fasc,file = 'D:/ADRENAL_PROJECT/FASCICULATA/Fasciculata_replicate/diff_FEMALE_fasc.csv')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "97187a4a",
   "metadata": {
    "lines_to_next_cell": 0,
    "name": "not run"
   },
   "outputs": [],
   "source": [
    "# markers genes \n",
    " # 0) all_markers\n",
    "      all_markers_replicat=FindAllMarkers(multiomics,assay = 'SCT_RUN_STERO',logfc.threshold = 0.1)\n",
    "    # Write as csv \n",
    "      write.csv(all_markers$gene[all_markers$p_val_adj<0.05&all_markers$avg_log2FC>1&all_markers$cluster==\"Immune Cells\"],file = 'all_markers.txt',quote = F,row.names = F,col.names = F)\n",
    "      \n",
    "      \n",
    "      \n",
    "      \n",
    "      \n",
    "      \n",
    "      \n",
    "      multiomics <- RenameIdents(multiomics,`Endothelial Cells`='Endothelial Cells',\n",
    "                     `Male specific Medulla`='Male Medulla',\n",
    "                     `Male Fasciculata Replicate 1`='Male Fasciculata Replicate 1',\n",
    "                     `Male Fasciculata Replicate 2`='Male Fasciculata Replicate 2',\n",
    "                     `Female Fasiculata Replicate 2`='Female Fasciculata Replicate 2',\n",
    "                     `Female Glomerulosa`='Female Glomerulosa',\n",
    "                     `Female Fasciculate Replicate 1`='Female Fasciculata Replicate 1',\n",
    "                     `Male Glomerulosa`='Male Glomerulosa',\n",
    "                     `Female specific Medulla`='Female Medulla',\n",
    "                     `Medulla progenitor Cells`='Medulla progenitor Cells',\n",
    "                     `Capsular Cells`='Capsular Cells',`Immune Cells`='Immune Cells',\n",
    "                     `Fetal adrenocortical Cells `= 'Stromal cells',\n",
    "                     `X zone `='X zone')\n",
    "      \n",
    "      multiomics@meta.data$cell_types=Idents(multiomics)\n",
    "      \n",
    "      \n",
    "      \n",
    "    # 0.bis) conserved markers\n",
    "\n",
    "      FindConservedMarkers(pbmc_small, ident.1 = 0, ident.2 = 1, grouping.var = \"groups\")\n",
    "\n",
    "      all_conserved=FindMarkers(object = multiomics,ident.1 = c(\"Male Glomerulosa\",\n",
    "                                      \"Male Fasciculata Replicate 1\",\n",
    "                                      \"Male Fasciculata Replicate 2\",\n",
    "                                      \"Female Fasciculata Replicate 1\",\n",
    "                                      \"Female Fasiculata Replicate 2\"))\n",
    "      \n",
    "      all_conserved$gene=rownames(all_conserved)\n",
    " # 1) reorder clusters :\n",
    "        # reorder clusters\n",
    "      # full dataset\n",
    "        multiomics@active.ident <- factor(multiomics@active.ident,   \n",
    "                                    levels=c(\"Female Glomerulosa\",\n",
    "                                              \"Male Glomerulosa\",\n",
    "                                              \"Male Fasciculata Replicate 1\",\n",
    "                                              \"Male Fasciculata Replicate 2\",\n",
    "                                              \"Female Fasciculata Replicate 1\",\n",
    "                                              \"Female Fasiculata Replicate 2\",\n",
    "                                              \"Capsular Cells\",\n",
    "                                              \"Stromal cells\",\n",
    "                                              \"Endothelial Cells\",\n",
    "                                              \"Medulla progenitor Cells\",\n",
    "                                              \"Male Medulla\",\n",
    "                                              \"Female Medulla\",\n",
    "                                              \"Immune Cells\",\n",
    "                                               \"X zone\"))\n",
    "\n",
    "\n",
    "    # only_STERO\n",
    "          only_STERO@active.ident <- factor(only_STERO@active.ident, \n",
    "                            levels=c(\"Male_Glomerulosa\",\n",
    "                                      \"Female_Glomerulosa\",\n",
    "                                      \"Female_tZone\",\n",
    "                                      \"Male_tZone\",\n",
    "                                      \"Male_Fasciculata\",\n",
    "                                      \"Female_Fasciculata\",\n",
    "                                      \"Capsular_cells\"))\n",
    "\n",
    "\n",
    "# FIND MARKERS REPLICATE DIFFERENCES : \n",
    "  \n",
    "Replicat=FindMarkers(multiomics,ident.1 =\"Female Fasciculata Replicate 1\",ident.2 = \"Female Fasiculata Replicate 2\",assay = 'SCT' )\n",
    "Replicat$genes=rownames(Replicat)\n",
    "write.table(Replicat$genes,file = 'ReplicatFEMALE.txt',quote = F,col.names = F,row.names = F)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a95493db",
   "metadata": {},
   "source": [
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c8a0a00c",
   "metadata": {
    "lines_to_next_cell": 2,
    "name": "FIND ALL MARKERS on LEIDEN clustering"
   },
   "outputs": [],
   "source": [
    "# use scanpy clustering on FA graph \n",
    "Idents(only_STERO)=only_STERO$leiden\n",
    "DimPlot(only_STERO,reduction = 'FA',pt.size = 3,label = T)+DimPlot(only_STERO,reduction = 'FA',pt.size = 3,group.by = 'identity')\n",
    "all_markers=FindAllMarkers(only_STERO,assay = 'SCT_RUN_STERO',logfc.threshold = 0.1,min.pct = 0.1)\n",
    "markers_foldchange=all_markers[all_markers$avg_log2FC>0.5,]\n",
    " # annotation cluster highlinghting sub zone \n",
    "only_STERO <- RenameIdents(only_STERO,\n",
    "                     `0`='Male_Fasciculata_1',  \n",
    "                     `1`='Female_Glomerulosa',\n",
    "                     `2`='Male_Glomerulosa',\n",
    "                     `3`='Female_Fasciculata_1',\n",
    "                     `4`='Male_Fasciculata_2',\n",
    "                     `5`='Female_Fasciculata_2',\n",
    "                     `6`='Male_tZone',\n",
    "                     `7`='Female_tZone',\n",
    "                     `8`='Capsular_cells'\n",
    "                     )\n",
    "only_STERO@meta.data$leiden_annotated=Idents(only_STERO)\n",
    "\n",
    "# Create new ident merging \n",
    "Idents(only_STERO)=only_STERO$leiden\n",
    "only_STERO <- RenameIdents(only_STERO,\n",
    "                     `0`='Male_Fasciculata',  \n",
    "                     `1`='Female_Glomerulosa',\n",
    "                     `2`='Male_Glomerulosa',\n",
    "                     `3`='Female_Fasciculata',\n",
    "                     `4`='Male_Fasciculata',\n",
    "                     `5`='Female_Fasciculata',\n",
    "                     `6`='Male_tZone',\n",
    "                     `7`='Female_tZone',\n",
    "                     `8`='Capsular_cells'\n",
    "                     )\n",
    "\n",
    "only_STERO$leiden_merging=Idents(only_STERO)\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "51adfabe",
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "#DE ANALYSIS , for merging  and sub clusters and sub clusters intra comparison "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1f3e9a3e",
   "metadata": {
    "lines_to_next_cell": 0,
    "name": "DE ANALYSIS"
   },
   "outputs": [],
   "source": [
    "\n",
    "# install packages\n",
    "\n",
    "library(biomaRt)\n",
    "library(BSgenome.Mmusculus.UCSC.mm10)\n",
    "\n",
    "DefaultAssay(only_STERO)='SCT_RUN_STERO'\n",
    "mart <- useMart(\"ENSEMBL_MART_ENSEMBL\", dataset = \"mmusculus_gene_ensembl\")\n",
    "\n",
    "genes.table <- try(biomaRt::getBM(attributes = c(\"ensembl_gene_id\", \"external_gene_name\",\n",
    "\"description\", \"gene_biotype\", \"chromosome_name\", \"start_position\"), mart = mart, useCache = F))   # fetch chromosome info plus some other annotations\n",
    "\n",
    "chrY.gene = genes.table$external_gene_name[genes.table$chromosome_name == \"Y\"] \n",
    "chrX.gene = genes.table$external_gene_name[genes.table$chromosome_name == \"X\"]\n",
    "chrX_and_Y=append(chrY.gene,chrX.gene)\n",
    "mouse_TF=read.table('D:/ADRENAL_PROJECT/mouseTF.txt')\n",
    "chrx_y=as.data.frame(chrX_and_Y)\n",
    "\n",
    "\n",
    "\n",
    "# 0) optional ) remove genes ==> create an assay object without genes xyz\n",
    "  remove_only_stero=only_STERO\n",
    "  remove_only_stero[['peaks']] <- NULL\n",
    "  counts <- GetAssayData(only_STERO, assay = \"SCT_RUN_STERO\")\n",
    "  counts <- counts[-(which(rownames(counts) %in% c('Xist','Tsix','Uty','Ddx3y'))),]\n",
    "  remove_only_stero <- subset(remove_only_stero, features = rownames(counts))\n",
    "  \n",
    "  \n",
    "# 1) take the right ident - old / new /subset \n",
    "  # COMPARING EACH MERGING CLUSTERS \n",
    "Idents(remove_only_stero)=remove_only_stero@meta.data$leiden_merging\n",
    "DefaultAssay(remove_only_stero)='SCT_RUN_STERO' #take good assay \n",
    "levels(remove_only_stero)\n",
    "vector_clusters_F=c( \"Female_Fasciculata\",\"Female_Glomerulosa\",\"Female_tZone\")\n",
    "vector_clusters_M=c( \"Male_Fasciculata\", \"Male_Glomerulosa\",\"Male_tZone\")\n",
    "\n",
    "for( i in 1:length(vector_clusters_F)){\n",
    "  assign(paste0('DE_STERO_',vector_clusters_F[i]),FindMarkers(remove_only_stero,ident.1 =vector_clusters_F[i],ident.2 = vector_clusters_M[i],assay = 'SCT_RUN_STERO',min.pct = 0.1))\n",
    "\n",
    "}\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# FIND all \n",
    "ALL_diffmarkers=FindMarkers(remove_only_stero,ident.1 =c(\"Female_Fasciculata\",\"Female_tZone\",\"Female_Glomerulosa\"),ident.2=c(\"Male_Fasciculata\",\"Male_tZone\",\"Male_Glomerulosa\"),assay = 'SCT_RUN_STERO' ,logfc.threshold = 0.1)\n",
    "ALL_diffmarkers<- ALL_diffmarkers[ALL_diffmarkers$p_val_adj < 0.05,]\n",
    "ALL_diffmarkers$gene_names <- rownames(ALL_diffmarkers)\n",
    "ALL_diffmarkers$TF <- mouse_TF$V1[match(ALL_diffmarkers$gene_names, mouse_TF$V1)]\n",
    "ALL_diffmarkers$sex <- chrx_y$chrX_and_Y[match(ALL_diffmarkers$gene_names, chrx_y$chrX_and_Y)]\n",
    "write.csv(ALL_diffmarkers,\"D:/ADRENAL_PROJECT/ALL_diffmarkers.csv\", row.names = FALSE)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# FEMALE ZF \n",
    "DE_STERO_Female_ZF<- DE_STERO_Female_Fasciculata[DE_STERO_Female_Fasciculata$p_val_adj < 0.05&DE_STERO_Female_Fasciculata$avg_log2FC>0.1,]\n",
    "DE_STERO_Female_ZF$gene_names <- rownames(DE_STERO_Female_ZF)\n",
    "DE_STERO_Female_ZF$TF <- mouse_TF$V1[match(DE_STERO_Female_ZF$gene_names, mouse_TF$V1)]\n",
    "DE_STERO_Female_ZF$sex <- chrx_y$chrX_and_Y[match(DE_STERO_Female_ZF$gene_names, chrx_y$chrX_and_Y)]\n",
    "write.csv(DE_STERO_Female_ZF,\"D:/ADRENAL_PROJECT/FASCICULATA/DE_STERO_Female_ZF.csv\", row.names = FALSE)\n",
    "\n",
    "\n",
    "\n",
    "# Male ZF \n",
    "DE_STERO_Male_ZF <- DE_STERO_Female_Fasciculata[DE_STERO_Female_Fasciculata$p_val_adj < 0.05&DE_STERO_Female_Fasciculata$avg_log2FC< - 0.1,]\n",
    "DE_STERO_Male_ZF$gene_names <- rownames(DE_STERO_Male_ZF)\n",
    "DE_STERO_Male_ZF$TF <- mouse_TF$V1[match(DE_STERO_Male_ZF$gene_names, mouse_TF$V1)]\n",
    "DE_STERO_Male_ZF$sex <- chrx_y$chrX_and_Y[match(DE_STERO_Male_ZF$gene_names, chrx_y$chrX_and_Y)]\n",
    "write.csv(DE_STERO_Male_ZF,\"D:/ADRENAL_PROJECT/FASCICULATA/DE_STERO_Male_ZF.csv\", row.names = FALSE)\n",
    "\n",
    "\n",
    "# Female ZG \n",
    "DE_STERO_Female_ZG<- DE_STERO_Female_Glomerulosa[DE_STERO_Female_Glomerulosa$p_val_adj < 0.05 & DE_STERO_Female_Glomerulosa$avg_log2FC>0.1,]\n",
    "DE_STERO_Female_ZG$gene_names <- rownames(DE_STERO_Female_ZG)\n",
    "DE_STERO_Female_ZG$TF <- mouse_TF$V1[match(DE_STERO_Female_ZG$gene_names, mouse_TF$V1)]\n",
    "DE_STERO_Female_ZG$sex <- chrx_y$chrX_and_Y[match(DE_STERO_Female_ZG$gene_names, chrx_y$chrX_and_Y)]\n",
    "write.csv(DE_STERO_Female_ZG,\"D:/ADRENAL_PROJECT/GLOMERULOSA/DE_STERO_Female_ZG.csv\", row.names = FALSE)\n",
    "\n",
    "\n",
    "\n",
    "# Male ZG \n",
    "DE_STERO_Male_ZG<- DE_STERO_Female_Glomerulosa[DE_STERO_Female_Glomerulosa$p_val_adj < 0.05&DE_STERO_Female_Glomerulosa$avg_log2FC< - 0.1,]\n",
    "DE_STERO_Male_ZG$gene_names <- rownames(DE_STERO_Male_ZG)\n",
    "DE_STERO_Male_ZG$TF <- mouse_TF$V1[match(DE_STERO_Male_ZG$gene_names, mouse_TF$V1)]\n",
    "DE_STERO_Male_ZG$sex <- chrx_y$chrX_and_Y[match(DE_STERO_Male_ZG$gene_names, chrx_y$chrX_and_Y)]\n",
    "write.csv(DE_STERO_Male_ZG,\"D:/ADRENAL_PROJECT/GLOMERULOSA/DE_STERO_Male_ZG.csv\", row.names = FALSE)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# Female tZ \n",
    "DE_STERO_Female_tZ<- DE_STERO_Female_tZone[DE_STERO_Female_tZone$p_val_adj < 0.05 & DE_STERO_Female_tZone$avg_log2FC>0.1,]\n",
    "DE_STERO_Female_tZ$gene_names <- rownames(DE_STERO_Female_tZ)\n",
    "DE_STERO_Female_tZ$TF <- mouse_TF$V1[match(DE_STERO_Female_tZ$gene_names, mouse_TF$V1)]\n",
    "DE_STERO_Female_tZ$sex <- chrx_y$chrX_and_Y[match(DE_STERO_Female_tZ$gene_names, chrx_y$chrX_and_Y)]\n",
    "write.csv(DE_STERO_Female_tZ,\"D:/ADRENAL_PROJECT/TZONE/DE_STERO_Female_tZ.csv\", row.names = FALSE)\n",
    "\n",
    "\n",
    "# Male tZ\n",
    "DE_STERO_Male_tZ<- DE_STERO_Female_tZone[DE_STERO_Female_tZone$p_val_adj < 0.05&DE_STERO_Female_tZone$avg_log2FC< - 0.1,]\n",
    "DE_STERO_Male_tZ$gene_names <- rownames(DE_STERO_Male_tZ)\n",
    "DE_STERO_Male_tZ$TF <- mouse_TF$V1[match(DE_STERO_Male_tZ$gene_names, mouse_TF$V1)]\n",
    "DE_STERO_Male_tZ$sex <- chrx_y$chrX_and_Y[match(DE_STERO_Male_tZ$gene_names, chrx_y$chrX_and_Y)]\n",
    "write.csv(DE_STERO_Male_tZ,\"D:/ADRENAL_PROJECT/TZONE/DE_STERO_Male_tZ.csv\", row.names = FALSE)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e8070ab0",
   "metadata": {},
   "source": [
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9ecc300f",
   "metadata": {
    "name": "DE ANALYSIS LINE PLOT"
   },
   "outputs": [],
   "source": [
    "\n",
    "library(reshape)\n",
    "\n",
    "test_data <-data.frame (Female_DEG = c('42','32','370'),\n",
    "                        Male_DEG = c('28','92','268'),\n",
    "                        Male_DAR = c('25','88','1326'),\n",
    "                        Female_DAR = c('2','16','3402'),\n",
    "                        \n",
    "                        pseudotime =  c('Glomerulosa','Tzone','Fasciculata'))\n",
    "            \n",
    "test_data$pseudotime =factor(test_data$pseudotime,levels=c('Glomerulosa','Tzone','Fasciculata'))\n",
    "test_data_long <- melt(test_data, id=\"pseudotime\") \n",
    "\n",
    "\n",
    "test_data_long$value <- factor(test_data_long$value, levels=c('2','16','25','28','32','42','88','92','268','370','1326','3402'))\n",
    "\n",
    "test_data_long %>% \n",
    "  ggplot(aes(x =pseudotime,y = value, color = variable, group = variable)) + \n",
    "  geom_line() + \n",
    "  geom_point() +\n",
    "  ylab(\"EI1 (Expected Influence with Neighbor)\") +\n",
    "  xlab(\"Variables\")\n",
    "\n",
    "\n",
    "df1 <- data.frame(x,y,day)\n",
    "df2 <- reshape::melt(test_data, id=\"pseudotime\")\n",
    "  ggplot(data = df2, aes(x = pseudotime, y = value, fill = variable)) + geom_bar(stat = \"identity\")+ facet_wrap(~ variable)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "test_DEG <-data.frame (Female_DEG = c('0','42','32','370'),\n",
    "                        Male_DEG = c('0','28','92','268'),\n",
    "                        pseudotime =  c('capsule','Glomerulosa','Tzone','Fasciculata'))\n",
    "\n",
    "test_DEG$pseudotime =factor(test_DEG$pseudotime,levels=c('capsule','Glomerulosa','Tzone','Fasciculata'))\n",
    "test_DEG <- melt(test_DEG, id=\"pseudotime\") \n",
    "\n",
    "\n",
    "test_DEG$value <- factor(test_DEG$value, levels=c('0','28','32','42','92','268','370'))\n",
    "\n",
    "PLOT_deg=test_DEG %>% \n",
    "  ggplot(aes(x =pseudotime,y = value, color = variable, group = variable)) + \n",
    "  geom_line() + \n",
    "  geom_point() +\n",
    "  ylab(\"Variable genes\") +\n",
    "  xlab(\"Cell types\")+\n",
    "  geom_line(size=2)+\n",
    "  scale_color_manual(values=c('#D16103','#4E84C4'))\n",
    "    \n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "test_DAR <-data.frame (\n",
    "                        Male_DAR = c('0','25','88','1326'),\n",
    "                        Female_DAR = c('0','2','16','3402'),\n",
    "                        \n",
    "                        pseudotime =  c('capsule','Glomerulosa','Tzone','Fasciculata'))\n",
    "            \n",
    "test_DAR$pseudotime =factor(test_DAR$pseudotime,levels=c('capsule','Glomerulosa','Tzone','Fasciculata'))\n",
    "test_DAR <- melt(test_DAR, id=\"pseudotime\") \n",
    "\n",
    "\n",
    "test_DAR$value <- factor(test_DAR$value, levels=c('0','2','16','25','88','1326','3402'))\n",
    "\n",
    "Plot_DAR=test_DAR %>% \n",
    "  ggplot(aes(x =pseudotime,y = value, color = variable, group = variable)) + \n",
    "  geom_line() + \n",
    "  geom_point() +\n",
    "  ylab(\"Variable region\") +\n",
    "  xlab(\"Cell types\")+\n",
    "  geom_line(size=2)+\n",
    "  scale_color_manual(values=c('#4E84C4','#D16103'))\n",
    "\n",
    "\n",
    "Plot_DAR \n",
    "PLOT_deg\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "164644f1",
   "metadata": {
    "lines_to_next_cell": 2,
    "name": "Leak peak to genes"
   },
   "outputs": [],
   "source": [
    "library(BSgenome.Mmusculus.UCSC.mm10)\n",
    "library(Signac)\n",
    "# # link peaks to genes Glomerulosa\n",
    "   #  compute gene to link with peaks :\n",
    "\n",
    "            DE_STERO_Female_tZ=DE_STERO_Female_tZ$gene_names\n",
    "            DE_STERO_Female_ZF=DE_STERO_Female_ZF$gene_names\n",
    "            DE_STERO_Female_ZG=DE_STERO_Female_ZG$gene_names\n",
    "            DE_STERO_Male_tZ=DE_STERO_Male_tZ$gene_names\n",
    "            DE_STERO_Male_ZF=DE_STERO_Male_ZF$gene_names\n",
    "            DE_STERO_Male_ZG=DE_STERO_Male_ZG$gene_names\n",
    "\n",
    "\n",
    "# LINK PEAK with DEG\n",
    "DefaultAssay(only_STERO)='peaks'\n",
    "Idents(only_STERO)=only_STERO@meta.data$leiden_merging\n",
    "\n",
    "            # 1) female TZ \n",
    "LINK_DE_STERO_Female_tZ<- LinkPeaks(\n",
    "      object = only_STERO,\n",
    "      peak.assay = \"peaks\",\n",
    "      expression.assay = \"SCT_RUN_STERO\",\n",
    "      genes.use = DE_STERO_Female_tZ,\n",
    "      )\n",
    "      # take peaks with significant correlation\n",
    "      LINK_DE_STERO_Female_tZ=data.frame(Links(LINK_DE_STERO_Female_tZ))\n",
    "      saveRDS(LINK_DE_STERO_Female_tZ,file = 'LINK_DE_STERO_Female_tZ')\n",
    "      LINK_Female_tZ_gene_peaks=LINK_DE_STERO_Female_tZ$peak[LINK_DE_STERO_Female_tZ$score>0.1 &LINK_DE_STERO_Female_tZ$pvalue<0.05 ]\n",
    "# Background\n",
    "open.peaks <- AccessiblePeaks(only_STERO, idents = c(\"Female_tZone\"),assay = 'peaks') # find a set of peaks that are open in a subset of cells\n",
    "meta.feature <- GetAssayData(only_STERO, assay = \"peaks\", slot = \"meta.features\")# match the overall GC content in the peak set\n",
    "peaks.matched <- MatchRegionStats(meta.feature = meta.feature[open.peaks, ],query.feature = meta.feature[LINK_Female_tZ_gene_peaks, ],n = 50000)\n",
    "# Motif recovery \n",
    "enriched.motifs_LINK_Female_tZ<- FindMotifs(\n",
    "object = only_STERO,\n",
    "features = LINK_Female_tZ_gene_peaks,           # take vector of peak\n",
    "background = peaks.matched\n",
    ")\n",
    "enriched.motifs_LINK_Female_tZ=enriched.motifs_LINK_Female_tZ[enriched.motifs_LINK_Female_tZ$pvalue<0.05,]\n",
    "\n",
    "\n",
    "        # 2) FEMALE ZF    \n",
    "LINK_DE_STERO_Female_ZF<- LinkPeaks(\n",
    "      object = only_STERO,\n",
    "      peak.assay = \"peaks\",\n",
    "      expression.assay = \"SCT_RUN_STERO\",\n",
    "      genes.use = DE_STERO_Female_ZF,\n",
    "      )\n",
    "          # take peaks with significant correlation\n",
    "      LINK_DE_STERO_Female_ZF=data.frame(Links(LINK_DE_STERO_Female_ZF))\n",
    "      saveRDS(LINK_DE_STERO_Female_ZF,file = 'LINK_DE_STERO_Female_ZF')\n",
    "      LINK_Female_ZF_gene_peaks=LINK_DE_STERO_Female_ZF$peak[LINK_DE_STERO_Female_ZF$score>0.1 &LINK_DE_STERO_Female_ZF$pvalue<0.05 ]\n",
    "\n",
    "       # Background\n",
    "open.peaks <- AccessiblePeaks(only_STERO, idents = c(\"Female_Fasciculata\"),assay = 'peaks') # find a set of peaks that are open in a subset of cells\n",
    "meta.feature <- GetAssayData(only_STERO, assay = \"peaks\", slot = \"meta.features\")# match the overall GC content in the peak set\n",
    "peaks.matched <- MatchRegionStats(meta.feature = meta.feature[open.peaks, ],query.feature = meta.feature[LINK_Female_ZF_gene_peaks, ],n = 50000)\n",
    "# Motif recovery \n",
    "enriched.motifs_LINK_Female_ZF <- FindMotifs(\n",
    "object = only_STERO,\n",
    "features = LINK_Female_ZF_gene_peaks  ,           # take vector of peak\n",
    "background = peaks.matched\n",
    ")\n",
    "\n",
    "  # 3 FEMALE ZG \n",
    "LINK_DE_STERO_Female_ZG<- LinkPeaks(\n",
    "      object = only_STERO,\n",
    "      peak.assay = \"peaks\",\n",
    "      expression.assay = \"SCT_RUN_STERO\",\n",
    "      genes.use = DE_STERO_Female_ZG,\n",
    "      )\n",
    "      # take peaks with significant correlation\n",
    "      LINK_DE_STERO_Female_ZG=data.frame(Links(LINK_DE_STERO_Female_ZG))\n",
    "      saveRDS(LINK_DE_STERO_Female_ZG,file = 'LINK_DE_STERO_Female_ZG')\n",
    "      LINK_Female_ZG_gene_peaks=LINK_DE_STERO_Female_ZG$peak[LINK_DE_STERO_Female_ZG$score>0.1 &LINK_DE_STERO_Female_ZG$pvalue<0.05 ]\n",
    "      \n",
    "             # Background\n",
    "open.peaks <- AccessiblePeaks(only_STERO, idents = c(\"Female_Glomerulosa\"),assay = 'peaks') # find a set of peaks that are open in a subset of cells\n",
    "meta.feature <- GetAssayData(only_STERO, assay = \"peaks\", slot = \"meta.features\")# match the overall GC content in the peak set\n",
    "peaks.matched <- MatchRegionStats(meta.feature = meta.feature[open.peaks, ],query.feature = meta.feature[LINK_Female_ZG_gene_peaks, ],n = 50000)\n",
    "# Motif recovery \n",
    "enriched.motifs_LINK_Female_ZG <- FindMotifs(\n",
    "object = only_STERO,\n",
    "features = LINK_Female_ZG_gene_peaks  ,           # take vector of peak\n",
    "background = peaks.matched\n",
    ")\n",
    "\n",
    "   \n",
    "      \n",
    "# ON MALE TZ        \n",
    "LINK_DE_STERO_Male_tZ<- LinkPeaks(\n",
    "      object = only_STERO,\n",
    "      peak.assay = \"peaks\",\n",
    "      expression.assay = \"SCT_RUN_STERO\",\n",
    "      genes.use = DE_STERO_Male_tZ,\n",
    "      )\n",
    "\n",
    "      # take peaks with significant correlation\n",
    "      LINK_DE_STERO_Male_tZ=data.frame(Links(LINK_DE_STERO_Male_tZ))\n",
    "      saveRDS(LINK_DE_STERO_Male_tZ,file = 'LINK_DE_STERO_Male_tZ')\n",
    "      LINK_DE_STERO_Male_tZ_gene_peaks=LINK_DE_STERO_Male_tZ$peak[LINK_DE_STERO_Male_tZ$score>0.1 &LINK_DE_STERO_Male_tZ$pvalue<0.05 ]\n",
    "      \n",
    "             # Background\n",
    "open.peaks <- AccessiblePeaks(only_STERO, idents = c(\"Male_tZone\"),assay = 'peaks') # find a set of peaks that are open in a subset of cells\n",
    "meta.feature <- GetAssayData(only_STERO, assay = \"peaks\", slot = \"meta.features\")# match the overall GC content in the peak set\n",
    "peaks.matched <- MatchRegionStats(meta.feature = meta.feature[open.peaks, ],query.feature = meta.feature[LINK_DE_STERO_Male_tZ_gene_peaks, ],n = 50000)\n",
    "# Motif recovery \n",
    "enriched.motifs_LINK_Male_tZ <- FindMotifs(\n",
    "object = only_STERO,\n",
    "features = LINK_DE_STERO_Male_tZ_gene_peaks  ,           # take vector of peak\n",
    "background = peaks.matched\n",
    ")\n",
    "\n",
    "\n",
    "\n",
    "LINK_DE_STERO_Male_ZF=readRDS(file = 'D:/Stage/TRUTH/LINK_DE_STERO_Male_ZF')\n",
    "\n",
    "         \n",
    "only_STERO<- LinkPeaks(\n",
    "      object = only_STERO,\n",
    "      peak.assay = \"peaks\",\n",
    "      expression.assay = \"SCT_RUN_STERO\",\n",
    "      genes.use = DE_STERO_Male_ZF$gene_names,\n",
    "      )\n",
    "      # take peaks with significant correlation\n",
    "      LINK_DE_STERO_Male_ZF=data.frame(Links(LINK_DE_STERO_Male_ZF))\n",
    "      saveRDS(LINK_DE_STERO_Male_ZF,file = 'LINK_DE_STERO_Male_ZF')\n",
    "      LINK_DE_STERO_Male_ZF_gene_peaks=LINK_DE_STERO_Male_ZF$peak[LINK_DE_STERO_Male_ZF$score>0.1 &LINK_DE_STERO_Male_ZF$pvalue<0.05 ]\n",
    "      \n",
    "             # Background\n",
    "open.peaks <- AccessiblePeaks(only_STERO, idents = c(\"Male_Fasciculata\"),assay = 'peaks') # find a set of peaks that are open in a subset of cells\n",
    "meta.feature <- GetAssayData(only_STERO, assay = \"peaks\", slot = \"meta.features\")# match the overall GC content in the peak set\n",
    "peaks.matched <- MatchRegionStats(meta.feature = meta.feature[open.peaks, ],query.feature = meta.feature[LINK_DE_STERO_Male_ZF_gene_peaks, ],n = 10000)\n",
    "# Motif recovery \n",
    "enriched.motifs_LINK_Male_ZF <- FindMotifs(\n",
    "object = only_STERO,\n",
    "features = LINK_DE_STERO_Male_ZF_gene_peaks  ,           # take vector of peak\n",
    "background = peaks.matched,assay = 'peaks'\n",
    ")\n",
    "\n",
    "enriched.motifs_LINK_Male_ZF=enriched.motifs_LINK_Male_ZF[enriched.motifs_LINK_Male_ZF$pvalue<0.001,]\n",
    "\n",
    "\n",
    "\n",
    "          \n",
    "LINK_DE_STERO_Male_ZG<- LinkPeaks(\n",
    "      object = only_STERO,\n",
    "      peak.assay = \"peaks\",\n",
    "      expression.assay = \"SCT_RUN_STERO\",\n",
    "      genes.use = DE_STERO_Male_ZG,\n",
    "      )\n",
    "      # take peaks with significant correlation\n",
    "      LINK_DE_STERO_Male_ZG=data.frame(Links(LINK_DE_STERO_Male_ZG))\n",
    "      saveRDS(LINK_DE_STERO_Male_ZG,file = 'LINK_DE_STERO_Male_ZG')\n",
    "      LINK_DE_STERO_Male_ZG_gene_peaks=LINK_DE_STERO_Male_ZG$peak[LINK_DE_STERO_Male_ZG$score>0.1 &LINK_DE_STERO_Male_ZG$pvalue<0.05 ]\n",
    "      \n",
    "             # Background\n",
    "open.peaks <- AccessiblePeaks(only_STERO, idents = c(\"Male_Glomerulosa\"),assay = 'peaks') # find a set of peaks that are open in a subset of cells\n",
    "meta.feature <- GetAssayData(only_STERO, assay = \"peaks\", slot = \"meta.features\")# match the overall GC content in the peak set\n",
    "peaks.matched <- MatchRegionStats(meta.feature = meta.feature[open.peaks, ],query.feature = meta.feature[LINK_DE_STERO_Male_ZG_gene_peaks, ],n = 50000)\n",
    "# Motif recovery \n",
    "enriched.motifs_LINK_Male_ZG <- FindMotifs(\n",
    "object = only_STERO,\n",
    "features = LINK_DE_STERO_Male_ZG_gene_peaks  ,           # take vector of peak\n",
    "background = peaks.matched\n",
    ")\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "74713f29",
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "TRY ALL TEST GENE REGULATORY NETWORK"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8c0d0c19",
   "metadata": {
    "lines_to_next_cell": 0,
    "name": "create motif x feature matrix"
   },
   "outputs": [],
   "source": [
    "\n",
    "library(motifmatchr)\n",
    "library(Signac)\n",
    "library(Seurat)\n",
    "library(JASPAR2020)\n",
    "library(TFBSTools)\n",
    "library(BSgenome.Mmusculus.UCSC.mm10)\n",
    "library(patchwork)\n",
    "LINK_DE_STERO_Female_tZ=readRDS('D:/Stage/TRUTH/LINK_DE_STERO_Female_tZ')\n",
    "\n",
    "# LINK PEAK with DEG\n",
    "DefaultAssay(only_STERO)='peaks'\n",
    "Idents(only_STERO)=only_STERO@meta.data$leiden_merging\n",
    "\n",
    "            # 1) female TZ \n",
    "only_STERO<- LinkPeaks(\n",
    "      object = only_STERO,\n",
    "      peak.assay = \"peaks\",\n",
    "      expression.assay = \"SCT_RUN_STERO\",\n",
    "      genes.use = DE_STERO_Male_ZF$gene_names,\n",
    "      )\n",
    "      # take peaks with significant correlation\n",
    "      LINK_DE_STERO_Female_tZ=data.frame(Links(LINK_DE_STERO_Female_tZ))\n",
    "  LINK_DE_STERO_Female_ZG=readRDS('D:/Stage/TRUTH/LINK_DE_STERO_Male_ZF')\n",
    "      LINK_DE_STERO_Female_ZG_peaks=LINK_DE_STERO_Female_ZG$peak[LINK_DE_STERO_Female_ZG$score > 0.1 &LINK_DE_STERO_Female_ZG$pvalue<0.05 ]\n",
    "      \n",
    "      \n",
    "# Background\n",
    "      \n",
    "Idents(only_STERO)=only_STERO$leiden_merging\n",
    "open.peaks <- AccessiblePeaks(only_STERO, idents = c(\"Male_Fasciculata\"),assay = 'peaks') # find a set of peaks that are open in a subset of cells\n",
    "meta.feature <- GetAssayData(only_STERO, assay = \"peaks\", slot = \"meta.features\")# match the overall GC content in the peak set\n",
    "peaks.matched <- MatchRegionStats(meta.feature = meta.feature[open.peaks, ],query.feature = meta.feature[LINK_DE_STERO_Female_ZG_peaks, ],n = 1000)\n",
    "# Motif recovery \n",
    "enriched.motifs_LINK_Female_tZ<- FindMotifs(\n",
    "object = only_STERO,\n",
    "features = LINK_DE_STERO_Female_ZG_peaks,           # take vector of peak\n",
    "background = peaks.matched,assay = 'peaks'\n",
    ")\n",
    "\n",
    "enriched.motifs_LINK_Female_tZ=enriched.motifs_LINK_Female_tZ[enriched.motifs_LINK_Female_tZ$pvalue<0.05,]\n",
    "\n",
    "# 1) prepare motif from jaspar \n",
    "pfm <- getMatrixSet(\n",
    "  x = JASPAR2020,\n",
    "  opts = list(collection = \"CORE\", tax_group = 'vertebrates', all_versions = FALSE)\n",
    ")\n",
    "\n",
    "# take Granges object from only_STERO\n",
    "motif.matrix <- CreateMotifMatrix(\n",
    "  features =StringToGRanges(LINK_DE_STERO_Female_ZG_peaks ,sep = c(\"-\", \"-\")),  # GET RIGHT PEAKS\n",
    "  pwm = pfm,\n",
    "  genome = BSgenome.Mmusculus.UCSC.mm10,score = F\n",
    ")\n",
    "\n",
    "\n",
    "motif.matrix =as.data.frame(motif.matrix)\n",
    "dim(motif.matrix)\n",
    "motif.matrix=motif.matrix[, intersect( colnames(motif.matrix),enriched.motifs_LINK_Female_tZ$motif)]\n",
    "\n",
    "\n",
    "motif.matrix$peaks=rownames(motif.matrix)\n",
    "motif.matrix$peaks = (gsub(\"\\\\.\", \"-\", motif.matrix$peaks))\n",
    "\n",
    "motif.matrix=left_join(enriched.motifs_LINK_Female_tZ,motif.matrix, by = c(\"peak\"=\"peaks\"))\n",
    "motif.matrix=motif.matrix[!apply(is.na(motif.matrix[,11:length(motif.matrix)]), 1, all),]\n",
    "\n",
    "######\n",
    "var_names <- enriched.motifs_LINK_Female_tZ %>% \n",
    "  # new variable names, old variable names\n",
    "  select(motif.name, motif) %>% \n",
    "  deframe()\n",
    "\n",
    "df_updated <-  motif.matrix %>% \n",
    "  rename(!!!var_names)\n",
    "\n",
    "# find all tf expressed in at least 10% of the cells \n",
    "Idents(only_STERO)=only_STERO$leiden_merging\n",
    "all_male_ZF=FindMarkers(only_STERO,assay = 'SCT_RUN_STERO',logfc.threshold = 0,min.pct = 0.2,ident.1 =\"Male_Fasciculata\" )  \n",
    "all_male_ZF$gene_names <- rownames(all_male_ZF)\n",
    "all_male_ZF$val2 <- mouse_TF$V1[match(all_male_ZF$gene_names, mouse_TF$V1)]\n",
    "\n",
    "\n",
    "drops=c(na.omit(c(toupper(all_male_ZF$val2),all_male_ZF$val2))) # vector of TF present in 20% (en maj et minuscule)\n",
    "\n",
    "\n",
    "retain=as.vector(drops)\n",
    "\n",
    "new_df <- df_updated[,which(names(df_updated) %in% names(df_updated)[1:10] | names(df_updated) %in% retain)]\n",
    "\n",
    "new_df <- new_df[new_df$Ar !=0,]\n",
    "\n",
    "\n",
    "idents.plot <- c(levels(only_STERO))\n",
    "\n",
    "\n",
    "DefaultAssay(only_STERO)='peaks'\n",
    " CoveragePlot(\n",
    "  object = only_STERO,\n",
    "  region = \"chr8-84634863-84635340\",\n",
    "  expression.assay = \"SCT_RUN_STERO\",features = c('Junb'),\n",
    "  idents = idents.plot,\n",
    "  extend.upstream = 5000,\n",
    "  extend.downstream = 5000,links = F\n",
    "  \n",
    ")\n",
    "\n",
    "Idents(only_STERO)=only_STERO$two_conditions\n",
    "idents.plot <- c(\"Male_tZone_male\",\"Male_Fasciculata_male\",\"Male_Glomerulosa_male\",\"Capsular_cells_male\",   \"Female_Glomerulosa_female\",\"Female_tZone_female\",\"Female_Fasciculata_female\", \"Capsular_cells_female\")\n",
    "\n",
    "Idents(only_STERO)=only_STERO$leiden_merging\n",
    "\n",
    "CoveragePlot(only_STERO, features = \"B4galt5\",region =\"B4galt5\" )\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "685c6e86",
   "metadata": {},
   "source": [
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "229e4743",
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "# MOTIF RECOVERY ON LINK PEAKS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0fd88ee3",
   "metadata": {
    "lines_to_next_cell": 2,
    "name": "MOTIF RECOVERY ON LINK PEAKS"
   },
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "  open.peaks <- AccessiblePeaks(multiomics, idents = c(\"Female Fasciculata\"),assay = 'peaks') # find a set of peaks that are open in a subset of cells\n",
    "      \n",
    "      meta.feature <- GetAssayData(multiomics, assay = \"peaks\", slot = \"meta.features\")# match the overall GC content in the peak set\n",
    "      peaks.matched <- MatchRegionStats(\n",
    "        meta.feature = meta.feature[open.peaks, ],\n",
    "        query.feature = meta.feature[peak_LINK_F_F, ],\n",
    "        n = 50000\n",
    "      )\n",
    "\n",
    "\n",
    "\n",
    "enriched.motifs_male <- FindMotifs(\n",
    "  object = multiomics,\n",
    "  features = peak_LINK_F_M  ,           # take vector of peak\n",
    "   background = peaks.matched\n",
    ")\n",
    "\n",
    "enriched.motifs_female <- FindMotifs(\n",
    "  object = multiomics,\n",
    "  features = peak_LINK_F_F  ,           # take vector of peak\n",
    "   background = peaks.matched\n",
    ")\n",
    "\n",
    "\n",
    "\n",
    "View(enriched.motifs_female)\n",
    "View(enriched.motifs_male)\n",
    "\n",
    "enriched.motifs_male=enriched.motifs_male[enriched.motifs_male$fold.enrichment>2 &enriched.motifs_male$pvalue<0.05 ,]\n",
    "\n",
    "enriched.motifs_female=enriched.motifs_female[enriched.motifs_female$fold.enrichment>2 &enriched.motifs_female$pvalue<0.05 ,]\n",
    "\n",
    "female_up = enriched.motifs_female$motif[order(enriched.motifs_female$fold.enrichment, decreasing = TRUE)]\n",
    "female_up=head(female_up)\n",
    "\n",
    "male_up = enriched.motifs_male$motif[order(enriched.motifs_male$fold.enrichment, decreasing = TRUE)]\n",
    "male_up=head(male_up)\n",
    "\n",
    "DefaultAssay(multiomics) <- 'chromvar'\n",
    "\n",
    "p2 <- FeaturePlot(\n",
    "  object = multiomics,\n",
    "  features = male_up,\n",
    "  min.cutoff = 'q10',\n",
    "  max.cutoff = 'q90',\n",
    "  pt.size = 0.1,reduction = 'wnn.umap'\n",
    ")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c72d3e1b",
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "# differential peaks accesibility AND motif accessibility"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e2127ae5",
   "metadata": {
    "lines_to_next_cell": 0,
    "name": "differential peaks accesibility AND motif accessibility"
   },
   "outputs": [],
   "source": [
    "\n",
    "Idents(only_STERO)=only_STERO@meta.data$leiden_merging\n",
    "DefaultAssay(only_STERO)='peaks'\n",
    "vector_clusters_F=c( \"Female_Fasciculata\",\"Female_Glomerulosa\",\"Female_tZone\")\n",
    "vector_clusters_M=c( \"Male_Fasciculata\", \"Male_Glomerulosa\",\"Male_tZone\")\n",
    "\n",
    "\n",
    "for( i in 1:length(vector_clusters_F)){ \n",
    "  assign(paste0('DA_STERO_',vector_clusters_F[i]),FindMarkers(only_STERO,\n",
    "                                                        ident.1 =vector_clusters_F[i],\n",
    "                                                        ident.2 = vector_clusters_M[i],\n",
    "                                                        assay = 'peaks',\n",
    "                                                        only.pos = F, test.use = 'LR',\n",
    "                                                        min.pct = 0.05,\n",
    "                                                        latent.vars = 'nCount_peaks'))\n",
    "}\n",
    "\n",
    "\n",
    "\n",
    "# FASCICULATA \n",
    "DA_STERO_Female_Fasciculata$peaks=rownames(DA_STERO_Female_Fasciculata)\n",
    "DA_STERO_Female_Fasciculata <- DA_STERO_Female_Fasciculata[DA_STERO_Female_Fasciculata$p_val_adj < 0.05,]\n",
    "\n",
    "peaks_Fem_FASC=DA_STERO_Female_Fasciculata$peaks[DA_STERO_Female_Fasciculata$p_val_adj < 0.05&DA_STERO_Female_Fasciculata$avg_log2FC>0.1]\n",
    "peaks_Male_FASC=DA_STERO_Female_Fasciculata$peaks[DA_STERO_Female_Fasciculata$p_val_adj < 0.05&DA_STERO_Female_Fasciculata$avg_log2FC< -0.1]\n",
    "\n",
    "\n",
    "# GLOMERULOSA \n",
    "DA_STERO_Female_Glomerulosa$peaks=rownames(DA_STERO_Female_Glomerulosa)\n",
    "DA_STERO_Female_Glomerulosa <- DA_STERO_Female_Glomerulosa[DA_STERO_Female_Glomerulosa$p_val_adj < 0.05,]\n",
    "\n",
    "peaks_Fem_GLO=DA_STERO_Female_Glomerulosa$peaks[DA_STERO_Female_Glomerulosa$p_val_adj < 0.05&DA_STERO_Female_Glomerulosa$avg_log2FC>0.1]\n",
    "peaks_Male_GLO=DA_STERO_Female_Glomerulosa$peaks[DA_STERO_Female_Glomerulosa$p_val_adj < 0.05&DA_STERO_Female_Glomerulosa$avg_log2FC< -0.1]\n",
    "\n",
    "\n",
    "#  Tzone\n",
    "DA_STERO_Female_tZone$peaks=rownames(DA_STERO_Female_tZone)\n",
    "DA_STERO_Female_tZone <- DA_STERO_Female_tZone[DA_STERO_Female_tZone$p_val_adj < 0.05,]\n",
    "\n",
    "peaks_Fem_tzone=DA_STERO_Female_tZone$peaks[DA_STERO_Female_tZone$p_val_adj < 0.05&DA_STERO_Female_tZone$avg_log2FC>0.1]\n",
    "peaks_Male_tzone=DA_STERO_Female_tZone$peaks[DA_STERO_Female_tZone$p_val_adj < 0.05&DA_STERO_Female_tZone$avg_log2FC< -0.1]\n",
    "\n",
    "\n",
    "# Background and motif enrichement :\n",
    "              \n",
    "              # Fasciculata : \n",
    "                      open.peaks <- AccessiblePeaks(only_STERO, idents = c(\"Female_Fasciculata\",\"Male_Fasciculata\"),assay = 'peaks') \n",
    "                      \n",
    "                      meta.feature <- GetAssayData(only_STERO, assay = \"peaks\", slot = \"meta.features\")\n",
    "                      # Female \n",
    "                            peaks.matched <- MatchRegionStats(\n",
    "                            meta.feature = meta.feature[open.peaks, ],\n",
    "                            query.feature = meta.feature[peaks_Fem_FASC, ],\n",
    "                            n = 50000\n",
    "                            )\n",
    "                              \n",
    "                            enrich_motif_Fem_FASC <- FindMotifs(object = only_STERO,features = peaks_Fem_FASC,background = peaks.matched)\n",
    "                            enrich_motif_Fem_FASC <- enrich_motif_Fem_FASC[enrich_motif_Fem_FASC$pvalue<0.05,]\n",
    "                                          \n",
    "                      # Male \n",
    "                            peaks.matched <- MatchRegionStats(\n",
    "                            meta.feature = meta.feature[open.peaks, ],\n",
    "                            query.feature = meta.feature[peaks_Male_FASC, ],\n",
    "                            n = 50000\n",
    "                            )\n",
    "                            enrich_motif_Male_FASC <- FindMotifs(object = only_STERO,features = peaks_Male_FASC,background = peaks.matched)\n",
    "                            enrich_motif_Male_FASC <- enrich_motif_Male_FASC[enrich_motif_Male_FASC$pvalue<0.05,]\n",
    "\n",
    "              # Glomerulosa : \n",
    "                      open.peaks <- AccessiblePeaks(only_STERO, idents = c(\"Male_Glomerulosa\",\"Female_Glomerulosa\"),assay = 'peaks') \n",
    "                      \n",
    "                      meta.feature <- GetAssayData(only_STERO, assay = \"peaks\", slot = \"meta.features\")\n",
    "                      # Female \n",
    "                            peaks.matched <- MatchRegionStats(\n",
    "                            meta.feature = meta.feature[open.peaks, ],\n",
    "                            query.feature = meta.feature[peaks_Fem_GLO, ],\n",
    "                            n = 50000\n",
    "                            )\n",
    "                              \n",
    "                            enrich_motif_Fem_GLO<- FindMotifs(object = only_STERO,features = peaks_Fem_GLO,background = peaks.matched)\n",
    "                            enrich_motif_Fem_GLO <- enrich_motif_Fem_GLO[enrich_motif_Fem_GLO$pvalue<0.05,]\n",
    "\n",
    "                                          \n",
    "                      # Male \n",
    "                            peaks.matched <- MatchRegionStats(\n",
    "                            meta.feature = meta.feature[open.peaks, ],\n",
    "                            query.feature = meta.feature[peaks_Male_GLO, ],\n",
    "                            n = 50000\n",
    "                            )\n",
    "                            enrich_motif_Male_GLO <- FindMotifs(object = only_STERO,features = peaks_Male_GLO,background = peaks.matched)\n",
    "                            enrich_motif_Male_GLO <- enrich_motif_Male_GLO[enrich_motif_Male_GLO$pvalue<0.05,]\n",
    "\n",
    "                            \n",
    "                # TZONE  : \n",
    "                      open.peaks <- AccessiblePeaks(only_STERO, idents = c(\"Female_tZone\",\"Male_tZone\"),assay = 'peaks') \n",
    "                      \n",
    "                      meta.feature <- GetAssayData(only_STERO, assay = \"peaks\", slot = \"meta.features\")\n",
    "                      # Female \n",
    "                            peaks.matched <- MatchRegionStats(\n",
    "                            meta.feature = meta.feature[open.peaks, ],\n",
    "                            query.feature = meta.feature[peaks_Fem_tzone, ],\n",
    "                            n = 50000\n",
    "                            )\n",
    "                              \n",
    "                            enrich_motif_Fem_TZONE<- FindMotifs(object = only_STERO,features = peaks_Fem_tzone,background = peaks.matched)\n",
    "                            enrich_motif_Fem_TZONE <- enrich_motif_Fem_TZONE[enrich_motif_Fem_TZONE$pvalue<0.05,]\n",
    "                                          \n",
    "                      # Male \n",
    "                            peaks.matched <- MatchRegionStats(\n",
    "                            meta.feature = meta.feature[open.peaks, ],\n",
    "                            query.feature = meta.feature[peaks_Male_tzone, ],\n",
    "                            n = 50000\n",
    "                            )\n",
    "                            enrich_motif_Male_TZONE <- FindMotifs(object = only_STERO,features = peaks_Male_tzone,background = peaks.matched)\n",
    "                            enrich_motif_Male_TZONE <- enrich_motif_Male_TZONE[enrich_motif_Male_TZONE$pvalue<0.05,]\n",
    "                            \n",
    "                            \n",
    "\n",
    "                          \n",
    "                            \n",
    "\n",
    "\n",
    "# ON SUB CLUSTERS\n",
    "\n",
    "Idents(only_STERO)=only_STERO@meta.data$new_ident\n",
    "DefaultAssay(only_STERO)='peaks'\n",
    "vector_clusters_F=c( \"Female_Fasciculata_1\", \"Male_Fasciculata_1\")\n",
    "vector_clusters_M=c( \"Female_Fasciculata_2\", \"Male_Fasciculata_2\")\n",
    "\n",
    "for( i in 1:length(vector_clusters_F)){\n",
    "  assign(paste0('DA_STERO_SUB',vector_clusters_F[i]),FindMarkers(only_STERO,\n",
    "                                                        ident.1 =vector_clusters_F[i],\n",
    "                                                        ident.2 = vector_clusters_M[i],\n",
    "                                                        assay = 'peaks',\n",
    "                                                        only.pos = F, test.use = 'LR',\n",
    "                                                        min.pct = 0.05,\n",
    "                                                        latent.vars = 'nCount_peaks'))\n",
    "}\n",
    "\n",
    "\n",
    "\n",
    "DA_STERO_SUBFemale_Fasciculata_1$peaks=rownames(DA_STERO_SUBFemale_Fasciculata_1)\n",
    "DA_STERO_SUBFemale_Fasciculata_1 <- DA_STERO_SUBFemale_Fasciculata_1[DA_STERO_SUBFemale_Fasciculata_1$p_val_adj < 0.05,]\n",
    "\n",
    "\n",
    "DA_STERO_SUBMale_Fasciculata_1$peaks=rownames(DA_STERO_SUBMale_Fasciculata_1)\n",
    "DA_STERO_SUBMale_Fasciculata_1<- DA_STERO_SUBMale_Fasciculata_1[DA_STERO_SUBMale_Fasciculata_1$p_val_adj < 0.05,]\n",
    "\n",
    "\n",
    "\n",
    "# do a bar plot of total number of DE AND DA between glomerulosa and fasciculata with and without sex\n",
    "\n",
    "# run it on homer for finding differentially MOTIF and TF recruitment \n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "360838af",
   "metadata": {},
   "source": [
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d40d7683",
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "# Motif get jaspar database"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "95136149",
   "metadata": {
    "lines_to_next_cell": 2,
    "name": "Motif get jaspar database"
   },
   "outputs": [],
   "source": [
    "\n",
    "# 1) prepare motif from jaspar \n",
    "pfm <- getMatrixSet(\n",
    "  x = JASPAR2020,\n",
    "  opts = list(collection = \"CORE\", tax_group = 'vertebrates', all_versions = FALSE)\n",
    ")\n",
    "\n",
    "# take Granges object from only_STERO\n",
    "motif.matrix <- as.data.frame( CreateMotifMatrix(\n",
    "  features =StringToGRanges( DA_STERO_Female_Fasciculata$peaks[DA_STERO_Female_Fasciculata$avg_log2FC< - 0.25 & DA_STERO_Female_Fasciculata$p_val_adj<0.01],sep = c(\"-\", \"-\")),  # GET RIGHT PEAKS\n",
    "  pwm = pfm,\n",
    "  genome = BSgenome.Mmusculus.UCSC.mm10,score = F\n",
    "))\n",
    "\n",
    "\n",
    "motif.matrix =as.data.frame(motif.matrix)\n",
    "dim(motif.matrix)\n",
    "motif.matrix=motif.matrix[, intersect( colnames(motif.matrix),enrich_motif_Male_FASC$motif)] # take enriched motif \n",
    "\n",
    "\n",
    "motif.matrix$peaks=rownames(motif.matrix)\n",
    "motif.matrix$peaks = (gsub(\"\\\\.\", \"-\", motif.matrix$peaks))\n",
    "\n",
    "motif.matrix=left_join(enrich_motif_Fem_FASC,motif.matrix, by = c(\"peak\"=\"peaks\"))\n",
    "motif.matrix=motif.matrix[!apply(is.na(motif.matrix[,11:length(motif.matrix)]), 1, all),]\n",
    "\n",
    "######\n",
    "var_names <- enriched.motifs_LINK_Female_tZ %>% \n",
    "  # new variable names, old variable names\n",
    "  select(motif.name, motif) %>% \n",
    "  deframe()\n",
    "\n",
    "df_updated <-  motif.matrix %>% \n",
    "  rename(!!!var_names)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f92ab3ee",
   "metadata": {
    "lines_to_next_cell": 0,
    "name": "MOTIF slot adding  PASS IF ALREADY DONE"
   },
   "outputs": [],
   "source": [
    "library(Seurat)\n",
    "library(Signac)\n",
    "library(TFBSTools)\n",
    "library(rlist)\n",
    "library(BSgenome.Mmusculus.UCSC.mm10)\n",
    "library(JASPAR2020)\n",
    "\n",
    "DefaultAssay(only_STERO)='peaks'\n",
    "\n",
    "# 1) prepare motif from jaspar \n",
    "pfm <- getMatrixSet(\n",
    "  x = JASPAR2020,\n",
    "  opts = list(collection = \"CORE\", tax_group = 'vertebrates', all_versions = FALSE)\n",
    ")\n",
    "  \n",
    "  # 2) add motif information\n",
    "      # construct a Motif object and adds it to our multiomics dataset, along with other information such as the base composition                 of each peak. \n",
    "only_STERO <- AddMotifs(\n",
    "  object = only_STERO,\n",
    "  genome = BSgenome.Mmusculus.UCSC.mm10,\n",
    "  pfm = pfm\n",
    ")\n",
    "\n",
    "only_STERO <- RegionStats(\n",
    "  only_STERO, \n",
    "  genome = BSgenome.Mmusculus.UCSC.mm10,\n",
    "  assay = 'peaks')  \n",
    "\n",
    "\n",
    " # Run chromvar  \n",
    "\n",
    "only_STERO <- RunChromVAR(\n",
    "  object = only_STERO,\n",
    "  genome = BSgenome.Mmusculus.UCSC.mm10,\n",
    "  assay = \"peaks\"\n",
    ")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "271e45da",
   "metadata": {},
   "source": [
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "48c62c2b",
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "####DATA FRAME annotation add ALL ID "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1141ee8f",
   "metadata": {
    "lines_to_next_cell": 2,
    "name": "DATA FRAME annotation add ALL ID"
   },
   "outputs": [],
   "source": [
    "# GSEA Fast comparison and dataframe annotation ENSEMBLID / ENTREZID (Kegg)\n",
    "\n",
    "# annotate according to Ensembl and kegg  DATABASE\n",
    "library(AnnotationHub)\n",
    "library(dplyr)\n",
    "library(clusterProfiler)\n",
    "library(fgsea)\n",
    "library(ggplot2)\n",
    "library(enrichplot)\n",
    "library(org.Mm.eg.db)\n",
    "library(DOSE)\n",
    "library(biomaRt)\n",
    "# get mouse genes mm10\n",
    "ensembl <- useMart(\"ensembl\", dataset=\"mmusculus_gene_ensembl\")\n",
    "\n",
    "                                              # DE_STERO_Female_tZ\n",
    "# get ensembl ID\n",
    "gene_IDs <- getBM(filters= 'mgi_symbol', attributes= c(\"ensembl_gene_id\",'mgi_symbol','entrezgene_id','description'),values = DE_STERO_Female_tZ$gene_names, mart= ensembl)\n",
    "# Add into DE DATAFRAME\n",
    "DE_STERO_Female_tZ=left_join(DE_STERO_Female_tZ,gene_IDs, by = c(\"gene_names\"=\"mgi_symbol\"))\n",
    "# check duplicate value \n",
    "DE_STERO_Female_tZ$gene[duplicated(DE_STERO_Female_tZ$gene_names)]\n",
    "# remove duplicate value\n",
    "DE_STERO_Female_tZ=DE_STERO_Female_tZ[!duplicated(DE_STERO_Female_tZ$gene_names), ]\n",
    "write.table(DE_STERO_Female_tZ$gene_names,file = 'DE_STERO_Female_tZ.txt',quote = F,row.names = F,col.names = F)\n",
    "\n",
    "                                              # DE_STERO_Male_tZ\n",
    "gene_IDs <- getBM(filters= 'mgi_symbol', attributes= c(\"ensembl_gene_id\",'mgi_symbol','entrezgene_id','description'),values = DE_STERO_Male_tZ$gene_names, mart= ensembl)\n",
    "# Add into DE DATAFRAME\n",
    "DE_STERO_Male_tZ=left_join(DE_STERO_Male_tZ,gene_IDs, by = c(\"gene_names\"=\"mgi_symbol\"))\n",
    "# check duplicate value \n",
    "DE_STERO_Male_tZ$gene[duplicated(DE_STERO_Male_tZ$gene_names)]\n",
    "# remove duplicate value\n",
    "DE_STERO_Male_tZ=DE_STERO_Male_tZ[!duplicated(DE_STERO_Male_tZ$gene_names), ]\n",
    "write.table(DE_STERO_Male_tZ$gene_names,file = 'DE_STERO_Male_tZ.txt',quote = F,row.names = F,col.names = F)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "674e9f7d",
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "# trajectory "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "05059166",
   "metadata": {
    "lines_to_next_cell": 0,
    "name": "DE FOR TRAJECTORY"
   },
   "outputs": [],
   "source": [
    "# re find clusters with high resolution \n",
    "#only_STERO<- FindClusters(only_STERO, graph.name = \"wsnn\", algorithm = 1, resolution = 1, verbose = FALSE)\n",
    "DimPlot(only_STERO,reduction  = 'wnn.umap',label = T,pt.size = 2)\n",
    "# create new IDENTS FOR SEPARING CAPSULAR CELL MALE AND FEMALE\n",
    "Idents(only_STERO)=only_STERO$merging_ident\n",
    "only_STERO$merging_ident_classed_by_sex <- paste(Idents(only_STERO), only_STERO$identity, sep = \"_\")\n",
    "levels(only_STERO)\n",
    "\n",
    "new.cluster.ids <- c(\"Male_Glomerulosa_male\"=\"Male_Glomerulosa\",\n",
    "                      \"Male_Fasciculata_male\"= \"Male_Fasciculata\", \n",
    "                      \"Capsular_Cells_male\" = \"Capsular_Cells_male\",\n",
    "                     \"Female_Glomerulosa_male\"=\"Female_Glomerulosa\",\n",
    "                     \"Female_Glomerulosa_female\"=\"Female_Glomerulosa\",\n",
    "                     \"Female_Fasciculata_female\" =\"Female_Fasciculata\",\n",
    "                     \"Capsular_Cells_female\"='Capsular_Cells_female' ,\n",
    "                     \"Male_Glomerulosa_female\" =\"Male_Glomerulosa\",\n",
    "                     \"Male_Fasciculata_female\"='Male_Fasciculata')\n",
    "\n",
    "\n",
    "\n",
    "names(new.cluster.ids) <- levels(only_STERO)\n",
    "only_STERO <- RenameIdents(only_STERO, new.cluster.ids)\n",
    "DimPlot(pbmc, reduction = \"umap\", label = TRUE, pt.size = 0.5) + NoLegend()\n",
    "only_STERO$merging_ident_classed_by_sex=Idents(only_STERO)\n",
    "\n",
    "\n",
    "# END \n",
    "\n",
    "# RUN  DE ==>\n",
    "\n",
    "mouse_TF=read.table('C:/Users/User/Documents/aposimz/mm_mgi_tfs.txt')\n",
    "\n",
    "mouse_TF=read.table('C:/Users/User/Documents/aposimz/mm_mgi_tfs.txt')\n",
    "chrx_y=as.data.frame(chrX_and_Y)\n",
    "\n",
    "\n",
    "# Findmarkers on the new idents \n",
    "DefaultAssay(remove_only_stero)='SCT_RUN_STERO'   \n",
    "Idents(remove_only_stero)=remove_only_stero$leiden_annotated\n",
    "# CLUSTERS  1 - MALE \n",
    "Male_Glomerulosa = FindMarkers(remove_only_stero,ident.1 =\"Male_Glomerulosa\",ident.2 = c(\"Male_tZone\",\"Female_Glomerulosa\"),assay = 'SCT_RUN_STERO')\n",
    "Male_Glomerulosa$gene_names=rownames(Male_Glomerulosa)\n",
    "Male_Glomerulosa$val2 <- mouse_TF$V1[match(Male_Glomerulosa$gene_names, mouse_TF$V1)]\n",
    "Male_Glomerulosa$sex <- chrx_y$chrX_and_Y[match(Male_Glomerulosa$gene_names, chrx_y$chrX_and_Y)]\n",
    "Male_Glomerulosa = Male_Glomerulosa[Male_Glomerulosa$p_val_adj<0.05&Male_Glomerulosa$avg_log2FC>0.5,]\n",
    "\n",
    "\n",
    "Male_tZone = FindMarkers(remove_only_stero,ident.1 ='Male_tZone',ident.2 = c(\"Male_Glomerulosa\",\"Male_Fasciculata_1\",\"Female_tZone\"),assay = 'SCT_RUN_STERO')\n",
    "Male_tZone$gene_names=rownames(Male_tZone)\n",
    "Male_tZone$val2 <- mouse_TF$V1[match(Male_tZone$gene_names, mouse_TF$V1)]\n",
    "Male_tZone$sex <- chrx_y$chrX_and_Y[match(Male_tZone$gene_names, chrx_y$chrX_and_Y)]\n",
    "Male_tZone = Male_tZone[Male_tZone$p_val_adj<0.05&Male_tZone$avg_log2FC>0.5,]\n",
    "\n",
    "\n",
    "Male_Fasciculata_1 = FindMarkers(remove_only_stero,ident.1 ='Male_Fasciculata_1',ident.2 = c(\"Male_tZone\",\"Male_Fasciculata_2\",\"Female_Fasciculata_1\"),assay = 'SCT_RUN_STERO')\n",
    "Male_Fasciculata_1$gene_names=rownames(Male_Fasciculata_1)\n",
    "Male_Fasciculata_1$val2 <- mouse_TF$V1[match(Male_Fasciculata_1$gene_names, mouse_TF$V1)]\n",
    "Male_Fasciculata_1$sex <- chrx_y$chrX_and_Y[match(Male_Fasciculata_1$gene_names, chrx_y$chrX_and_Y)]\n",
    "Male_Fasciculata_1 = Male_Fasciculata_1[Male_Fasciculata_1$p_val_adj<0.05&Male_Fasciculata_1$avg_log2FC>0.5,]\n",
    "\n",
    "\n",
    "\n",
    "Female_Glomerulosa = FindMarkers(remove_only_stero,ident.1 ='Female_Glomerulosa',ident.2 = c(\"Female_tZone\",'Male_Glomerulosa'),assay = 'SCT_RUN_STERO')\n",
    "Female_Glomerulosa$gene_names=rownames(Female_Glomerulosa)\n",
    "Female_Glomerulosa$val2 <- mouse_TF$V1[match(Female_Glomerulosa$gene_names, mouse_TF$V1)]\n",
    "Female_Glomerulosa$sex <- chrx_y$chrX_and_Y[match(Female_Glomerulosa$gene_names, chrx_y$chrX_and_Y)]\n",
    "Female_Glomerulosa = Female_Glomerulosa[Female_Glomerulosa$p_val_adj<0.05&Female_Glomerulosa$avg_log2FC>0.5,]\n",
    "\n",
    "\n",
    "getwd()\n",
    "Female_tZone = FindMarkers(remove_only_stero,ident.1 ='Female_tZone',ident.2 = c(\"Female_Glomerulosa\",\"Female_Fasciculata_1\",\"Male_tZone\"),assay = 'SCT_RUN_STERO')\n",
    "Female_tZone$gene_names=rownames(Female_tZone)\n",
    "Female_tZone$val2 <- mouse_TF$V1[match(Female_tZone$gene_names, mouse_TF$V1)]\n",
    "Female_tZone$sex <- chrx_y$chrX_and_Y[match(Female_tZone$gene_names, chrx_y$chrX_and_Y)]\n",
    "Female_tZone = Female_tZone[Female_tZone$p_val_adj<0.05&Female_tZone$avg_log2FC>0.5,]\n",
    "\n",
    "\n",
    "Female_Fasciculata_1 = FindMarkers(remove_only_stero,ident.1 ='Female_Fasciculata_1',ident.2 = c(\"Female_tZone\",\"Female_Fasciculata_2\",\"Male_Fasciculata_1\"),assay = 'SCT_RUN_STERO')\n",
    "Female_Fasciculata_1$gene_names=rownames(Female_Fasciculata_1)\n",
    "Female_Fasciculata_1$val2 <- mouse_TF$V1[match(Female_Fasciculata_1$gene_names, mouse_TF$V1)]\n",
    "Female_Fasciculata_1$sex <- chrx_y$chrX_and_Y[match(Female_Fasciculata_1$gene_names, chrx_y$chrX_and_Y)]\n",
    "Female_Fasciculata_1 = Female_Fasciculata_1[Female_Fasciculata_1$p_val_adj<0.05&Female_Fasciculata_1$avg_log2FC>0.5,]\n",
    "\n",
    "\n",
    "Female_Fasciculata_2 = FindMarkers(remove_only_stero,ident.1 ='Female_Fasciculata_2',ident.2 = c('Female_Fasciculata_1',\"Male_Fasciculata_2\"),assay = 'SCT_RUN_STERO')\n",
    "Female_Fasciculata_2$gene_names=rownames(Female_Fasciculata_2)\n",
    "Female_Fasciculata_2$val2 <- mouse_TF$V1[match(Female_Fasciculata_2$gene_names, mouse_TF$V1)]\n",
    "Female_Fasciculata_2$sex <- chrx_y$chrX_and_Y[match(Female_Fasciculata_2$gene_names, chrx_y$chrX_and_Y)]\n",
    "Female_Fasciculata_2 = Female_Fasciculata_2[Female_Fasciculata_2$p_val_adj<0.05&Female_Fasciculata_2$avg_log2FC>0.5,]\n",
    "\n",
    "\n",
    "Male_Fasciculata_2 = FindMarkers(remove_only_stero,ident.1 ='Male_Fasciculata_2',ident.2 = c(\"Male_Fasciculata_1\",\"Female_Fasciculata_2\"),assay = 'SCT_RUN_STERO')\n",
    "Male_Fasciculata_2$gene_names=rownames(Male_Fasciculata_2)\n",
    "Male_Fasciculata_2$val2 <- mouse_TF$V1[match(Male_Fasciculata_2$gene_names, mouse_TF$V1)]\n",
    "Male_Fasciculata_2$sex <- chrx_y$chrX_and_Y[match(Male_Fasciculata_2$gene_names, chrx_y$chrX_and_Y)]\n",
    "Male_Fasciculata_2 = Male_Fasciculata_2[Male_Fasciculata_2$p_val_adj<0.05&Male_Fasciculata_2$avg_log2FC>0.5,]\n",
    "\n",
    "\n",
    "# CAPSULAR CELL BY CONDITIONS \n",
    "Idents(only_STERO)=only_STERO$leiden_merging\n",
    "only_STERO$two_conditions <- paste(Idents(only_STERO), only_STERO$identity, sep = \"_\")\n",
    "Idents(only_STERO) <- only_STERO$two_conditions\n",
    "\n",
    "\n",
    "Capsular_cells_female = FindMarkers(only_STERO,ident.1 ='Capsular_cells_female',ident.2 = c('Capsular_cells_male'),logfc.threshold = 0.1,assay = 'SCT_RUN_STERO')\n",
    "Capsular_cells_female$gene_names=rownames(Capsular_cells_female)\n",
    "Capsular_cells_female$val2 <- mouse_TF$V1[match(Capsular_cells_female$gene_names, mouse_TF$V1)]\n",
    "Capsular_cells_female$sex <- chrx_y$chrX_and_Y[match(Capsular_cells_female$gene_names, chrx_y$chrX_and_Y)]\n",
    "Capsular_cells_female = Capsular_cells_female[Capsular_cells_female$p_val_adj<0.05,]\n",
    "\n",
    "\n",
    "Capsular_cells_male = FindMarkers(remove_only_stero,ident.1 ='Capsular_cells_male',ident.2 = c('Capsular_cells_female'))\n",
    "Capsular_cells_male$gene_names=rownames(Capsular_cells_male)\n",
    "Capsular_cells_male$val2 <- mouse_TF$V1[match(Capsular_cells_male$gene_names, mouse_TF$V1)]\n",
    "Capsular_cells_male$sex <- chrX_and_Y$chrX_and_Y[match(Capsular_cells_male$gene_names, chrX_and_Y$chrX_and_Y)]\n",
    "Capsular_cells_male = Capsular_cells_male[Capsular_cells_male$p_val_adj<0.05,]\n",
    "\n",
    "\n",
    "\n",
    "Capsular_cells_male = FindMarkers(remove_only_stero,ident.1 ='Capsular_cells_male',ident.2 = c('Capsular_cells_female'))\n",
    "\n",
    "\n",
    "Capsular_cells_female = FindMarkers(only_STERO,ident.1 ='Capsular_cells_female',ident.2 = c('Capsular_cells_male'),assay = 'peaks', test.use = 'LR', latent.vars = 'nCount_peaks', min.pct = 0.05)\n",
    "\n",
    "\n",
    "\n",
    "for( i in 1:length(vector_clusters_F)){ \n",
    "  assign(paste0('DA_STERO_',vector_clusters_F[i]),FindMarkers(only_STERO,\n",
    "                                                        ident.1 =vector_clusters_F[i],\n",
    "                                                        ident.2 = vector_clusters_M[i],\n",
    "                                                        assay = 'peaks',\n",
    "                                                        only.pos = F, test.use = 'LR',\n",
    "                                                        min.pct = 0.05,\n",
    "                                                        latent.vars = 'nCount_peaks'))\n",
    "}\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "Idents(only_STERO) <- only_STERO$leiden_annotated\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1a4dc91c",
   "metadata": {},
   "source": [
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "440f6ecd",
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "can pass if already run "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0cf4a7ef",
   "metadata": {
    "lines_to_next_cell": 2,
    "name": "MOTIF slot adding  PASS IF ALREADY DONE"
   },
   "outputs": [],
   "source": [
    "library(Seurat)\n",
    "library(Signac)\n",
    "library(TFBSTools)\n",
    "library(rlist)\n",
    "library(BSgenome.Mmusculus.UCSC.mm10)\n",
    "library(JASPAR2020)\n",
    "\n",
    "DefaultAssay(only_STERO)='peaks'\n",
    "\n",
    "# 1) prepare motif from jaspar \n",
    "pfm <- getMatrixSet(\n",
    "  x = JASPAR2020,\n",
    "  opts = list(collection = \"CORE\", tax_group = 'vertebrates', all_versions = FALSE)\n",
    ")\n",
    "  \n",
    "  # 2) add motif information\n",
    "      # construct a Motif object and adds it to our multiomics dataset, along with other information such as the base composition                 of each peak. \n",
    "only_STERO <- AddMotifs(\n",
    "  object = only_STERO,\n",
    "  genome = BSgenome.Mmusculus.UCSC.mm10,\n",
    "  pfm = pfm\n",
    ")\n",
    "\n",
    "only_STERO <- RegionStats(\n",
    "  only_STERO, \n",
    "  genome = BSgenome.Mmusculus.UCSC.mm10,\n",
    "  assay = 'peaks')  \n",
    "\n",
    "\n",
    " # Run chromvar  \n",
    "\n",
    "only_STERO <- RunChromVAR(\n",
    "  object = only_STERO,\n",
    "  genome = BSgenome.Mmusculus.UCSC.mm10,\n",
    "  assay = \"peaks\"\n",
    ")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5bed96dd",
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "# plot motif along pseudotime"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "77abe34c",
   "metadata": {
    "lines_to_next_cell": 0,
    "name": "HEATMAP plot motif along pseudotime"
   },
   "outputs": [],
   "source": [
    "\n",
    "library(dittoSeq)\n",
    "Idents(only_STERO) <- only_STERO$two_conditions\n",
    "Idents(only_STERO) <- only_STERO$leiden_merging\n",
    "\n",
    "Idents(only_STERO)=only_STERO$two_conditions\n",
    "\n",
    "\n",
    "\n",
    "#Female\n",
    "\n",
    "issou=subset(only_STERO,idents = c(\"Female_Glomerulosa_female\",\"Female_tZone_female\",\"Female_Fasciculata_female\",\"Male_Glomerulosa_male\",\"Male_tZone_male\",\"Male_Fasciculata_male\"))\n",
    "\n",
    "female_traj=read.csv(file = 'E:/TRUTH/Pseudotime_female.csv')\n",
    "names(female_traj)[1]='index'\n",
    "male_traj=read.csv(file = 'E:/TRUTH/Pseudotime_male.csv')\n",
    "names(male_traj)[1]='index'\n",
    "out=rbind(x = female_traj,male_traj)\n",
    "rownames(out)=out$index\n",
    "\n",
    "\n",
    "xd$new_issoux <- out$Pseudotime\n",
    "xd$Pseudotime <- out$Pseudotime[match(xd$index, out$index)]\n",
    "\n",
    "\n",
    "issox= c('identity','val2')\n",
    "\n",
    "names(xd@meta.data)[46]='Cell types'\n",
    "names(xd@meta.data)[4]='Sex'\n",
    "\n",
    "\n",
    "\n",
    "dittoHeatmap(xd,genes=c('Fat3','Ly6d','Lynx1','Crem','Fam129a','Slc25a3','Pde10a','Akr1b7','Srd5a2','Slc25a30','Abcb1b','Aldh1a1','Runx2','Agtr1b','Vsnl1','Cdh6','Ecrg4','Pip5k1b','Elmo1','Efna5','Ror1','Adh1','Pdzrn3','A530020G20Rik','Akr1d1','Lama4','Cyp2f2')\n",
    "             ,show_colnames = F\n",
    "             ,order.by = c('Sex','Pseudotime')\n",
    "             ,annot.by = c('Cell types','Sex','Pseudotime')\n",
    "             ,assay = 'SCT_RUN_STERO'\n",
    "             ,slot = 'scale.data'\n",
    "             ,cluster_rows=FALSE\n",
    "             ,complex = TRUE,\n",
    "heatmap.colors=colorRampPalette(c(\"royalblue\", \"white\", \"red\"))(50),\n",
    "breaks=seq(-4, 4, length.out=51)\n",
    ")\n",
    "\n",
    "    breaks = seq(0, 0.65, length.out = 26)\n",
    "\n",
    "\n",
    "\n",
    "heatmap.colors=colorRampPalette(c(\"blue\", \"white\", \"red\"))(50),\n",
    "breaks=seq(-4, 4, length.out=50)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "genix=c(\"Agtr1b\",\"Vsnl1\",\"Elmo1\",\"Igfbp7\",\"Agtr1b\",\"Vsnl1\",\"Pde7a\" ,\"Pdzrn3\",\"Akr1d1\" ,\"Prr16\",\"Il1rapl2\") \n",
    "\n",
    "breaks = seq(-6, 6, 0.06))\n",
    "\n",
    "\n",
    "genix=c(\"Agtr1b\",\"Vsnl1\",\"Elmo1\",\"Igfbp7\",\"Agtr1b\",\"Vsnl1\",\"Pde7a\" ,\"Pdzrn3\",\"Akr1d1\" ,\"Prr16\",\"Il1rapl2\") \n",
    "dittoHeatmap(Male,order.by = 'new_X1z',annot.by = 'leiden_merging',assay = 'SCT_RUN_STERO',genes=all_markers_MG,show_colnames = F,  complex = TRUE,use_raster = TRUE,scale = \"none\",cluster_cols = FALSE,heatmap.colors = colorRampPalette(c(\"blue\",\"firebrick4\",\"firebrick4\",\"firebrick4\"))(4),scaled.to.max = T,slot = 'scale.data')\n",
    "\n",
    "\n",
    "\n",
    "dittoHeatmap(x,order.by = 'new_X1z',annot.by = 'leiden_merging',assay = 'SCT_RUN_STERO',genes=all_markers_MG,show_colnames = F,  complex = TRUE,use_raster = TRUE,scale = \"none\",cluster_cols = FALSE,heatmap.colors = colorRampPalette(c(\"blue\",\"firebrick4\",\"firebrick4\",\"firebrick4\"))(4),scaled.to.max = T,slot = 'scale.data')\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "all_markers_FEMALE=c('Agtr1b','Vsnl1','cdh6','Ecgr4',\n",
    "                    'Pip5k1b','Elmo1','Efna5','Ror1',\n",
    "                    'Adh1','Pdzrn3','A530020G20Rik',\n",
    "                    'Akr1d1','Cyp2f2','Lama4','Cyp2f2'\n",
    "                    )\n",
    "\n",
    "Female_tZone_genes=Female_tZone$gene_names[Female_tZone$p_val_adj<0.05]\n",
    "Male_GLOMER8ULOSA_genes=Female_Glomerulosa$gene_names[Female_Glomerulosa$p_val_adj<0.05]\n",
    "Male_tZone_genes=Female_Fasciculata_2$gene_names[Female_Fasciculata_2$p_val_adj<0.05&Female_Fasciculata_2$avg_log2FC>1]\n",
    "Male_tZone_genes=Male_tZone$gene_names[Male_tZone$p_val_adj<0.05,]\n",
    "Male_tZone_genes=Male_tZone$gene_names[Male_tZone$p_val_adj<0.05]\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "library(ComplexHeatmap)\n",
    "library(circlize)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "af263319",
   "metadata": {},
   "source": [
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "348e4454",
   "metadata": {
    "lines_to_next_cell": 2,
    "name": "Reticulate"
   },
   "outputs": [],
   "source": [
    "library(reticulate)\n",
    "py$anndata['paga']\n",
    "\n",
    "\n",
    "# add embending \n",
    "embedding <- py$anndata$obsm[\"X_draw_graph_fr\"]\n",
    "rownames(embedding) <- py$anndata$obs_names$to_list()\n",
    "colnames(embedding) <- c(\"FA_1\", \"FA_2\")\n",
    "only_STERO[[\"FA\"]] <- CreateDimReducObject(embedding, key = \"FA_\")\n",
    "# ad dmeta data\n",
    "only_STERO <- AddMetaData(only_STERO, py$anndata$obs)\n",
    "\n",
    "# FIND ALL MARKERS FOR LEIDEN RE CLUSTERING \n",
    "Idents(only_STERO)=only_STERO$leiden\n",
    "allclusters=FindAllMarkers(only_STERO)\n",
    "write.csv(allclusters,file = 'allclusters.csv',sep = ,)\n",
    "write.csv(data, \"data.csv\") \n",
    "DimPlot(only_STERO,reduction = 'FA')|DimPlot(only_STERO,reduction = 'FA',group.by = 'identity')\n",
    "setwd('D:/Stage/')\n",
    "# DEG\n",
    "all_stero_diff=FindMarkers(only_STERO,ident.1 = c('1','7','3','5'),ident.2 = c('2','6','0','4'),split.by = 'identity')\n",
    "\n",
    "all_stero_diff=all_stero_diff[all_stero_diff$p_val_adj<0.05,]\n",
    "all_stero_diff$gene_names=rownames(all_stero_diff)\n",
    "\n",
    "# male\n",
    "male_diff=all_stero_diff[all_stero_diff$avg_log2FC<0,]\n",
    "male_diff$TF <- mouse_TF$V1[match(male_diff$gene_names, mouse_TF$V1)]\n",
    "male_diff$sex <- chrx_y$chrX_and_Y[match(male_diff$gene_names, chrx_y$chrX_and_Y)]\n",
    "\n",
    "# female\n",
    "female_diff=all_stero_diff[all_stero_diff$avg_log2FC > 0,]\n",
    "female_diff$TF <- mouse_TF$V1[match(female_diff$gene_names, mouse_TF$V1)]\n",
    "female_diff$sex <- chrx_y$chrX_and_Y[match(female_diff$gene_names, chrx_y$chrX_and_Y)]\n",
    "\n",
    "\n",
    "#\n",
    "all_genes=subset(only_STERO,idents =\"Capsular_Cells\" ,invert=T,)\n",
    "virus=FindAllMarkers(all_genes,logfc.threshold=0,min.pct=0.05,assay = 'SCT_RUN_STERO')\n",
    "\n",
    "    X1iX1=FindMarkers(only_STERO,ident.1 = c('1','7','3','5'),ident.2 = c('2','6','0','4'),min.pct = 0.05,logfc.threshold = 0,assay = 'SCT_RUN_STERO')\n",
    "    \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1d0b1ac1",
   "metadata": {
    "lines_to_next_cell": 0,
    "name": "PLOT motif"
   },
   "outputs": [],
   "source": [
    "library(lemon)\n",
    "library(stringr)\n",
    "library(data.table)\n",
    "library(stringr)\n",
    "library(ggplot2)\n",
    "\n",
    "# 1) loading objects\n",
    "# trajectory coordinates generated by scan^py\n",
    "trajFA = read.csv(\"D:/Stage/Pseudotime_RNA.csvFA2_components_ATAC.csv\")\n",
    "colnames(trajFA) = c(\"index\",\"FA1\",\"FA2\")\n",
    "\n",
    "minmaxthis = function(x) { \n",
    "  (x - min(x))/(max(x) - min(x))\n",
    "}\n",
    "\n",
    "trajFA=only_STERO@reductions[[\"FA\"]]@cell.embeddings\n",
    "\n",
    "# plot figure \n",
    "TFsnames = c(\"MA0007.3\", \"MA0113.3\", \"MA0727.1\", \"MA1655.1\" ,\"MA0102.4\", \"MA0836.2\",\"MA1636.1\")\n",
    "plot_list = list() \n",
    "\n",
    "    for (x in TFsnames) {\n",
    "      print(x)\n",
    "      \n",
    "      DefaultAssay(only_STERO) = \"chromvar\"\n",
    "      p10 = FeaturePlot(\n",
    "        object = only_STERO,\n",
    "        features = str_replace_all(x, fixed(\"_\"), \"-\"),\n",
    "        min.cutoff = 'q0',\n",
    "        max.cutoff = 'q100',\n",
    "        pt.size = 0.3)\n",
    "      \n",
    "      chromvardf = p10$data\n",
    "      chromvardf$index = rownames(chromvardf)\n",
    "      \n",
    "      mergedFAchromvar = merge(chromvardf, trajFA, by = \"index\")\n",
    "      \n",
    "      \n",
    "      TFactplot = ggplot(mergedFAchromvar, aes( x = FA1, y = FA2, \n",
    "                                                colour = minmaxthis(mergedFAchromvar[,5]))) + \n",
    "        geom_point(size = 1, alpha = 1) + \n",
    "        scale_color_gradient(low = \"gray93\",high = \"red1\",breaks = seq(0,1,0.2))+\n",
    "        xlab(\"FA1\") + ylab(\"FA2\") +\n",
    "        theme(panel.border=element_blank(), \n",
    "              axis.line = element_line(colour = 'black', size = 1), \n",
    "              axis.ticks = element_blank(),\n",
    "              axis.text = element_blank(), \n",
    "              panel.grid.major = element_blank(),\n",
    "              panel.grid.minor = element_blank(),\n",
    "              panel.background = element_blank(),\n",
    "              legend.key=element_blank(),\n",
    "              legend.title=element_blank(),\n",
    "              legend.text=element_text(size=12)) + \n",
    "        # guides(color = guide_legend(override.aes max.cutoff= list(size=5))) + \n",
    "        coord_capped_cart(bottom='right', left='none')  + \n",
    "      ggtitle(x)\n",
    "      TFactplot\n",
    "      \n",
    "      \n",
    "      print(TFactplot)\n",
    "  \n",
    "}\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "x=FindMarkers(only_STERO,ident.1 =\"Female_Fasciculata_1\",ident.2 = \"Female_Fasciculata_2\",assay = 'chromvar')\n",
    "\n",
    "\n",
    "\n",
    "TFsnames = c(\"MA0007.3\", \"MA0113.3\", \"MA0727.1\", \n",
    "             \n",
    "             \"MA1655.1\" ,\"MA0102.4\", \"MA0836.2\",\"MA1636.1\")\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "trajFA=as.data.frame(only_STERO@reductions[[\"wnn.umap\"]]@cell.embeddings)\n",
    "trajFA$index=rownames(trajFA)\n",
    "\n",
    "minmaxthis = function(x) { \n",
    "  (x - min(x))/(max(x) - min(x))\n",
    "}\n",
    "\n",
    "\n",
    "plot_list = list() \n",
    "\n",
    "\n",
    "    for (x in TFsnames) {\n",
    "      print(x)\n",
    "      \n",
    "      DefaultAssay(only_STERO) = \"chromvar\"\n",
    "      p10 = FeaturePlot(\n",
    "        object = only_STERO,\n",
    "        features = str_replace_all(x, fixed(\"_\"), \"-\"),\n",
    "        min.cutoff = 'q0',\n",
    "        max.cutoff = 'q100',\n",
    "        pt.size = 0.2)\n",
    "      \n",
    "      chromvardf = p10$data\n",
    "      chromvardf$index = rownames(chromvardf)\n",
    "      \n",
    "      mergedFAchromvar = merge(chromvardf, trajFA, by = \"index\")\n",
    "      \n",
    "      \n",
    "      TFactplot = ggplot(mergedFAchromvar, aes( x = FA_1, y=FA_2,colour = minmaxthis(mergedFAchromvar[,5]))) + \n",
    "        geom_point(size = 1, alpha = 1) + \n",
    "        scale_color_gradient(low = \"gray93\",high = \"red1\",breaks = seq(0,1,0.2))+\n",
    "        xlab(\"FA1\") + ylab(\"FA2\") +\n",
    "        theme(panel.border=element_blank(), \n",
    "              axis.line = element_line(colour = 'black', size = 1), \n",
    "              axis.ticks = element_blank(),\n",
    "              axis.text = element_blank(), \n",
    "              panel.grid.major = element_blank(),\n",
    "              panel.grid.minor = element_blank(),\n",
    "              panel.background = element_blank(),\n",
    "              legend.key=element_blank(),\n",
    "              legend.title=element_blank(),\n",
    "              legend.text=element_text(size=12)) + \n",
    "        # guides(color = guide_legend(override.aes = list(size=5))) + \n",
    "        coord_capped_cart(bottom='right', left='none')  + \n",
    "      ggtitle(x)\n",
    "      TFactplot\n",
    "      \n",
    "      \n",
    "      print(TFactplot)\n",
    "  \n",
    "}\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cec680e0",
   "metadata": {},
   "source": [
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "52936441",
   "metadata": {
    "lines_to_next_cell": 0,
    "name": "Leak peak to genes"
   },
   "outputs": [],
   "source": [
    "library(BSgenome.Mmusculus.UCSC.mm10)\n",
    "library(Signac)\n",
    "# # link peaks to genes Glomerulosa\n",
    "   #  compute gene to link with peaks :\n",
    "\n",
    "            Female_Fasc_gene=DE_STERO_Female_Fasciculata$gene_names[DE_STERO_Female_Fasciculata$avg_log2FC> 0]   #Testing 300 genes and 178944 peaks Found gene coordinates for 278 genes\n",
    "            Male_Fasc_gene=DE_STERO_Female_Fasciculata$gene_names[DE_STERO_Female_Fasciculata$avg_log2FC< 0]     #Testing 287 genes and 178944 peaks Found gene coordinates for 279 genes\n",
    "            \n",
    "            Female_Glo_gene=DE_STERO_Female_Glomerulosa$gene_names[DE_STERO_Female_Glomerulosa$avg_log2FC> 0]    #Testing 69 genes and 178944 peaks Found gene coordinates for 65 genes\n",
    "            Male_Glo_gene=DE_STERO_Female_Glomerulosa$gene_names[DE_STERO_Female_Glomerulosa$avg_log2FC< 0]      #Testing 39 genes and 178944 peaks Found gene coordinates for 37 genes\n",
    "\n",
    "# LINK PEAK with DEG\n",
    "LINK_Female_Fasc_gene<- LinkPeaks(\n",
    "      object = only_STERO,\n",
    "      peak.assay = \"peaks\",\n",
    "      expression.assay = \"SCT_RUN_STERO\",\n",
    "      genes.use = Female_Fasc_gene,\n",
    "      )\n",
    "LINK_Male_Fasc_gene<- LinkPeaks(\n",
    "    object = only_STERO,\n",
    "    peak.assay = \"peaks\",\n",
    "    expression.assay = \"SCT_RUN_STERO\",\n",
    "    genes.use = Male_Fasc_gene,\n",
    "    )\n",
    "LINK_Female_Glo_gene<- LinkPeaks(\n",
    "    object = only_STERO,\n",
    "    peak.assay = \"peaks\",\n",
    "    expression.assay = \"SCT_RUN_STERO\",\n",
    "    genes.use = Female_Glo_gene,\n",
    "    )\n",
    "LINK_Male_Glo_gene<- LinkPeaks(\n",
    "    object = only_STERO,\n",
    "    peak.assay = \"peaks\",\n",
    "    expression.assay = \"SCT_RUN_STERO\",\n",
    "    genes.use = Male_Glo_gene,\n",
    "    )\n",
    "\n",
    "\n",
    "# take peaks with significant correlation\n",
    "LINK_Female_Fasc_gene=data.frame(Links(LINK_Female_Fasc_gene))\n",
    "saveRDS(LINK_Female_Fasc_gene,file = 'LINK_Female_Fasc_gene')\n",
    "LINK_Female_Fasc_gene_peaks=LINK_Female_Fasc_gene$peak[LINK_Female_Fasc_gene$score>0.1 &LINK_Female_Fasc_gene$pvalue<0.05 ]\n",
    "\n",
    "\n",
    "LINK_Male_Fasc_gene=data.frame(Links(LINK_Male_Fasc_gene))\n",
    "saveRDS(LINK_Male_Fasc_gene,file = 'LINK_Male_Fasc_gene')\n",
    "LINK_Male_Fasc_gene_peaks=LINK_Male_Fasc_gene$peak[LINK_Male_Fasc_gene$score>0.1 &LINK_Male_Fasc_gene$pvalue<0.05]\n",
    "\n",
    "\n",
    "LINK_Female_Glo_gene=data.frame(Links(LINK_Female_Glo_gene))\n",
    "saveRDS(LINK_Female_Glo_gene,file = 'LINK_Female_Glo_gene')\n",
    "LINK_Female_Glo_gene_peaks=LINK_Female_Glo_gene$peak[LINK_Female_Glo_gene$score>0.1 &LINK_Female_Glo_gene$pvalue<0.05 ]\n",
    "\n",
    "\n",
    "LINK_Male_Glo_gene=data.frame(Links(LINK_Male_Glo_gene))\n",
    "saveRDS(LINK_Male_Glo_gene,file = 'LINK_Male_Glo_gene')\n",
    "LINK_Male_Glo_gene_peaks=LINK_Male_Glo_gene$peak[LINK_Male_Glo_gene$score>0.1 &LINK_Male_Glo_gene$pvalue<0.05 ]\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "50958cc2",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "c65b6248",
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "https://bookdown.org/ytliu13207/SingleCellMultiOmicsDataAnalysis/scenic.html"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0a47031f",
   "metadata": {
    "lines_to_next_cell": 0,
    "name": "RUN SCENIC FOUND REGULON"
   },
   "outputs": [],
   "source": [
    "library(Seurat)\n",
    "library(tidyverse)\n",
    "library(magrittr)\n",
    "library(SCENIC)\n",
    "library(SingleCellExperiment)\n",
    "\n",
    "cellInfo <- data.frame(only_STERO=Idents(only_STERO))\n",
    "getwd()\n",
    "only_STERO=readRDS('D:/Stage/aposimz/only_STERO_FINAL')\n",
    "\n",
    "\n",
    "\n",
    "LINK_Male_Fasc_gene_peaks"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5beb3873",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "805a25ee",
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "https://stdworkflow.com/282/single-cell-transcriptional-regulatory-network-analysis-tool-tf-motifs-scenic"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4e71652d",
   "metadata": {
    "name": "RUN scenic another tutorial"
   },
   "outputs": [],
   "source": [
    "# 1)Prepare cell meta information\n",
    "cellInfo <- data.frame(only_STERO@meta.data)\n",
    "colnames(cellInfo)[which(colnames(cellInfo)==\"orig.ident\")] <- \"sample\"\n",
    "colnames(cellInfo)[which(colnames(cellInfo)==\"identity\")] <- \"sexe\"\n",
    "colnames(cellInfo)[which(colnames(cellInfo)==\"nFeature_SCT_RUN_STERO\")] <- \"nGene\"\n",
    "colnames(cellInfo)[which(colnames(cellInfo)==\"nCount_SCT_RUN_STERO\")] <- \"nUMI\"\n",
    "colnames(cellInfo)[which(colnames(cellInfo)==\"new_ident\")] <- \"celltype\"\n",
    "colnames(cellInfo)[which(colnames(cellInfo)==\"merging_ident\")] <- \"celltype_merged\"\n",
    "\n",
    "cellInfo <- cellInfo[,c(\"sample\",\"nGene\",\"nUMI\",\"celltype\",\"celltype_merged\",\"sexe\")]\n",
    "saveRDS(cellInfo, file=\"cellInfo.Rds\")\n",
    "\n",
    "\n",
    "# Set the celltype color\n",
    "colVars <- list(celltype=c(\"Male_Fasciculata_1\"=\"forestgreen\", \n",
    "                           \"Female_Fasciculata_2\"=\"darkorange\", \n",
    "                           \"Male_Glomerulosa\"=\"magenta4\", \n",
    "                           \"Female_Glomerulosa\"=\"hotpink\", \n",
    "                           \"Male_Fasciculata_2\"=\"red3\", \n",
    "                           \"Female_Fasciculata_1\"=\"skyblue\", \n",
    "                           \"Capsular_Cells\"=\"darkblue\"))\n",
    "colVars$celltype <- colVars$celltype[intersect(names(colVars$celltype), cellInfo$celltype)]\n",
    "\n",
    "\n",
    "# 2) Initialize SCENIC settings, set up the analysis environment\n",
    "org <- \"mgi\"\n",
    "dbFiles <- 'C:/Users/User/Documents/aposimz/Rcis_target/'\n",
    "myDatasetTitle <- \"SCENIC mouse adrenal \" # choose a name for your analysis\n",
    "data(defaultDbNames)\n",
    "# change to mm10\n",
    "defaultDbNames$mgi[1] <- \"mm10__refseq-r80__10kb_up_and_down_tss.mc9nr.feather\"\n",
    "defaultDbNames$mgi[2] <- \"mm10__refseq-r80__500bp_up_and_100bp_down_tss.mc9nr.feather\"\n",
    "\n",
    "# create scenic object\n",
    "scenicOptions <- initializeScenic(org= \"mgi\", datasetTitle = myDatasetTitle, dbDir=dbFiles,dbs =defaultDbNames[['mgi']],nCores = 4 )\n",
    "\n",
    "scenicOptions@inputDatasetInfo$cellInfo <- cellInfo\n",
    "scenicOptions@inputDatasetInfo$colVars <- colVars\n",
    "\n",
    "# create matrix Build a Co-expression network\n",
    "\n",
    "exprMat <- as.matrix(only_STERO@assays$SCT_RUN_STERO@counts) #Prepare the expression matrix. In order to save computing resources, only some cells are randomly selected to calculate the co-expression network\n",
    "\n",
    "genesKept <- geneFiltering(exprMat, \n",
    "                           scenicOptions=scenicOptions,minCountsPerGene=3*.01*ncol(exprMat),minSamples=ncol(exprMat)*.01) #Gene filtering/selection, remove genes that are most likely to be noise\n",
    "\n",
    "exprMat_filtered <- exprMat[genesKept, ]\n",
    "dim(exprMat_filtered)\n",
    "runCorrelation(exprMat_filtered, scenicOptions) ##Calculate the correlation matrix, 1.2_corrMat.Rds: the correlation matrix between genes\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "saveRDS(scenicOptions, file=\"int/scenicOptions.Rds\")\n",
    "saveRDS(cellInfo, file=\"int/cellInfo.Rds\")\n",
    "saveRDS(colVars, file=\"int/colVars.Rds\")\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "##GENIE3, the input of GENIE3 is usually an expression matrix and a list of candidate regulators.\n",
    "exprMat_filtered <- log2(exprMat_filtered+1) #complete matrix\n",
    "runGenie3(exprMat_filtered, scenicOptions, nParts = 10) #nParts parameter, is to divide the expression matrix into n parts and calculate separately\n",
    "#1.4, GENIE3_linkList.Rds: The final result of GENIE3 is to merge the files starting with \"1.3_\"\n",
    "\n",
    "\n",
    "# Build and score the GRN (runSCENIC_…)\n",
    "\n",
    "exprMat_log <- log2(exprMat+1) #log standard words original matrix\n",
    "#scenicOptions <- readRDS(\"int/scenicOptions.Rds\")\n",
    "scenicOptions@settings$verbose <- TRUE\n",
    "scenicOptions@settings$nCores <- 1\n",
    "scenicOptions@settings$seed <- 123\n",
    "scenicOptions <- runSCENIC_1_coexNetwork2modules(scenicOptions) #1. Get coexpression module\n",
    "scenicOptions <- runSCENIC_2_createRegulons(scenicOptions) #2. Get regulons\n",
    "#You can use the parameter coexMethod=c(\"top5perTarget\"), coexMethod is available. The author tried a variety of strategies to filter low-relevance TF-Target, and suggested that all six filtering criteria should be used. In fact, it does not take much time. Both are calculated by default.\n",
    "#w001: Use each TF as the core to reserve genes with weight>0.001 to form a co-expression module; #w005, #top50\n",
    "#top5perTarget: Each gene retains the weight value of top5 TF to get a simplified TF-Target association table, and then assign the gene to TF to construct a co-expression module; #top10perTarget; #top50perTarget\n",
    "scenicOptions <- runSCENIC_3_scoreCells(scenicOptions, exprMat_log) #3. Scoring the GRN (regulator) in the cell\n",
    "saveRDS(scenicOptions, file=\"int/scenicOptions.Rds\")\n",
    "runSCENIC_4_aucell_binarize(scenicOptions, exprMat=exprMat_log)\n",
    "\n",
    "\n",
    "# ADDING IN SEURAT \n",
    "\n",
    "\n",
    "##Import the original regulonAUC matrix, which is the AUC matrix generated after running runSCENIC_3_scoreCells, you can view the AUC activity score of each GRN in each cell\n",
    "\n",
    "setwd(\"C:/Users/User/Documents/aposimz\")\n",
    "AUCmatrix <- readRDS(\"int/3.4_regulonAUC.Rds\")\n",
    "\n",
    "\n",
    "AUCmatrix <- data.frame(t(AUCmatrix@assays@data@listData$AUC), check.names=F)\n",
    "RegulonName_AUC <- colnames(AUCmatrix)\n",
    "RegulonName_AUC <- gsub(' \\\\(','_',RegulonName_AUC) #replace \"(\" with \"_\"\n",
    "RegulonName_AUC <- gsub('\\\\)','',RegulonName_AUC) # Remove \")\" and finally replace KLF3_extended (79g) with KLF3_extended_79g\n",
    "colnames(AUCmatrix) <- RegulonName_AUC\n",
    "only_STERO <- AddMetaData(only_STERO, AUCmatrix) # Add the AUC matrix to the metadata information of pbmc\n",
    "saveRDS(only_STERO,'only_STERO_final')\n",
    "\n",
    "\n",
    "AUCmatrix <- readRDS(\"int/3.4_regulonAUC.Rds\")\n",
    "regulonAUC <-  AUCmatrix@assays@data$AUC\n",
    "rownames(regulonAUC) <- gsub(\"[(+)]\", \"\", rownames(regulonAUC))\n",
    "\n",
    "only_STERO[['AUC']] <- CreateAssayObject(data = regulonAUC)\n",
    "  \n",
    "# Identify differential regulons based AUC scores\n",
    "\n",
    "DefaultAssay(only_STERO) <- 'AUC'\n",
    "\n",
    "vector_clusters_F=c( \"Female_Fasciculata\",\"Female_Glomerulosa\")\n",
    "vector_clusters_M=c( \"Male_Fasciculata\", \"Male_Glomerulosa\")\n",
    "\n",
    "for( i in 1:length(vector_clusters_F)){\n",
    "  assign(paste0('REGULON_STERO_',vector_clusters_F[i]),FindMarkers(only_STERO,ident.1 =vector_clusters_F[i],ident.2 = vector_clusters_M[i],assay = 'AUC',logfc.threshold = 0.005))\n",
    "\n",
    "}\n",
    "\n",
    "REGULON_STERO_Female_Fasciculata<- REGULON_STERO_Female_Fasciculata[REGULON_STERO_Female_Fasciculata$p_val_adj < 0.05,]\n",
    "REGULON_F_F_up <- REGULON_STERO_Female_Fasciculata[which(REGULON_STERO_Female_Fasciculata$avg_log2FC > 0), ]\n",
    "REGULON_M_F_up <- REGULON_STERO_Female_Fasciculata[which(REGULON_STERO_Female_Fasciculata$avg_log2FC < 0), ]\n",
    "\n",
    "\n",
    "deg.ls <- list(REGULON_F_F_up, REGULON_M_F_up)\n",
    "names(deg.ls) <- c('Female_Fasciculata', 'Male_Fasciculata')\n",
    "\n",
    "gene.ls <- lapply(deg.ls, rownames)\n",
    "\n",
    "DefaultAssay(only_STERO) <- 'AUC'\n",
    "\n",
    "only_STERO <- ScaleData(only_STERO, assay = 'AUC', features = rownames(only_STERO))\n",
    "\n",
    "mg <- unique(Reduce('c', gene.ls[c(2)]))\n",
    "\n",
    "DoHeatmap(only_STERO, features = mg, raster = F,slot = 'scale.data',assay = 'AUC') + scale_fill_viridis()\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "30f8db44",
   "metadata": {
    "lines_to_next_cell": 2,
    "name": "monocle3 trajectory analysis"
   },
   "outputs": [],
   "source": [
    "library(monocle3)\n",
    "library(monocle)\n",
    "library(Seurat)\n",
    "library(Signac)\n",
    "library(Signac)\n",
    "library(Seurat)\n",
    "library(SeuratWrappers)\n",
    "library(monocle3)\n",
    "library(Matrix)\n",
    "library(ggplot2)\n",
    "library(patchwork)\n",
    "\n",
    "#####choose datasets - FEMALE ####\n",
    "female_STERO<-only_STERO[,  only_STERO$identity %in% c('female')]\n",
    "# Subset\n",
    "female_STERO <- SplitObject(female_STERO, split.by = \"Replicat\") # list \n",
    "\n",
    "############# RNA REPROCESSING ##############\n",
    "for (i in 1:length(female_STERO)) {\n",
    "    female_STERO[[i]] <- SCTransform(female_STERO[[i]],variable.features.n =NULL,\n",
    "    variable.features.rv.th =1.1,return.only.var.genes = FALSE,\n",
    "    assay = 'RNA',new.assay.name = 'SCT_INT',vars.to.regress = 'percent.mt')\n",
    "    }\n",
    "# Select the most variable features to use for integration\n",
    "integ_features <-SelectIntegrationFeatures(object.list = female_STERO, nfeatures = 3000) \n",
    "# Prepare the SCT list object for integration\n",
    "female_STERO  <- PrepSCTIntegration(object.list = female_STERO, \n",
    "                                   anchor.features =integ_features)\n",
    "\n",
    "integ_anchors <- FindIntegrationAnchors(object.list = female_STERO, \n",
    "                                        normalization.method = \"SCT\", \n",
    "                                        anchor.features = integ_features,reduction=\"cca\")\n",
    "\n",
    "# Integrate across conditions\n",
    "female_STERO <- IntegrateData(anchorset = integ_anchors,normalization.method = \"SCT\")                 # replace seurat list by object\n",
    "\n",
    "# CLustering \n",
    "DefaultAssay(female_STERO)='integrated'\n",
    "female_STERO=RunPCA(female_STERO, verbose = FALSE,assay=\"integrated\",reduction.name='PCA_INT',npcs = 50)\n",
    "female_STERO=RunICA(female_STERO,nics = 20)\n",
    "ElbowPlot(female_STERO,reduction ='PCA_INT',ndims = 50 )# choose 15\n",
    "female_STERO=FindNeighbors(female_STERO,assay ='integrated',reduction ='PCA_INT', dims = 1:10)\n",
    "female_STERO=FindClusters(female_STERO, resolution = 0.3,algorithm = 1,graph.name = 'integrated_nn')\n",
    "female_STERO=RunUMAP(female_STERO,dims = 1:10, reduction.name = 'UMAP_PCA_RNA_INT',reduction = 'PCA_INT')\n",
    "DimPlot(female_STERO)\n",
    "#####\n",
    "###########\n",
    "###############\n",
    "####################\n",
    "                                    ## ATAC processing  AND PEAKS  generating ## \n",
    "\n",
    "DefaultAssay(female_STERO) <- \"peaks\"\n",
    "female_STERO <- FindTopFeatures(female_STERO, min.cutoff = 5)\n",
    "female_STERO <- RunTFIDF(female_STERO)\n",
    "female_STERO <- RunSVD(female_STERO)\n",
    "DepthCor(female_STERO,n = 50) | ElbowPlot(female_STERO,reduction = 'lsi')\n",
    "\n",
    "female_STERO <- RunUMAP(object = female_STERO,reduction = 'lsi',dims = 2:5,reduction.name = 'UMAP_LSI_ATAC')\n",
    "female_STERO <- FindNeighbors(object = female_STERO,reduction = 'lsi',dims = 2:5) # use SNN for findclusters \n",
    "#only_STERO <- FindClusters(object = only_STERO,algorithm = 1,resolution = 0.7,verbose = FALSE)\n",
    "DimPlot(object = female_STERO, label = TRUE,group.by = 'Replicat',reduction = 'UMAP_LSI_ATAC')+DimPlot(female_STERO,reduction = 'UMAP_PCA_RNA_INT',group.by = 'Replicat')\n",
    "\n",
    "##################################INTEGRATION ON REPLICAT  for atac \n",
    "\n",
    "crazy=female_STERO\n",
    "\n",
    "crazy <- SplitObject(crazy, split.by = \"Replicat\") # list \n",
    "\n",
    "\n",
    "x=SelectIntegrationFeatures(\n",
    "  crazy,\n",
    "  nfeatures = 3000,\n",
    "  assay = c('peaks','peaks'),\n",
    "  verbose = TRUE\n",
    ")\n",
    "\n",
    "\n",
    "# find integration anchors\n",
    "integration.anchors <- FindIntegrationAnchors(\n",
    "  object.list = crazy,\n",
    "  anchor.features =x,\n",
    "  reduction = \"rlsi\",\n",
    "  dims = 2:30\n",
    ")\n",
    "\n",
    "rm(crazy)\n",
    "X1=female_STERO\n",
    "\n",
    "# integrate LSI embeddings\n",
    "integrated <- IntegrateEmbeddings(\n",
    "  anchorset = integration.anchors,\n",
    "  reductions = X1[[\"lsi\"]],\n",
    "  new.reduction.name = \"integrated_lsi\",\n",
    "  dims.to.integrate = 2:30\n",
    ")\n",
    "\n",
    "integrated <- RunUMAP(integrated, reduction = \"integrated_lsi\", dims = 2:20)\n",
    "\n",
    "DimPlot(integrated ,group.by = 'Replicat')\n",
    " \n",
    " \n",
    " #######################ADD INT LSI ON OBJECT \n",
    "\n",
    "female_STERO@reductions$UMAP_LSI_ATAC_INT=integrated@reductions[[\"umap\"]]\n",
    "female_STERO@reductions$integrated_lsi=integrated@reductions[[\"integrated_lsi\"]]\n",
    "\n",
    "\n",
    "\n",
    "DimPlot(object = female_STERO, label = TRUE,group.by = 'Replicat',reduction = 'UMAP_LSI_ATAC')+DimPlot(female_STERO,reduction = 'UMAP_PCA_RNA_INT',group.by = 'Replicat')+DimPlot(object = female_STERO, label = TRUE,group.by = 'Replicat',reduction = 'UMAP_LSI_ATAC_INT')\n",
    "\n",
    "\n",
    "\n",
    "female_STERO <- FindMultiModalNeighbors(\n",
    "  female_STERO, reduction.list = list(\"PCA_INT\", \"integrated_lsi\"), \n",
    "  dims.list = list(1:10, 2:5),weighted.nn.name = 'X1'\n",
    ")\n",
    "female_STERO <- RunUMAP(female_STERO, nn.name = \"X1\", reduction.name = \"wnn.umap\", reduction.key = \"wnnUMAP_\")\n",
    "female_STERO=FindClusters(female_STERO,graph.name = 'wknn',resolution = 0.5)\n",
    "\n",
    "DimPlot(female_STERO,reduction = 'wnn.umap')\n",
    "female_STERO$new_ident=Idents(female_STERO)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "                                            ### TRAJECTORY FEMALE ####\n",
    "\n",
    "Female_glomerulosa_root <- female_STERO[, female_STERO$new_ident %in% c(\"4\")]\n",
    "vector_root_F=colnames(Female_glomerulosa_root)\n",
    "\n",
    "\n",
    "female_STERO@reductions$UMAP=female_STERO@reductions[[\"UMAP_PCA_RNA_INT\"]]\n",
    "Female.cds <- as.cell_data_set(female_STERO,assay = 'SCT_RUN_STERO')\n",
    "Female.cds <- cluster_cells(cds = Female.cds, reduction_method = \"UMAP\",resolution=30)\n",
    "\n",
    "# find all possible partitions\n",
    "all_partitions <- unique(Female.cds@clusters$UMAP$partitions)\n",
    "all_partitions <- all_partitions[all_partitions != \"1\"]\n",
    "# set all partitions to 1\n",
    "Female.cds@clusters$UMAP$partitions[Female.cds@clusters$UMAP$partitions %in% all_partitions] <- \"1\"\n",
    "\n",
    "Female.cds <- learn_graph(Female.cds, use_partition = F)\n",
    "\n",
    "Female.cds <- order_cells(Female.cds, reduction_method = \"UMAP\" ,root_cells =vector_root_F )\n",
    "plot_cells(\n",
    "  cds = Female.cds ,\n",
    "  color_cells_by = \"pseudotime\",\n",
    "  show_trajectory_graph = TRUE,cell_size = 2,trajectory_graph_segment_size = 2,graph_label_size = 2,label_cell_groups = F,label_branch_points = T,)\n",
    "\n",
    "## add to only stero\n",
    "\n",
    "only_STERO<- AddMetaData(\n",
    "  object = only_STERO,\n",
    "  metadata = Female.cds@principal_graph_aux@listData$UMAP$pseudotime,\n",
    "  col.name = \"RNA_ONLY_Female_trajectory\"\n",
    ")\n",
    "\n",
    "FeaturePlot(only_STERO, c(\"WNN_ONLY_Female_trajectory\"), pt.size = 1,reduction = 'wnn.umap',label = T) & scale_color_viridis_c()\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "#####choose datasets - MALE ####\n",
    "male_STERO<-only_STERO[,  only_STERO$identity %in% c('male')]\n",
    "# Subset\n",
    "male_STERO <- SplitObject(male_STERO, split.by = \"Replicat\") # list \n",
    "\n",
    "############# RNA REPROCESSING ##############\n",
    "for (i in 1:length(male_STERO)) {\n",
    "    male_STERO[[i]] <- SCTransform(male_STERO[[i]],variable.features.n =NULL,\n",
    "    variable.features.rv.th =1.1,return.only.var.genes = FALSE,\n",
    "    assay = 'RNA',new.assay.name = 'SCT_INT',vars.to.regress = 'percent.mt')\n",
    "    }\n",
    "# Select the most variable features to use for integration\n",
    "integ_features <-SelectIntegrationFeatures(object.list = male_STERO, nfeatures = 3000) \n",
    "# Prepare the SCT list object for integration\n",
    "male_STERO  <- PrepSCTIntegration(object.list = male_STERO, \n",
    "                                   anchor.features =integ_features)\n",
    "\n",
    "integ_anchors <- FindIntegrationAnchors(object.list = male_STERO, \n",
    "                                        normalization.method = \"SCT\", \n",
    "                                        anchor.features = integ_features,reduction=\"cca\")\n",
    "\n",
    "# Integrate across conditions\n",
    "male_STERO <- IntegrateData(anchorset = integ_anchors,normalization.method = \"SCT\")                 # replace seurat list by object\n",
    "\n",
    "# CLustering \n",
    "DefaultAssay(male_STERO)='integrated'\n",
    "male_STERO=RunPCA(male_STERO, verbose = FALSE,assay=\"integrated\",reduction.name='PCA_INT',npcs = 50)\n",
    "ElbowPlot(male_STERO,reduction ='PCA_INT',ndims = 50 ) # choose 15\n",
    "male_STERO=FindNeighbors(male_STERO,assay ='integrated',reduction ='PCA_INT', dims = 1:20)\n",
    "male_STERO=FindClusters(male_STERO, resolution = 0.3,algorithm = 1,graph.name = 'integrated_nn')\n",
    "male_STERO=RunUMAP(male_STERO,dims = 1:20, reduction.name = 'UMAP_PCA_RNA_INT',reduction = 'PCA_INT')\n",
    "DimPlot(male_STERO)\n",
    "#####\n",
    "###########\n",
    "###############\n",
    "####################\n",
    "                                    ## ATAC processing  AND PEAKS  generating ## \n",
    "\n",
    "DefaultAssay(male_STERO) <- \"peaks\"\n",
    "male_STERO <- FindTopFeatures(male_STERO, min.cutoff = 5)\n",
    "male_STERO <- RunTFIDF(male_STERO)\n",
    "male_STERO <- RunSVD(male_STERO)\n",
    "DepthCor(male_STERO,n = 50)\n",
    "\n",
    "male_STERO <- RunUMAP(object = male_STERO,reduction = 'lsi',dims = 2:20,reduction.name = 'UMAP_LSI_ATAC')\n",
    "male_STERO <- FindNeighbors(object = male_STERO,reduction = 'lsi',dims = 2:20) # use SNN for findclusters \n",
    "#only_STERO <- FindClusters(object = only_STERO,algorithm = 1,resolution = 0.7,verbose = FALSE)\n",
    "DimPlot(object = male_STERO, label = TRUE,group.by = 'Replicat',reduction = 'UMAP_LSI_ATAC')+DimPlot(male_STERO,reduction = 'UMAP_PCA_RNA_INT',group.by = 'Replicat')\n",
    "\n",
    "##################################INTEGRATION ON REPLICAT  for atac \n",
    "\n",
    "crazy=male_STERO\n",
    "\n",
    "crazy <- SplitObject(crazy, split.by = \"Replicat\") # list \n",
    "\n",
    "\n",
    "x=SelectIntegrationFeatures(\n",
    "  crazy,\n",
    "  nfeatures = 3000,\n",
    "  assay = c('peaks','peaks'),\n",
    "  verbose = TRUE\n",
    ")\n",
    "\n",
    "\n",
    "# find integration anchors\n",
    "integration.anchors <- FindIntegrationAnchors(\n",
    "  object.list = crazy,\n",
    "  anchor.features =x,\n",
    "  reduction = \"rlsi\",\n",
    "  dims = 2:30\n",
    ")\n",
    "\n",
    "rm(crazy)\n",
    "X1=male_STERO\n",
    "\n",
    "# integrate LSI embeddings\n",
    "integrated <- IntegrateEmbeddings(\n",
    "  anchorset = integration.anchors,\n",
    "  reductions = X1[[\"lsi\"]],\n",
    "  new.reduction.name = \"integrated_lsi\",\n",
    "  dims.to.integrate = 2:30\n",
    ")\n",
    "\n",
    "integrated <- RunUMAP(integrated, reduction = \"integrated_lsi\", dims = 2:20)\n",
    "\n",
    "DimPlot(integrated ,group.by = 'Replicat')\n",
    " \n",
    " \n",
    " #######################ADD INT LSI ON OBJECT \n",
    "\n",
    "male_STERO@reductions$UMAP_LSI_ATAC_INT=integrated@reductions[[\"umap\"]]\n",
    "male_STERO@reductions$integrated_lsi=integrated@reductions[[\"integrated_lsi\"]]\n",
    "\n",
    "\n",
    "\n",
    "DimPlot(object = male_STERO, label = TRUE,group.by = 'Replicat',reduction = 'UMAP_LSI_ATAC')+DimPlot(male_STERO,reduction = 'UMAP_PCA_RNA_INT',group.by = 'Replicat')+DimPlot(object = male_STERO, label = TRUE,group.by = 'Replicat',reduction = 'UMAP_LSI_ATAC_INT')\n",
    "\n",
    "\n",
    "\n",
    "male_STERO <- FindMultiModalNeighbors(\n",
    "  male_STERO, reduction.list = list(\"PCA_INT\", \"integrated_lsi\"), \n",
    "  dims.list = list(1:20, 2:30),weighted.nn.name = 'X1'\n",
    ")\n",
    "male_STERO <- RunUMAP(male_STERO, nn.name = \"X1\", reduction.name = \"wnn.umap\", reduction.key = \"wnnUMAP_\")\n",
    "male_STERO=FindClusters(male_STERO,graph.name = 'wknn',resolution = 0.5)\n",
    "\n",
    "DimPlot(male_STERO,reduction = 'wnn.umap')\n",
    "male_STERO$new_ident=Idents(male_STERO)\n",
    "\n",
    "\n",
    "###  MALE TRAJECTORY ####\n",
    "\n",
    "male_glomerulosa_root <- male_STERO[, male_STERO$new_ident %in% c(\"3\")]\n",
    "vector_root_M=colnames(male_glomerulosa_root)\n",
    "\n",
    "\n",
    "male_STERO@reductions$UMAP=male_STERO@reductions[[\"UMAP_PCA_RNA_INT\"]]\n",
    "male.cds <- as.cell_data_set(male_STERO,assay = 'SCT_RUN_STERO')\n",
    "male.cds <- cluster_cells(cds = male.cds, reduction_method = \"UMAP\",resolution=30)\n",
    "\n",
    "\n",
    "# find all possible partitions\n",
    "all_partitions <- unique(male.cds@clusters$UMAP$partitions)\n",
    "all_partitions <- all_partitions[all_partitions != \"1\"]\n",
    "# set all partitions to 1\n",
    "male.cds@clusters$UMAP$partitions[male.cds@clusters$UMAP$partitions %in% all_partitions] <- \"1\"\n",
    "\n",
    "male.cds <- learn_graph(male.cds, use_partition = FALSE)\n",
    "\n",
    "male.cds <- order_cells(male.cds, reduction_method = \"UMAP\" ,root_cells =vector_root_M )\n",
    "plot_cells(\n",
    "  cds = male.cds ,\n",
    "  color_cells_by = \"pseudotime\",\n",
    "  show_trajectory_graph = TRUE,cell_size = 2,trajectory_graph_segment_size = 2,graph_label_size = 2,label_cell_groups = F,label_branch_points = T,)\n",
    "\n",
    "## add to only stero\n",
    "\n",
    "only_STERO<- AddMetaData(\n",
    "  object = only_STERO,\n",
    "  metadata = male.cds@principal_graph_aux@listData$UMAP$pseudotime,\n",
    "  col.name = \"RNA_ONLY_male_trajectory\"\n",
    ")\n",
    "\n",
    "FeaturePlot(only_STERO, c(\"RNA_ONLY_male_trajectory\",\"RNA_ONLY_Female_trajectory\"), pt.size = 1,reduction = 'wnn.umap',label = T) & scale_color_viridis_c()\n",
    "\n",
    "\n",
    "# with algo \n",
    "\n",
    "get_earliest_principal_node <- function(cds){\n",
    "  cell_ids <- which(colData(cds)[, \"new_ident\"] %in% c(\"3\"))\n",
    "  \n",
    "  closest_vertex <-\n",
    "  cds@principal_graph_aux[[\"UMAP\"]]$pr_graph_cell_proj_closest_vertex\n",
    "  closest_vertex <- as.matrix(closest_vertex[colnames(cds), ])\n",
    "  root_pr_nodes <-\n",
    "  igraph::V(principal_graph(cds)[[\"UMAP\"]])$name[as.numeric(names\n",
    "  (which.max(table(closest_vertex[cell_ids,]))))]\n",
    "  \n",
    "  root_pr_nodes\n",
    "}\n",
    "\n",
    "cds_male <- order_cells(male.cds, root_pr_nodes=get_earliest_principal_node(male.cds))\n",
    "cds_female <- order_cells(female.cds, root_pr_nodes=get_earliest_principal_node(female.cds))\n",
    "\n",
    "\n",
    "\n",
    "# add  umap pseudotime value in seurat object \n",
    "\n",
    "\n",
    "only_STERO<- AddMetaData(\n",
    "  object = only_STERO,\n",
    "  metadata = cds_female@principal_graph_aux@listData$UMAP$pseudotime,\n",
    "  col.name = \"female_trajectory\"\n",
    ")\n",
    "\n",
    "\n",
    "only_STERO <- AddMetaData(\n",
    "  object = only_STERO,\n",
    "  metadata = cds_male@principal_graph_aux@listData$UMAP$pseudotime,\n",
    "  col.name = \"male_trajectory\"\n",
    ")\n",
    "\n",
    "FeaturePlot(only_STERO, c(\"male_trajectory\", \"female_trajectory\"), pt.size = 1,reduction = 'wnn.umap',label = T) & scale_color_viridis_c()\n",
    "\n",
    "Idents(only_STERO)=only_STERO$new_ident\n",
    "FeaturePlot(only_STERO,features = c('male_trajectory','female_trajectory'),reduction = 'wnn.umap')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "38c4150d",
   "metadata": {
    "lines_to_next_cell": 2,
    "name": "monocle3 trajectory analysis"
   },
   "outputs": [],
   "source": [
    "##########\n",
    "\n",
    "genes=rownames(all.cds)\n",
    "genes=head(genes)\n",
    "\n",
    "\n",
    "\n",
    "pt.matrix <- exprs(all.cds)[match(genes,rownames(rowData(all.cds))),order(pseudotime(all.cds))]\n",
    "#Can also use \"normalized_counts\" instead of \"exprs\" to use various normalization methods, for example:\n",
    "#normalized_counts(cds, norm_method = \"log\")\n",
    "\n",
    "pt.matrix <- t(apply(pt.matrix,1,function(x){smooth.spline(x,df=3)$y}))\n",
    "pt.matrix <- t(apply(pt.matrix,1,function(x){(x-mean(x))/sd(x)}))\n",
    "rownames(pt.matrix) <- genes;\n",
    "\n",
    "\n",
    "hthc <- Heatmap(\n",
    "  pt.matrix,\n",
    "  name                         = \"z-score\",\n",
    "  col                          = colorRamp2(seq(from=-2,to=2,length=11),rev(brewer.pal(11, \"Spectral\"))),\n",
    "  show_row_names               = TRUE,\n",
    "  show_column_names            = T,\n",
    "  row_names_gp                 = gpar(fontsize = 6),\n",
    "  clustering_method_rows = \"ward.D2\",\n",
    "  clustering_method_columns = \"ward.D2\",\n",
    "  row_title_rot                = 0,\n",
    "  cluster_rows                 = TRUE,\n",
    "  cluster_row_slices           = T,\n",
    "  cluster_columns              = T)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "### take reduction \n",
    "all_STERO=only_STERO\n",
    "\n",
    "all_STERO@reductions$UMAP=all_STERO@reductions[[\"wnn.umap\"]]\n",
    "\n",
    "# root for trajectory \n",
    "all_root <- all_STERO[, all_STERO$new_ident %in% c( 'Capsular_Cells',\"Female_Glomerulosa\",\"Male_Glomerulosa\")]\n",
    "vector_root_all=colnames(all_root)\n",
    "\n",
    "\n",
    "\n",
    "# Building trajectories with Monocle 3\n",
    "\n",
    "\n",
    "\n",
    "all.cds <- as.cell_data_set(all_STERO)\n",
    "all.cds <- cluster_cells(cds = all.cds, reduction_method = \"UMAP\",resolution = 1,max.overlaprs=10)\n",
    "\n",
    "\n",
    "\n",
    "# find all possible partitions\n",
    "all_partitions <- unique(all.cds@clusters$UMAP$partitions)\n",
    "all_partitions <- all_partitions[all_partitions != \"1\"]\n",
    "# set all partitions to 1\n",
    "all.cds@clusters$UMAP$partitions[all.cds@clusters$UMAP$partitions %in% all_partitions] <- \"1\"\n",
    "all.cds <- learn_graph(all.cds, use_partition = FALSE)\n",
    "\n",
    "\n",
    "# order cells\n",
    "all.cds <- order_cells(all.cds, reduction_method = \"UMAP\" ,root_cells =vector_root_all )\n",
    "\n",
    "\n",
    "\n",
    "x=plot_cells(\n",
    "  cds = all.cds,\n",
    "  color_cells_by = \"pseudotime\",\n",
    "  show_trajectory_graph = TRUE,cell_size = 2,trajectory_graph_segment_size = 2,graph_label_size = 2,label_cell_groups = F,label_branch_points = T,)+\n",
    "  plot_cells(\n",
    "  cds = all.cds,\n",
    "  color_cells_by = \"pseudotime\",\n",
    "  show_trajectory_graph = TRUE,cell_size = 2,trajectory_graph_segment_size = 2,graph_label_size = 2,label_cell_groups = F,label_branch_points = T)\n",
    "\n",
    "\n",
    "\n",
    "# # a helper function to identify the root principal points:\n",
    "get_earliest_principal_node <- function(cds){\n",
    "  cell_ids <- which(colData(cds)[, \"new_ident\"] %in% c('Capsular_Cells',\"Female_Glomerulosa\",\"Male_Glomerulosa\"))\n",
    "  \n",
    "  closest_vertex <-\n",
    "  cds@principal_graph_aux[[\"UMAP\"]]$pr_graph_cell_proj_closest_vertex\n",
    "  closest_vertex <- as.matrix(closest_vertex[colnames(cds), ])\n",
    "  root_pr_nodes <-\n",
    "  igraph::V(principal_graph(cds)[[\"UMAP\"]])$name[as.numeric(names\n",
    "  (which.max(table(closest_vertex[cell_ids,]))))]\n",
    "  \n",
    "  root_pr_nodes\n",
    "}\n",
    "cds_male <- order_cells(male.cds, root_pr_nodes=get_earliest_principal_node(male.cds))\n",
    "cds_female <- order_cells(female.cds, root_pr_nodes=get_earliest_principal_node(female.cds))\n",
    "all.cds <- order_cells(all.cds, root_pr_nodes=get_earliest_principal_node(all.cds))\n",
    "\n",
    "\n",
    "# add  umap pseudotime value in seurat object \n",
    "\n",
    "\n",
    "only_STERO <- AddMetaData(\n",
    "  object = only_STERO,\n",
    "  metadata = all.cds@principal_graph_aux@listData$UMAP$pseudotime,\n",
    "  col.name = \"all_sexe_trajectory_umap\"\n",
    ")\n",
    "\n",
    "\n",
    "multiomics <- AddMetaData(\n",
    "  object = multiomics,\n",
    "  metadata = male.cds@principal_graph_aux@listData$UMAP$pseudotime,\n",
    "  col.name = \"male\"\n",
    ")\n",
    "\n",
    "FeaturePlot(multiomics, c(\"female\", \"male\"), pt.size = 1,reduction = 'wnn.umap',label = T) & scale_color_viridis_c()\n",
    "Idents(only_STERO)=only_STERO$new_ident\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e111371f",
   "metadata": {
    "lines_to_next_cell": 2,
    "name": "DYNIVERSE TRAJECTORY"
   },
   "outputs": [],
   "source": [
    "\n",
    "library(dyno)\n",
    "library(tidyverse)\n",
    "\n",
    "DefaultAssay(only_STERO)='SCT_RUN_STERO'\n",
    "# PREPARING THE DATA\n",
    "\n",
    "object_counts <- Matrix::t(as(as.matrix(only_STERO@assays$SCT_RUN_STERO@counts), 'sparseMatrix'))\n",
    "object_expression <- Matrix::t(as(as.matrix(only_STERO@assays$SCT_RUN_STERO@data), 'sparseMatrix'))\n",
    "object_cellinfo <- only_STERO@meta.data #optional, i needed the cell type IDs\n",
    "object_cellinfo <- only_STERO@meta.data[[\"merging_ident\"]]\n",
    "\n",
    "\n",
    "dyno_object <- wrap_expression(\n",
    "  counts = object_counts,\n",
    "  expression = object_expression\n",
    "  )\n",
    "\n",
    "\n",
    " dyno_object <- add_grouping(\n",
    "   dyno_object,\n",
    "  only_STERO@meta.data$merging_ident\n",
    " )\n",
    "\n",
    "dyno_object <- add_dimred(dyno_object,\n",
    "                          as.matrix(only_STERO@reductions$wnn.umap@cell.embeddings)\n",
    "                          )\n",
    "\n",
    " \n",
    "# guidelines : \n",
    " guidelines_shiny(dataset = dyno_object)\n",
    " \n",
    " guidelines <- guidelines(\n",
    "  dyno_object,\n",
    "  answers = answer_questions(\n",
    "    dyno_object,\n",
    "    multiple_disconnected = FALSE,\n",
    "    expect_topology = TRUE,\n",
    "    expected_topology = \"linear\"\n",
    "  )\n",
    ")\n",
    " \n",
    "# INFERRING TRAJECTORIES\n",
    "\n",
    " \n",
    "model <- infer_trajectory(dyno_object, ti_paga_tree())\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "dyno_object <-add_dimred(dimred =as.matrix(only_STERO@reductions$wnn.umap@cell.embeddings),expression_source = dyno_object$expression)\n",
    "\n",
    "dyno_object <- add_dimred(dyno_object,as.matrix(only_STERO@reductions$wnn.umap@cell.embeddings))\n",
    "\n",
    "\n",
    "\n",
    "plot_dimred(\n",
    "  model, \n",
    "  expression_source = dyno_object$expression, \n",
    "  grouping = object_cellinfo \n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cfc409f5",
   "metadata": {
    "lines_to_next_cell": 2,
    "name": "SLINGSHOT TRAJECTORY"
   },
   "outputs": [],
   "source": [
    "\n",
    "library(slingshot)\n",
    "library(BUSpaRse)\n",
    "library(tidyverse)\n",
    "library(tidymodels)\n",
    "library(Seurat)\n",
    "library(scales)\n",
    "library(viridis)\n",
    "library(Matrix)\n",
    "\n",
    "sce <- SingleCellExperiment(female_STERO)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bc5dbbd8",
   "metadata": {
    "lines_to_next_cell": 2,
    "name": "SCANPY Convert to anndata"
   },
   "outputs": [],
   "source": [
    "library(SeuratDisk)\n",
    "setwd('D:/Stage')\n",
    "\n",
    "female_STERO=readRDS('D:/Stage/female_stero')\n",
    "female_STERO@assays[[\"peaks\"]]@motifs =NULL # error if features in motif slot \n",
    "i <- sapply(female_STERO@meta.data, is.factor)\n",
    "female_STERO@meta.data[i] <- lapply(female_STERO@meta.data[i], as.character)\n",
    "SaveH5Seurat(female_STERO, filename = \"female_STERO\",overwrite = T)\n",
    "Convert(\"female_STERO.h5seurat\", dest = \"h5ad\",overwrite =T  ,assay= c('integrated'))\n",
    "\n",
    "male_STERO=readRDS('D:/Stage/male_STERO')\n",
    "male_STERO@assays[[\"peaks\"]]@motifs =NULL # error if features in motif slot \n",
    "i <- sapply(male_STERO@meta.data, is.factor)\n",
    "male_STERO@meta.data[i] <- lapply(male_STERO@meta.data[i], as.character)\n",
    "SaveH5Seurat(male_STERO, filename = \"male_STERO\",overwrite = T)\n",
    "Convert(\"male_STERO.h5seurat\", dest = \"h5ad\",overwrite =T  ,assay= c('integrated'))\n",
    "\n",
    "\n",
    " # 1) Convert seurat object to anndata \n",
    "      # object seurat should not have a motif slot , otherwise ==> bug \n",
    "only_STERO_nomotif=only_STERO\n",
    "only_STERO_nomotif@assays[[\"peaks\"]]@motifs =NULL # error if features in motif slot\n",
    "#metadata need to be as.character \n",
    "i <- sapply(only_STERO_nomotif@meta.data, is.factor) \n",
    "only_STERO_nomotif@meta.data[i] <- lapply(only_STERO@meta.data[i], as.character)\n",
    "# To import lsi dim \n",
    "only_STERO_nomotif@reductions[[\"integrated_lsi\"]]@assay.used ='integrated'\n",
    "# Save file \n",
    "SaveH5Seurat(only_STERO_nomotif, filename = \"only_STERO_nomotif_convert.h5Seurat\",overwrite = T)\n",
    "# Convert file \n",
    "# just take the assay you want \n",
    "Convert(\"only_STERO_nomotif_convert.h5Seurat\", dest = \"h5ad\",overwrite =T  ,assay= c('integrated'))\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ea998f39",
   "metadata": {
    "lines_to_next_cell": 2,
    "name": "run scanpy on python"
   },
   "outputs": [],
   "source": [
    "%%python\n",
    "import scanpy  # pip install in  R terminal\n",
    "import scanpy as sc\n",
    "import numpy as np\n",
    "print(np.__version__)\n",
    "import pandas as pd\n",
    "from matplotlib import rcParams\n",
    "from pyWNN import pyWNN\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import networkx as nwx\n",
    "import os\n",
    "\n",
    "\n",
    "sc.pl.paga_compare(\n",
    "  anndata,\n",
    "  basis = \"X_draw_graph_fr\",\n",
    "  threshold = 0.15,\n",
    "  edge_width_scale = 0.5,\n",
    ")\n",
    "\n",
    "################\n",
    "anndata=sc.read_h5ad('D:/Stage/TRUTH/multiomics_convert.h5ad')\n",
    "del anndata.uns['neighbors']\n",
    "del anndata.obsp['distances']\n",
    "del anndata.varm['ICA_STERO']\n",
    "del anndata.obsm['X_ICA_STERO']\n",
    "del anndata.obsm['X_UMAP_LSI_ATAC']\n",
    "del anndata.obsm['X_UMAP_PCA_RNA_INT']\n",
    "del anndata.obsm['X_UMAP_LSI_ATAC_INT']\n",
    "#del anndata.obsm['X_PCA_INT']\n",
    "del anndata.varm['PCA_INT']\n",
    "\n",
    "\n",
    "\n",
    "sc.pp.neighbors(anndata, n_neighbors=50, use_rep='X_PCA_INT')\n",
    " # sc.tl.leiden(anndata, resolution = 1)\n",
    "\n",
    "\n",
    "\n",
    "sc.tl.paga(anndata,groups='leiden_merging')\n",
    "sc.pl.paga(anndata)\n",
    "sc.tl.draw_graph(anndata)\n",
    "\n",
    "anndata.uns['iroot'] = np.flatnonzero(anndata.obs['leiden_merging']  =='Capsular_cells')[0]\n",
    "sc.tl.draw_graph(anndata, \n",
    "                init_pos='paga')# root =np.flatnonzero(anndata.obs['leiden_merging']=='Capsular_cells')[0])\n",
    "\n",
    "\n",
    "sc.tl.dpt(anndata)\n",
    "sc.pl.draw_graph(anndata, color=['merging_ident', 'dpt_pseudotime','leiden','identity'], legend_loc='on data')\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# plot \n",
    "\n",
    "myColors = ['#e6194b', '#3cb44b', '#ffe119', '#4363d8', '#f58231',\n",
    "            '#911eb4', '#46f0f0', '#f032e6', '#bcf60c', '#fabebe',\n",
    "            '#008080', '#e6beff', '#9a6324', '#000000', '#800000',\n",
    "            '#aaffc3', '#808000', '#ffd8b1', '#000075', '#808080', \n",
    "            '#ffffff', '#fffac8']\n",
    "\n",
    "f, axs = plt.subplots(1,2,figsize=(36,12))\n",
    "sns.set(font_scale=1.5)\n",
    "sns.set_style(\"white\")\n",
    "\n",
    "sc.pl.draw_graph(anndata, \n",
    "                 color           = 'merging_ident', \n",
    "                 palette         = myColors,\n",
    "                 legend_loc      = 'on data', \n",
    "                 show            = True, \n",
    "                 size            = 100, \n",
    "                 legend_fontsize = 'large')\n",
    "                 \n",
    "                 \n",
    "sc.pl.draw_graph(anndata, \n",
    "                 color           = 'dpt_pseudotime', \n",
    "                 legend_loc      = 'on data', \n",
    "                 show            = True, \n",
    "                 size            = 400, \n",
    "                 color_map       = \"viridis\",\n",
    "                 legend_fontsize = 'large')\n",
    "\n",
    "                 \n",
    "sns.despine(offset=10, trim=False)\n",
    "plt.tight_layout()\n",
    "plt.show(block=False)\n",
    "plt.close(\"all\")\n",
    "\n",
    "\n",
    "\n",
    "# save value\n",
    "\n",
    "dpt_pseudotime = pd.DataFrame(index   = anndata.obs.index.tolist(),\n",
    "                              columns = [\"Pseudotime\"],\n",
    "                              data    = anndata.obs[\"dpt_pseudotime\"].tolist())\n",
    "\n",
    "dpt_pseudotime.to_csv(\"D:/Stage/Pseudotime_RNA.csv\")\n",
    "\n",
    "\n",
    "\n",
    "FA2_comp = pd.DataFrame(index   = anndata.obs.index.tolist(),\n",
    "                        columns = [\"FA1\", \"FA2\"],\n",
    "                        data    = anndata.obs[\"X_draw_graph_fa\"])\n",
    "\n",
    "FA2_comp.to_csv(\"D:/Stage/Pseudotime_RNA.csvFA2_components_ATAC.csv\")\n",
    "\n",
    "\n",
    "# save paga embending \n",
    "clust_embedding = pd.DataFrame(anndata.uns['paga']['pos'], columns=['X', 'Y'])\n",
    "clust_embedding['Cluster'] = range(clust_embedding.shape[0])\n",
    "clust_embedding = clust_embedding[['Cluster', 'X', 'Y']]\n",
    "\n",
    "clust_embedding.to_csv('D:/Stage/TRUTH/PAGA/cluster_embedding.csv'),\n",
    "                       index=False)\n",
    "\n",
    "# save object \n",
    "sc.read_h5ad(anndata)\n",
    "\n",
    "\n",
    "\n",
    "names=None,\n",
    "sparse=True\n",
    "con=clust_con\n",
    "\n",
    "\n",
    "def con2edges(con, names=None, sparse=True):\n",
    "    print('Converting connectivity matrix to edges...')\n",
    "    n = con.shape[0]\n",
    "    edges = pd.DataFrame(columns=['From', 'To', 'Connectivity'])\n",
    "\n",
    "    for i in range(n):\n",
    "        for j in range(i + 1, n):\n",
    "            if names is not None:\n",
    "                fr = names[i]\n",
    "                to = names[j]\n",
    "            else:\n",
    "                fr = str(i)\n",
    "                to = str(j)\n",
    "\n",
    "            connectivity = con[i, j]\n",
    "            if sparse and connectivity == 0:\n",
    "                continue\n",
    "\n",
    "            entry = {'From' : fr, 'To' : to,\n",
    "                     'Connectivity' : con[i, j]}\n",
    "            edges = edges.append(entry, ignore_index=True)\n",
    "\n",
    "    return edges\n",
    "\n",
    "\n",
    "clust_con = anndata.uns['paga']['connectivities'].toarray()\n",
    "clust_edges = con2edges(clust_con)\n",
    "clust_edges.to_csv(os.path.join(out_dir, 'cluster_edges.csv'),\n",
    "                   index=False)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "print('outputting cluster edges...')\n",
    "clust_con = anndata.uns['paga']['connectivities'].toarray()\n",
    "clust_edges = con2edges(clust_con)\n",
    "clust_edges.to_csv(os.path.join(out_dir, 'cluster_edges.csv'),\n",
    "                   index=False)\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "732b38db",
   "metadata": {
    "lines_to_next_cell": 2,
    "name": "graph PAGA"
   },
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "\n",
    "\n",
    "clust_embedding <- read_csv(\n",
    "    here::here('D:/Stage/TRUTH/PAGA/cluster_embedding.csv'),\n",
    "    col_types = cols(\n",
    "        .default = col_double()))\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "clust_plot <- ggplot(clust_embedding, aes(x = X, y = Y)) +\n",
    "    geom_segment(data = clust_edges,\n",
    "                 aes(x = FromX, y = FromY, xend = ToX, yend = ToY,\n",
    "                     colour = Connectivity, alpha = Connectivity), size = 4) +\n",
    "    scale_colour_viridis_c(limits = c(0, 1)) +\n",
    "    scale_alpha_continuous(limits = c(0, 1), range = c(0, 1), guide = FALSE) +\n",
    "    geom_point(aes(fill = Cluster, size = Size), shape = 21) +\n",
    "    geom_text(aes(label = Cluster)) +\n",
    "    scale_size(range = c(6, 20), guide = FALSE) +\n",
    "    scale_fill_discrete(guide = FALSE) +\n",
    "    guides(colour = guide_colourbar(barwidth = 20)) +\n",
    "    ggtitle(\"PAGA cluster graph\") +\n",
    "    theme_void() +\n",
    "    theme(plot.title = element_text(size = rel(1.2), hjust = 0.1, \n",
    "                                    vjust = 1, margin = margin(5)),\n",
    "          legend.position = \"bottom\")\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e8a57b2a",
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "Create new object without sexual chromosome genes and normalize/ clustering / plot "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "34bc8e1e",
   "metadata": {
    "lines_to_next_cell": 0,
    "name": "remove sexual chromosome genes and re clustering"
   },
   "outputs": [],
   "source": [
    "library(biomaRt)\n",
    "DefaultAssay(only_STERO)='SCT'\n",
    "mart <- useMart(\"ENSEMBL_MART_ENSEMBL\", dataset = \"mmusculus_gene_ensembl\")\n",
    "\n",
    "genes.table <- try(biomaRt::getBM(attributes = c(\"ensembl_gene_id\", \"external_gene_name\",\n",
    "\"description\", \"gene_biotype\", \"chromosome_name\", \"start_position\"), mart = mart, useCache = F))   # fetch chromosome info plus some other annotations\n",
    "\n",
    "chrY.gene = genes.table$external_gene_name[genes.table$chromosome_name == \"Y\"] \n",
    "chrX.gene = genes.table$external_gene_name[genes.table$chromosome_name == \"X\"]\n",
    "chrX_and_Y=append(chrY.gene,chrX.gene)\n",
    "\n",
    "counts <- GetAssayData(multiomics, assay = \"RNA\")\n",
    "counts <- counts[-(which(rownames(counts) %in% chrX_and_Y)),]\n",
    "multiomics_sex_removed <- subset(multiomics, features = rownames(counts)) \n",
    "\n",
    "\n",
    "# re compute a UMAP\n",
    "\n",
    "# re normalize , find variable feature and scaling\n",
    "multiomics_sex_removed=SCTransform(multiomics_sex_removed,assay = 'RNA',variable.features.rv.th = 1.1,variable.features.n = NULL,new.assay.name = 'SCT_nosex',return.only.var.genes = FALSE)\n",
    "DefaultAssay(multiomics_sex_removed)='SCT_nosex'\n",
    "\n",
    "# reduce dimension , umap and clustering\n",
    "multiomics_sex_removed=RunPCA(multiomics_sex_removed, verbose = FALSE,assay=\"SCT_nosex\",reduction.name='PCA_nosex')\n",
    "multiomics_sex_removed=FindNeighbors(multiomics_sex_removed,assay ='SCT_nosex',reduction ='PCA_nosex', dims = 1:50)\n",
    "   # multiomics_sex_removed=FindClusters(multiomics_sex_removed, resolution = 0.5,algorithm = 1)\n",
    "multiomics_sex_removed=RunUMAP(multiomics_sex_removed,dims = 1:50, reduction.name = 'UMAP_PCA_RNA_NOSEX')\n",
    "\n",
    "\n",
    "DimPlot(multiomics_sex_removed,reduction = 'UMAP_PCA_RNA_NOSEX',label = T) | DimPlot(object = multiomics_sex_removed, label = TRUE,group.by = 'identity',reduction = 'UMAP_PCA_RNA_NOSEX') |DimPlot(object = multiomics_sex_removed, label = TRUE,group.by = 'Replicat',reduction = 'UMAP_PCA_RNA_NOSEX')\n",
    "\n",
    "# glomerulosa don't split between sexe . fasiculata split with sexe , and by replicat\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f6838e05",
   "metadata": {},
   "source": [
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "945da4be",
   "metadata": {
    "lines_to_next_cell": 0,
    "name": "SCENIC"
   },
   "outputs": [],
   "source": [
    "# http://htmlpreview.github.io/?https://github.com/aertslab/SCENIC/blob/master/inst/doc/SCENIC_Setup.html\n",
    "# http://htmlpreview.github.io/?https://github.com/aertslab/SCENIC/blob/master/inst/doc/SCENIC_Running.html\n",
    "\n",
    "BiocManager::install(c(\"AUCell\", \"RcisTarget\"))\n",
    "BiocManager::install(c(\"GENIE3\")) # Optional. Can be replaced by GRNBoost\n",
    "BiocManager::install(c(\"GRNBoost\"))\n",
    "BiocManager::install(c(\"zoo\", \"mixtools\", \"rbokeh\"))\n",
    "BiocManager::install(c(\"DT\", \"NMF\", \"ComplexHeatmap\", \"R2HTML\", \"Rtsne\"))\n",
    "BiocManager::install(c(\"doMC\", \"doRNG\"))\n",
    "devtools::install_github(\"aertslab/SCopeLoomR\", build_vignettes = TRUE)     #To export/visualize in http://scope.aertslab.org\n",
    "devtools::install_github(\"aertslab/SCENIC\") \n",
    "packageVersion(\"SCENIC\")\n",
    "\n",
    "# loading with wget \"\", linux :\n",
    "dbFiles <- c(\"https://resources.aertslab.org/cistarget/databases/homo_sapiens/hg19/refseq_r45/mc9nr/gene_based/hg19-500bp-upstream-7species.mc9nr.feather\",\n",
    "\"https://resources.aertslab.org/cistarget/databases/homo_sapiens/hg19/refseq_r45/mc9nr/gene_based/hg19-tss-centered-10kb-7species.mc9nr.feather\")\n",
    "# mc9nr: Motif collection version 9: 24k motifs\n",
    "dbFiles <- c(\"https://resources.aertslab.org/cistarget/databases/mus_musculus/mm9/refseq_r45/mc9nr/gene_based/mm9-500bp-upstream-7species.mc9nr.feather\",\n",
    "\"https://resources.aertslab.org/cistarget/databases/mus_musculus/mm9/refseq_r45/mc9nr/gene_based/mm9-tss-centered-10kb-7species.mc9nr.feather\")\n",
    "# mc9nr: Motif collection version 9: 24k motifs\n",
    "dbFiles <- c(\"https://resources.aertslab.org/cistarget/databases/drosophila_melanogaster/dm6/flybase_r6.02/mc8nr/gene_based/dm6-5kb-upstream-full-tx-11species.mc8nr.feather\")\n",
    "# mc8nr: Motif collection version 8: 20k motifs\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e1976a13",
   "metadata": {},
   "source": [
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "56f88d67",
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "Enrichment analysis (DE) with Gprofiler , here we just need the gene name as input"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b1ef5447",
   "metadata": {
    "lines_to_next_cell": 0,
    "name": "Enrichment analysis (DE) with Gprofiler"
   },
   "outputs": [],
   "source": [
    "library(gprofiler2)\n",
    "#Analysing multiple gene lists\n",
    "\n",
    "# get the significant up/down-regulated genes\n",
    "X1=FindMarkers(only_STERO,ident.1 = c(\"Male_Fasciculata\",\"Male_Glomerulosa\",\"Male_tZone\"),ident.2 = c(\"Female_tZone\",\"Female_Fasciculata\",\"Female_Glomerulosa\"))\n",
    "X1$gene_names=rownames(X1)\n",
    "female_up=X1[X1$p_val_adj<0.05&X1$avg_log2FC>0,]\n",
    "male_up=X1[X1$p_val_adj<0.05&X1$avg_log2FC<0,]\n",
    "\n",
    "\n",
    "female_up = DE_STERO_Female_ZF\n",
    "male_up = DE_STERO_Male_ZF\n",
    "\n",
    "# order genes by log2FC\n",
    "female_up = female_up[order(female_up$avg_log2FC, decreasing = TRUE),]\n",
    "\n",
    "male_up = male_up[order(male_up$avg_log2FC, decreasing = FALSE),] # higher value first \n",
    "\n",
    "### Background set \n",
    "background=rownames(only_STERO@assays[[\"SCT_RUN_STERO\"]]@data) # take the good assay ! \n",
    "\n",
    "# ordered enrichment analysis\n",
    "\n",
    "gp_up_ordered_F = gost(list(female_up$gene_names,male_up$gene_names), organism = 'mmusculus',\n",
    " ordered_query = TRUE,sources = c('GO:BP', 'GO:MF', 'GO:CC','KEGG', 'REAC', 'TF','WP'),domain_scope=\"custom\",custom_bg = background)\n",
    "\n",
    "# run individually \n",
    "gp_up_ordered_F = gost(female_up$gene_names, organism = 'mmusculus',\n",
    " ordered_query = TRUE,sources = c('GO:BP', 'GO:MF', 'GO:CC','KEGG', 'REAC', 'TF','WP'),domain_scope=\"custom\",custom_bg = background,correction_method = \"bonferroni\",user_threshold = 0.01)\n",
    "\n",
    "gp_up_ordered_M= gost(male_up$gene_names, organism = 'mmusculus',\n",
    " ordered_query = TRUE,sources = c('GO:BP', 'GO:MF', 'GO:CC','KEGG', 'REAC', 'TF','WP'),domain_scope=\"custom\",custom_bg = background,correction_method = \"bonferroni\",user_threshold = 0.01)\n",
    "\n",
    "gp_up_ordered_M\n",
    "gp_up_ordered_F\n",
    "#plot\n",
    "gostplot(gp_up_ordered_F,capped = FALSE,interactive = TRUE) # need to highlight the symbol on the plot\n",
    "gostplot(gp_up_ordered_M,capped = FALSE,interactive = TRUE) # need to highlight the symbol on the plot\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "34f7e5cb",
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "# clusterprofiler"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6ece796e",
   "metadata": {
    "name": "And again ... with clusterprofiler"
   },
   "outputs": [],
   "source": [
    "library(ggplot2)\n",
    "library(enrichplot)\n",
    "library(clusterProfiler)\n",
    "library(org.Mm.eg.db)\n",
    "library(DOSE)\n",
    "#######################################ENSEMBL Gene Set Enrichment Analysis\n",
    "\n",
    "# we want the log2 fold change\n",
    "DE_Female_Fasciculata_log2FC<-DE_STERO_Female_tZ$avg_log2FC\n",
    "# name the vector for each log2fc get the gene names \n",
    "names(DE_Female_Fasciculata_log2FC)<- DE_STERO_Female_tZ$ensembl_gene_id # take ensemble id\n",
    "# remove any NA values \n",
    "DE_Female_Fasciculata_log2FC<-na.omit(DE_Female_Fasciculata_log2FC)\n",
    "DE_Female_Fasciculata_log2FC<-DE_Female_Fasciculata_log2FC[!is.na(names(DE_Female_Fasciculata_log2FC))]\n",
    "# sort the list in decreasing order (required for clusterProfiler)\n",
    "DE_Female_Fasciculata_log2FC = sort(DE_Female_Fasciculata_log2FC, decreasing = TRUE)\n",
    "##################################################### output \n",
    "gse <- gseGO(geneList=DE_Female_Fasciculata_log2FC, \n",
    "             ont =\"MF\", \n",
    "             keyType = \"ENSEMBL\", \n",
    "             minGSSize = 3, \n",
    "             maxGSSize = 800, \n",
    "             pvalueCutoff = 0.05, \n",
    "             verbose = TRUE, \n",
    "             OrgDb = org.Mm.eg.db, \n",
    "             pAdjustMethod = \"none\")\n",
    "\n",
    "# plot\n",
    "dotplot(gse, showCategory=10, split=\".sign\") + facet_grid(.~.sign)\n",
    "\n",
    "\n",
    "#######################################KEGG Gene Set Enrichment Analysis\n",
    "# we want the log2 fold change\n",
    "DE_Female_Fasciculata_log2FC<-`DE_Female Fasciculata`$avg_log2FC\n",
    "# name the vector for each log2fc get the gene names \n",
    "names(DE_Female_Fasciculata_log2FC)<- `DE_Female Fasciculata`$ENTREZID\n",
    "# remove any NA values \n",
    "DE_Female_Fasciculata_log2FC<-na.omit(DE_Female_Fasciculata_log2FC)\n",
    "DE_Female_Fasciculata_log2FC<-DE_Female_Fasciculata_log2FC[!is.na(names(DE_Female_Fasciculata_log2FC))]\n",
    "# sort the list in decreasing order (required for clusterProfiler)\n",
    "DE_Female_Fasciculata_log2FC = sort(DE_Female_Fasciculata_log2FC, decreasing = TRUE)\n",
    "\n",
    "##########################################Create gseKEGG object\n",
    "KEG <- gseKEGG(geneList     = DE_Female_Fasciculata_log2FC,\n",
    "               organism     = 'mmu',\n",
    "               minGSSize    = 3,\n",
    "               maxGSSize    = 800,\n",
    "               pvalueCutoff = 0.05,\n",
    "               pAdjustMethod = \"none\",\n",
    "               keyType       = \"ncbi-geneid\",)\n",
    "\n",
    "# plot\n",
    "dotplot(KEG, title = \"Enriched Pathways\" , split=\".sign\") + facet_grid(.~.sign)\n",
    "dotplot(gse, split=\".sign\") + facet_grid(.~.sign)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0961278f",
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "Comparing multiple gene lists"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b82d7265",
   "metadata": {
    "lines_to_next_cell": 2,
    "name": "Comparing multiple gene lists"
   },
   "outputs": [],
   "source": [
    "\n",
    "all_sex_dif=FindMarkers(remove_only_stero,ident.1 =c(\"Female_tZone_female\",\"Female_Fasciculata_2_female\",\"Female_Fasciculata_1_female\",\"Capsular_cells_female\", \"Female_Glomerulosa_female\"),ident.2=c(\"Male_tZone_male\",\"Male_Fasciculata_2_male\",\"Male_Fasciculata_1_male\",\"Capsular_cells_male\", \"Male_Glomerulosa_male\"),assay = 'SCT_RUN_STERO',logfc.threshold = 0.2)\n",
    "all_sex_dif=all_sex_dif[all_sex_dif$p_val_adj<0.05,]\n",
    "all_sex_dif$gene_names=rownames(all_sex_dif)\n",
    "all_sex_dif$val2 <- mouse_TF$V1[match(all_sex_dif$gene_names, mouse_TF$V1)]\n",
    "all_sex_dif$sex <- chrx_y$chrX_and_Y[match(all_sex_dif$gene_names, chrx_y$chrX_and_Y)]\n",
    "\n",
    "ensembl <- useMart(\"ensembl\", dataset=\"mmusculus_gene_ensembl\")\n",
    "gene_IDs <- getBM(filters= 'mgi_symbol', attributes= c(\"ensembl_gene_id\",'mgi_symbol','entrezgene_id','description'),values = all_sex_dif$gene_names, mart= ensembl)\n",
    "# Add into DE DATAFRAME\n",
    "all_sex_dif=left_join(all_sex_dif,gene_IDs, by = c(\"gene_names\"=\"mgi_symbol\"))\n",
    "# check duplicate value \n",
    "all_sex_dif$gene[duplicated(all_sex_dif$gene_names)]\n",
    "# remove duplicate value\n",
    "all_sex_dif=all_sex_dif[!duplicated(all_sex_dif$gene_names), ]\n",
    "write.table(all_sex_dif$gene_names,file = 'all_sex_dif.txt',quote = F,row.names = F,col.names = F)\n",
    "\n",
    "\n",
    "\n",
    "mydf <- data.frame(Entrez=all_sex_dif$entrezgene_id,all_sex_dif)\n",
    "mydf <- mydf[abs(mydf$avg_log2FC) > 0,]\n",
    "mydf$group <- \"Female upregulated\"\n",
    "mydf$group[mydf$avg_log2FC < 0] <- \"Male upregulated\"\n",
    "mydf$othergroup <- \"no significant\"\n",
    "mydf$othergroup[(mydf$avg_log2FC > 0.7)] <- \"Female highly upregulated\"\n",
    "mydf$othergroup[(mydf$avg_log2FC < -0.7)] <- \"Male highly upregulated\"\n",
    "\n",
    "\n",
    "\n",
    "x=mydf$gene_names[mydf$avg_log2FC>0&mydf$p_val_adj<0.05]\n",
    "write.table(x = x,file = 'all_female.txt',quote = F,row.names = F,col.names = F)\n",
    "getwd()\n",
    "\n",
    "formula_res_GO <- compareCluster(Entrez~group, data=mydf, OrgDb=\"org.Mm.eg.db\", fun=\"groupGO\")\n",
    "formula_res_KEGG <- compareCluster(Entrez~group, data=mydf, fun=\"enrichKEGG\",org='mmu')\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "top_markers=all_sex_dif\n",
    "\n",
    "\n",
    "\n",
    "top_markers$sign <- 0\n",
    "top_markers$sign_logFC[top_markers$avg_log2FC > 0] <- 1\n",
    "top_markers$description <- gsub(' \\\\[Source.*$', '', top_markers$description)\n",
    "\n",
    "ggplot(top_markers[is.na(top_markers$description) == F, ], aes(reorder(gene_id, avg_logFC), avg_logFC))+\n",
    "  geom_bar(stat='identity', aes(fill = factor(sign_logFC)))+\n",
    "  coord_flip()+\n",
    "  scale_x_discrete(labels = top_markers$description[is.na(top_markers$description) == F][order(top_markers$avg_logFC[is.na(top_markers$description) == F])])+\n",
    "  labs(x = 'gene description', y = 'Average logFC')+\n",
    "  guides(fill = F)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "head(formula_res_KEGG)\n",
    "dotplot(formula_res_GO,title = \" Fasciculata differentially enriched pathways\", x=\"group\",size='count',showCategory =NULL ) + facet_grid(~othergroup)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "24507eb2",
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "# Volcano plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "161f1b28",
   "metadata": {
    "lines_to_next_cell": 2,
    "name": "Volcano plot"
   },
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "all_sex_dif=FindMarkers(remove_only_stero,ident.1 =c(\"Female_tZone_female\",\"Female_Fasciculata_female\",\"Capsular_cells_female\", \"Female_Glomerulosa_female\"),ident.2=c(\"Male_tZone_male\",\"Male_Fasciculata_male\",\"Capsular_cells_male\", \"Male_Glomerulosa_male\"),assay = 'SCT_RUN_STERO',logfc.threshold = 0)\n",
    "all_sex_dif=all_sex_dif[all_sex_dif$p_val_adj<0.05,]\n",
    "all_sex_dif$gene_names=rownames(all_sex_dif)\n",
    "all_sex_dif$val2 <- mouse_TF$V1[match(all_sex_dif$gene_names, mouse_TF$V1)]\n",
    "all_sex_dif$sex <- chrx_y$chrX_and_Y[match(all_sex_dif$gene_names, chrx_y$chrX_and_Y)]\n",
    "\n",
    "  write.table(all_sex_dif$gene_names[all_sex_dif$avg_log2FC< - 0.5 & all_sex_dif$p_val_adj<0.05],file = 'male_all_cells.txt',quote = F,row.names = F,col.names = F)\n",
    "  write.table(all_sex_dif$gene_names[all_sex_dif$avg_log2FC> 0.5 & all_sex_dif$p_val_adj<0.05],file = 'female_all_cells.txt',quote = F,row.names = F,col.names = F)\n",
    "\n",
    "  \n",
    "sexy=all_sex_dif\n",
    "sexy<-sexy[!is.na(sexy$sex),]\n",
    "\n",
    "\n",
    "sexF=all_sex_dif$sex[all_sex_dif$p_val<0.05& all_sex_dif$avg_log2FC >  0.1]\n",
    "sexF<-sexF[!is.na(sexF)]\n",
    "\n",
    "\n",
    "\n",
    "  # create custom key-value pairs for 'high', 'low', 'mid' expression by fold-change\n",
    "  # this can be achieved with nested ifelse statements\n",
    "  keyvals <- ifelse(all_sex_dif$sex %in% sexy$sex==TRUE, 'red',\n",
    "              ifelse(all_sex_dif$avg_log2FC > 0 & all_sex_dif$p_val_adj < 0.05, '#D16103',\n",
    "              ifelse(all_sex_dif$avg_log2FC < 0 & all_sex_dif$p_val_adj < 0.05, '#4E84C4','black')))\n",
    "\n",
    "  keyvals[is.na(keyvals)] <- 'black'\n",
    "  names(keyvals)[keyvals == '#D16103'] <- 'Female'\n",
    "  names(keyvals)[keyvals == 'black'] <- 'Not significant '\n",
    "  names(keyvals)[keyvals == '#4E84C4'] <- 'Male'\n",
    "  names(keyvals)[keyvals == 'red'] <- 'sexual chromosome'\n",
    "\n",
    "  \n",
    "  \n",
    "vector1=all_sex_dif$gene_names[all_sex_dif$p_val_adj< 0.005&all_sex_dif$avg_log2FC>1.1]\n",
    "vector2=all_sex_dif$gene_names[all_sex_dif$p_val_adj< 0.005&all_sex_dif$avg_log2FC< -1.1]\n",
    "vector2=append(vector1,vector2)\n",
    "vector2=append(x = vector2,values = 'Smarca1')\n",
    "  \n",
    "  lab_italics <- paste0(\"italic('\", rownames(all_sex_dif), \"')\")\n",
    "  selectLab_italics = paste0(\n",
    "    \"italic('\",\n",
    "    vector2, \"')\")\n",
    "\n",
    "\n",
    " EnhancedVolcano(all_sex_dif,\n",
    "    lab = lab_italics,\n",
    "    x = 'avg_log2FC',\n",
    "    y = 'p_val_adj',\n",
    "  pCutoff = 0.05,\n",
    "    FCcutoff = 0.9,\n",
    "    cutoffLineType = 'twodash',\n",
    "    cutoffLineWidth = 0.5,\n",
    "    pointSize = 4.0,\n",
    "    labSize = 6.0,\n",
    "    colAlpha = 1,\n",
    "    legendLabels=c('Not sig.','Log (base 2) FC','p-value',\n",
    "      'p-value & Log (base 2) FC'),\n",
    "    legendPosition = 'right',\n",
    "    legendLabSize = 16,\n",
    "    legendIconSize = 5.0,\n",
    "  colCustom = keyvals,  \n",
    "\n",
    "    parseLabels = TRUE, title = 'STERODOGENIC GENE EXPRESSION ')+ coord_flip()\n",
    " \n",
    "\n",
    "\n",
    ")\n",
    " \n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5f450f71",
   "metadata": {},
   "source": [
    "# Remove sexual chromosome peaks and do a clustering"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "83016ff4",
   "metadata": {
    "lines_to_next_cell": 2,
    "name": "Remove sexual chromosome peaks and do a clustering"
   },
   "outputs": [],
   "source": [
    "DefaultAssay(multiomics)='peaks'\n",
    "counts_peaks=as.data.frame(rownames(multiomics)) # first i'm taking all the peaks\n",
    "names(counts_peaks)[1] <- \"newName\"# rename columns \n",
    "\n",
    "\n",
    "counts_peak_grep=counts_peaks$newName[!grepl(\"chrY.*\\\\d\",counts_peaks$newName)] # get rid of all the peaks beginning with Y\n",
    "counts_peak_grep=counts_peaks$newName[!grepl(\"chrX.*\\\\d\",counts_peaks$newName)] # get rid of all the peaks beginning with X\n",
    "length(counts_peak_grep)\n",
    "\n",
    "# Subset this peaks with PEAKS ASSAY of seurat object \n",
    "\n",
    "toRemove <- (counts_peak_grep) \n",
    "no_Y <- multiomics[rownames(multiomics) %in% toRemove,] # subset from seurat object \n",
    "no_Y=RenameAssays(no_Y, peaks='peaks_nosex') # rename assay\n",
    "\n",
    "\n",
    "# re compute a UMAP \n",
    "\n",
    "# re run find variable features and normalisation\n",
    "no_Y <- FindTopFeatures(no_Y, min.cutoff = 'q5',assay = 'peaks_nosex')\n",
    "no_Y <- RunTFIDF(no_Y)\n",
    "no_Y <- RunSVD(no_Y,reduction.name = 'lsi_nosex')\n",
    "DepthCor(no_Y) # remove first component ( classic)\n",
    "\n",
    "# UMAP AND CLUSTERING\n",
    "no_Y <- RunUMAP(object = no_Y,reduction = 'lsi_nosex',dims = 2:50)\n",
    "no_Y <- FindNeighbors(object = no_Y,reduction = 'lsi_nosex',dims = 2:50,assay ='peaks_nosex') # use SNN for findclusters \n",
    "no_Y <- FindClusters(object = no_Y,algorithm = 1,resolution = 0.7,verbose = FALSE,graph.name = 'peaks_nosex_snn')\n",
    "\n",
    "DimPlot(object = no_Y, label = TRUE) | DimPlot(object = no_Y, label = TRUE,group.by = 'identity') |DimPlot(object = no_Y, label = TRUE,group.by = 'Replicat')\n",
    "\n",
    "VlnPlot(multiomics, features = 'nFeature_peaks')\n",
    "# same as RNA, the glomerulosa don't split , sexe chromsome must explain a dimorphism in expression , try di"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "188536be",
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "#  HOCOMOCCO PWM MATRIX"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4f85dd9a",
   "metadata": {
    "lines_to_next_cell": 2,
    "name": "HOCOMOCCO PWM MATRIX"
   },
   "outputs": [],
   "source": [
    "\n",
    "\n",
    " library(universalmotif)\n",
    "  library(TFBSTools)\n",
    "  library(rlist)\n",
    "  # data taken from https://hocomoco11.autosome.ru/final_bundle/hocomoco11/full/HUMAN/mono/HOCOMOCOv11_full_pwm_HUMAN_mono.tar.gz\n",
    "  pathToPWMs = \"D:/ADRENAL_PROJECT/hocomoco_motif/HOCOMOCOv11_full_pwm_MOUSE_mono/pwm/\"\n",
    "  pwmfiles = list.files(pathToPWMs)\n",
    "\n",
    "  empty.list <- list()\n",
    "  \n",
    "    for (x in 1:length(pwmfiles)) {\n",
    "  \n",
    "    test1 = read_matrix(paste0(pathToPWMs,pwmfiles[x]), \n",
    "                      headers = \">\", sep = \"\", positions = \"rows\")\n",
    "    y  =convert_motifs(test1, class = \"TFBSTools-PWMatrix\")\n",
    "  \n",
    "    empty.list <- c(empty.list, y)\n",
    "  \n",
    "  }\n",
    "\n",
    "  pfm.list <- do.call(PWMatrixList, empty.list)\n",
    "  \n",
    "for (i in 1:length(pfm.list)){\n",
    "  print(i)\n",
    "  names(pfm.list)[i]=pfm.list@listData[[i]]@name\n",
    "}\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7c1a4f21",
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "#  Convert genome position to dna string and get motif position"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "211b9b2d",
   "metadata": {
    "lines_to_next_cell": 2,
    "name": "Convert genome position to dna string and get motif position"
   },
   "outputs": [],
   "source": [
    "library(TFBSTools)\n",
    "library(tidyr)\n",
    "library(dplyr)\n",
    "library(motifmatchr)\n",
    "library(seqinr)\n",
    "library(stringr)\n",
    "library(JASPAR2020)\n",
    "\n",
    "#0) Create motif matrix \n",
    "  # prepare motif from jaspar \n",
    "  pfm <- getMatrixSet(\n",
    "    x = JASPAR2020,\n",
    "    opts = list(collection = \"CORE\", tax_group = 'vertebrates', all_versions = FALSE)\n",
    "  )\n",
    "  # Create motif matrix with my peaks \n",
    "  motif.matrix <- as.data.frame( CreateMotifMatrix(\n",
    "    features =StringToGRanges( DA_STERO_Female_Fasciculata$peaks[DA_STERO_Female_Fasciculata$avg_log2FC< - 0.25 & DA_STERO_Female_Fasciculata$p_val_adj<0.01],sep = c(\"-\", \"-\")),  # GET RIGHT PEAKS\n",
    "    pwm = pfm,\n",
    "    genome = BSgenome.Mmusculus.UCSC.mm10,score = F\n",
    "  ))\n",
    "\n",
    "\n",
    "#1) Take Peaks on AR and create Dataframe\n",
    "HOMER_FASC_F=as.data.frame(rownames(motif.matrix)[motif.matrix$MA0007.3!=0])\n",
    "# TAKE DATAFRAME WITH FIRST ROW AS PEAKS POSITION , convert into dna strings \n",
    "Peakspos_to_DNA <- function(x) {\n",
    "      names(x)[1]='chr'\n",
    "      x=as.data.frame(str_split_fixed(x$chr,\"-\",3))\n",
    "      names(x)[2]='START'\n",
    "      names(x)[3]='END'# separate columns \n",
    "      for (i in 1:length(x$START)){\n",
    "      x$ID[i]=paste(\"ID\", i,sep = '')# add unique ID columns\n",
    "      } \n",
    "      # convert seq pos to dna seq \n",
    "      Start=as.integer(x$START)\n",
    "      End=as.integer(x$END)\n",
    "      Chr=c(x$V1)\n",
    "      for (i in 1:length(Chr)){\n",
    "        x$DNA_STRINGs[i] <- as.character(Biostrings::getSeq(BSgenome.Mmusculus.UCSC.mm10, Chr[i], Start[i],End[i])) \n",
    "      }\n",
    "      return(x)\n",
    "}\n",
    "# Run function\n",
    "HOMER_FASC_F=Peakspos_to_DNA(HOMER_FASC_F)\n",
    "names(HOMER_FASC_F)[1]='chr'\n",
    "\n",
    "      \n",
    "#3) get AR PWM from jaspar\n",
    "pwm <- getMatrixSet(\n",
    "  x = JASPAR2020,\n",
    "  opts = list(collection = \"UNVALIDATED\", tax_group = 'vertebrates', all_versions = TRUE,matrixtype='PWM',species=\"Mus musculus\")\n",
    ")\n",
    "    # TAKE only Ar\n",
    "    Ar= pwm$MA0007.3\n",
    "    \n",
    "    # Get motif positions within peaks for example motifs in peaks \n",
    "    motif_ix <- matchMotifs(pwms = Ar, subject=HOMER_FASC_F$DNA_STRINGs, genome = \"mm10\",out = \"positions\")\n",
    "  \n",
    "    # View motif Ar as a dataframe \n",
    "    motif_ix_dataframe=as.data.frame(motif_ix)\n",
    "    motif_ix_dataframe=motif_ix_dataframe[motif_ix_dataframe$strand=='+',]\n",
    "    \n",
    "    # merge  motif_ix_dataframe and HOMER_FASC\n",
    "    HOMER_FASC_F$Column5=as.integer(rownames(HOMER_FASC_F))\n",
    "    x=left_join(HOMER_FASC_F,motif_ix_dataframe, by = c('Column5'=\"group\"))\n",
    "    \n",
    "    # remove NA \n",
    "    x= x %>% drop_na('score')\n",
    "    \n",
    "    # new columns =  subset DNA seq based on AR neighboors\n",
    "    for (i in 1:length(x$chr)){\n",
    "    print(i)\n",
    "    x$p300[i]=x$end[i]+300\n",
    "    x$m300[i]=x$start[i]-300\n",
    "    x$subsetDNA[i]=substr( x$DNA_STRINGs[i],start = x$m300[i],stop = x$p300[i])\n",
    "    # add new pos\n",
    "    x$new_start[i]=(as.double(x$V2[i])+((x$start[i]))-300)\n",
    "    x$new_end[i]=(as.double(x$V2[i])+((x$end[i]))+300)\n",
    "\n",
    "\n",
    "    }\n",
    "    \n",
    "for (i in 1:length(x$chr)){\n",
    "  print(i)\n",
    "  x$AR_M300.start[i]=(as.double(x$START[i])+x$start[i])-300\n",
    "  x$AR_M300.end[i]=(as.double(x$START[i])+x$start[i])-1\n",
    "  x$AR_P300.start[i]=(as.double(x$END[i])+x$end[i])+1\n",
    "  x$AR_P300.end[i]=(as.double(x$END[i])+x$end[i])+300\n",
    "  Start=as.integer(x$AR_M300.start)\n",
    "  End=as.integer(x$AR_M300.end)\n",
    "  Chr=c(x$chr)\n",
    "  x$DNA_LEFT_TO_AR[i]<- as.character(Biostrings::getSeq(BSgenome.Mmusculus.UCSC.mm10, Chr[i], Start[i],End[i])) \n",
    "  Start_right=as.integer(x$AR_P300.start)\n",
    "  End_right=as.integer(x$AR_P300.end)\n",
    "  x$DNA_right_TO_AR[i]<- as.character(Biostrings::getSeq(BSgenome.Mmusculus.UCSC.mm10, Chr[i], Start_right[i],End_right[i])) \n",
    "\n",
    "\n",
    "}\n",
    "\n",
    "    write.fasta(sequences =as.list(x$DNA_LEFT_TO_AR),file.out = 'FASTA_MALE_FASC_LEFT_AR.fasta',names = x$X1) # Sequence of interest  \n",
    "    write.fasta(sequences =as.list(x$DNA_LEFT_TO_AR),file.out = 'FASTA_MALE_FASC_Right_AR.fasta',names = x$X1) # Sequence of interest  \n",
    "\n",
    "\n",
    "    \n",
    "    \n",
    "    # TAKE only NRC1/NRC2\n",
    "    NR3C1= pwm$MA0113.2\n",
    "    # Get motif positions within peaks for example motifs in peaks \n",
    "    motif_ix_NR3C1 <- as.data.frame(matchMotifs(pwms = NR3C1, subject= x$subsetDNA, genome = \"mm10\",out = \"positions\"))\n",
    "    motif_ix_NR3C1=motif_ix_NR3C1[motif_ix_NR3C1$strand=='+',]\n",
    "    y=left_join(x,motif_ix_NR3C1, by = c('Column5'=\"group\"))\n",
    "    y= y %>% drop_na('score.y')\n",
    "\n",
    "    \n",
    "    # compute average distance of NRC to AR \n",
    "    for (i in 1:length(y$chr)){\n",
    "    AR_loc=y$start.x[i] + (y$width.x[i]/2)\n",
    "    NRC_loc=y$start.y[i] + (y$width.y[i]/2)\n",
    "    y$AR_NRC[i]=(AR_loc-NRC_loc)*sign(AR_loc - NRC_loc)\n",
    "    print(i)\n",
    "      \n",
    "    }\n",
    "    \n",
    "     # TAKE only TATAA\n",
    "    TATAA= pwm$\n",
    "    # Get motif positions within peaks for example motifs in peaks \n",
    "    motif_ix_NR3C1 <- as.data.frame(matchMotifs(pwms = NR3C1, subject= x$subsetDNA, genome = \"mm10\",out = \"positions\"))\n",
    "    motif_ix_NR3C1=motif_ix_NR3C1[motif_ix_NR3C1$strand=='+',]\n",
    "    y=left_join(x,motif_ix_NR3C1, by = c('Column5'=\"group\"))\n",
    "    y= y %>% drop_na('score.y')\n",
    "\n",
    "    \n",
    "    # compute average distance of NRC to AR \n",
    "    for (i in 1:length(y$chr)){\n",
    "    AR_loc=y$start.x[i] + (y$width.x[i]/2)\n",
    "    NRC_loc=y$start.y[i] + (y$width.y[i]/2)\n",
    "    y$AR_NRC[i]=(AR_loc-NRC_loc)*sign(AR_loc - NRC_loc)\n",
    "    print(i)\n",
    "      \n",
    "    }\n",
    "\n",
    "\n",
    "    # take Granges object from only_STERO\n",
    "      motif.matrix <- as.data.frame(\n",
    "      CreateMotifMatrix(\n",
    "        features =StringToGRanges(x$subsetDNA,sep = c(\"-\", \"-\")),  # GET RIGHT PEAKS\n",
    "        pwm = pfm,\n",
    "        genome = BSgenome.Mmusculus.UCSC.mm10,score = F\n",
    "      ))\n",
    "\n",
    "    # write as fasta file Sequence of interest\n",
    "    for (i in 1:length(x$chr)){\n",
    "    x$X1[i]=paste(\"ID\", i,sep = '')  # add unique ID columns\n",
    "   }\n",
    "    write.fasta(sequences =as.list(x$subsetDNA),file.out = 'FASTA_MALE_FASC.fasta',names = x$X1) # Sequence of interest  \n",
    "\n",
    "    \n",
    "    \n",
    "    # write as fasta file Background sequence \n",
    "    all_peaks=as.data.frame(rownames(only_STERO))\n",
    "    for (i in 1:length(all_peaks$V1)){\n",
    "    all_peaks$ID[i]=paste(\"ID\", i,sep = '')  # add unique ID columns\n",
    "   }\n",
    "    write.fasta(sequences =as.list(all_peaks$DNA_STRINGs),file.out = 'all_peaks.fasta',names = all_peaks$ID) # Background sequence \n",
    "\n",
    "    \n",
    "    \n",
    "                                                       f= FindMarkers(only_STERO,\n",
    "                                                        ident.1 =\"Male_Fasciculata\",\n",
    "                                                        ident.2 = \"Female_Fasciculata\",\n",
    "                                                        assay = 'peaks',\n",
    "                                                        only.pos = F, test.use = 'LR',\n",
    "                                                        min.pct = 0,logfc.threshold = 0,\n",
    "                                                        latent.vars = 'nCount_peaks')\n",
    "\n",
    "\n",
    "\n",
    "    \n",
    "    # Get DNAstring \n",
    "DNA_STRING=DNAStringSet(HOMER_FASC_F$DNA_STRINGs, start=NA, end=NA, width=NA, use.names=TRUE)\n",
    "\n",
    "\n",
    "\n",
    "#  1) selection of the peaks enrichment i\n",
    "# 2) select AR \n",
    "# 3) identifie withuing peaks enrichment of toher TF\n",
    "# 4) among those enrich TF their postiion related to AR binding SITE \n",
    "# peaks who have AR and whatevers CO factor ( glucorticoides ?) \n",
    "\n",
    "# 5 ) put\n",
    "\n",
    "\n",
    "# scan AR motif in sequence \n",
    "motifScanHits(DNA_STRING, Ar, minScore = \"80%\", seqOrder = c(1:length(regionsSeq)))\n",
    "\n",
    "write.table(HOMER_FASC_F$DNA_STRINGs,file = 'fasc.txt',append = F,quote = F,col.names = F,row.names = T)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# write as fasta file\n",
    "write.fasta(sequences =as.list(x$subsetDNA),file.out = 'FASTA_MALE_FASC.fasta',names = HOMER_FASC_F$column1)\n",
    "\n",
    "\n",
    "print(motif_ix)\n",
    "library(motifmatchr)\n",
    "library(GenomicRanges)\n",
    "\n",
    "motifMatches(motif_ix) # Extract matches matrix from result\n",
    "\n",
    "  # example peak set we're interested in\n",
    "    peaks.use <- x$Combined\n",
    "    motif.use <- motif.matrix$MA0006.1\n",
    "\n",
    "    for ( i in 0:length(peaks.use)){\n",
    "       sum(motif.matrix[peaks.use, colnames(motif.matrix[i])]) / length(x$Combined)\n",
    "       sum(motif.matrix[peaks.use, colnames(motif.matrix[i])]) / length(x$Combined)\n",
    "      \n",
    "    }\n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bb05ed50",
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "Run HOMER with DA data frame  and create .txt file "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "452ab42e",
   "metadata": {
    "name": "HOMER RUN"
   },
   "outputs": [],
   "source": [
    "library(stringr)\n",
    "#HOMER peak files should have at minimum 5 columns (separated by TABs, additional columns will be ignored):\n",
    "\n",
    "    #Column1: Unique Peak ID\n",
    "    #Column2: chromosome\n",
    "    #Column3: starting position\n",
    "    #Column4: ending position\n",
    "    #Column5: not used\n",
    "    #Column6: Strand (+/- or 0/1, where 0=\"+\", 1=\"-\")\n",
    "\n",
    "# try with DA FASCICULATA :\n",
    "\n",
    "HOMER_FASC_F=as.data.frame(peaks_Male_FASC)\n",
    "HOMER_FASC_F=as.data.frame(DA_STERO_Female_Fasciculata$peaks[DA_STERO_Female_Fasciculata$avg_log2FC< 0 & DA_STERO_Female_Fasciculata$p_val_adj<0.01]) # create new object wwith peak\n",
    "names(HOMER_FASC_F)[1]='Column1'\n",
    "\n",
    "HOMER_FASC_F=as.data.frame(str_split_fixed(HOMER_FASC_F$Column1,\"-\",3)) # separate columns \n",
    "\n",
    "HOMER_FASC_F$column1=rownames(HOMER_FASC_F)\n",
    "HOMER_FASC_F$column1 <- paste(\"ID\", HOMER_FASC_F$column1, sep=\"_\")  # add unique ID columns\n",
    "\n",
    "HOMER_FASC_F=HOMER_FASC_F[c(4,1,2,3)]  # reorder columns ( take index of the columns)\n",
    "\n",
    "HOMER_FASC_F$columnn5=\"\"\n",
    "\n",
    "HOMER_FASC_F$Column6= '+'                # STRAND ?\n",
    "\n",
    "#HOMER_FASC_F$V1 <- sub(\"^\", \">\", HOMER_FASC_F$V1 )\n",
    "\n",
    "\n",
    "names(HOMER_FASC_F)[2]<-paste(\"Column2\")\n",
    "names(HOMER_FASC_F)[3]<-paste(\"Column3\")\n",
    "names(HOMER_FASC_F)[4]<-paste(\"Column4\") \n",
    "\n",
    "# save \n",
    "write.table(HOMER_FASC_F,'/home/absta/Documents/DATA_RUN/HOMER_FASC_F.bed',  sep = \" \", row.names = F ,col.names = F,quote = F)\n",
    "\n",
    "### get HOMER TO BED FORMAT \n",
    " # remove\n",
    "  bedfile=df = subset(HOMER_FASC_F, select = -c(columnn5,Column6))\n",
    " # reorder \n",
    "  bedfile <- bedfile[, c(2,3,4,1)]\n",
    "\n",
    "\n",
    "\n",
    "# STRAND ?\n",
    "\n",
    "# get background\n",
    "\n",
    "HOMER_FASC_M=as.data.frame(`DA_Female Fasciculata`$peaks[`DA_Female Fasciculata`$avg_log2FC<0]) # create new object wwith peak\n",
    "names(HOMER_FASC_M)[1]='Column1'\n",
    "\n",
    "HOMER_FASC_M=as.data.frame(str_split_fixed(HOMER_FASC_M$Column1,\"-\",3)) # separate columns \n",
    "\n",
    "HOMER_FASC_M$column1=rownames(HOMER_FASC_M)\n",
    "HOMER_FASC_M$column1 <- paste(\"ID\", HOMER_FASC_M$column1, sep=\"_\")  # add unique ID columns\n",
    "\n",
    "HOMER_FASC_M=HOMER_FASC_M[c(4,1,2,3)]  # reorder columns ( take index of the columns)\n",
    "\n",
    "#HOMER_FASC_M$V1 <- sub(\"^\", \">\", HOMER_FASC_M$V1 )\n",
    "\n",
    "HOMER_FASC_M$columnn5=\"\"\n",
    "\n",
    "HOMER_FASC_M$Column6= '+'            # STRAND ?\n",
    "\n",
    "\n",
    "names(HOMER_FASC_M)[2]<-paste(\"Column2\")\n",
    "names(HOMER_FASC_M)[3]<-paste(\"Column3\")\n",
    "names(HOMER_FASC_M)[4]<-paste(\"Column4\") \n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# save \n",
    "write.table(HOMER_FASC_M,'/home/absta/Documents/DATA_RUN/HOMER_FASC_M.bed', sep=\" \",row.names = F,col.names = F,quote = F)\n",
    "\n",
    "\n",
    "# run homer : findMotifsGenome.pl HOMER_FASC_F.csv mm10 /home/absta/Documents/DATA_RUN/outPUTFIRSTHOMER -size 200 -bg HOMER_FASC_M.csv\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d785cd38",
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "Link  peak to genes "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "814a2377",
   "metadata": {
    "lines_to_next_cell": 2,
    "name": "top Link peak to genes"
   },
   "outputs": [],
   "source": [
    "library(BSgenome.Mmusculus.UCSC.mm10)\n",
    "library(Signac)\n",
    "DefaultAssay(multiomics)='peaks'\n",
    "# first compute the GC content for each peak\n",
    "multiomics <- RegionStats(multiomics, genome = BSgenome.Mmusculus.UCSC.mm10)\n",
    "\n",
    "`DE_Female Glomerulosa`$gene_names=rownames(`DE_Female Glomerulosa`)\n",
    "`DE_Female Fasciculata`$gene_names=rownames(`DE_Female Fasciculata`)\n",
    "\n",
    "\n",
    "# # link peaks to genes Glomerulosa\n",
    "   #  compute gene to link with peaks :\n",
    "\n",
    "            Female_Fasc_gene=`DE_Female Fasciculata`$gene_names[`DE_Female Fasciculata`$avg_log2FC> 0]   #Testing 300 genes and 178944 peaks Found gene coordinates for 278 genes\n",
    "            Male_Fasc_gene=`DE_Female Fasciculata`$gene_names[`DE_Female Fasciculata`$avg_log2FC< 0]     #Testing 287 genes and 178944 peaks Found gene coordinates for 279 genes\n",
    "            \n",
    "            Female_Glo_gene=`DE_Female Glomerulosa`$gene_names[`DE_Female Glomerulosa`$avg_log2FC> 0]    #Testing 69 genes and 178944 peaks Found gene coordinates for 65 genes\n",
    "            Male_Glo_gene=`DE_Female Glomerulosa`$gene_names[`DE_Female Glomerulosa`$avg_log2FC< 0]      #Testing 39 genes and 178944 peaks Found gene coordinates for 37 genes\n",
    "\n",
    "x<- LinkPeaks(\n",
    "  object = multiomics,\n",
    "  peak.assay = \"peaks\",\n",
    "  expression.assay = \"SCT\",\n",
    "  genes.use = Male_Fasc_gene,\n",
    ")\n",
    "\n",
    "LINK_F_M=data.frame(Links(x))\n",
    "saveRDS(LINK_F_F,file = 'LINK_F_F')\n",
    "\n",
    "vec=Genes2peaks$peak[Genes2peaks$gene=='Akr1d1']\n",
    "FeaturePlot(multiomics,features = vec,reduction = 'umap.rna.pca')\n",
    "\n",
    "\n",
    "# i have run it on glomerulosa and fasciculata "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8895017e",
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "# Gene to peak correlation "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1c8c0a7e",
   "metadata": {
    "lines_to_next_cell": 0,
    "name": "take significant link"
   },
   "outputs": [],
   "source": [
    "\n",
    "peak_LINK_F_F=link_F_f$peak[link_F_f$score>0.2 & link_F_f$pvalue< 0.05]\n",
    "peak_LINK_F_M=link_F_M$peak[link_F_M$score>0.2 & link_F_M$pvalue< 0.05]\n",
    "peak_LINK_G_F=link_G_f$peak[link_G_f$score>0.2 & link_G_f$pvalue< 0.05]\n",
    "peak_LINK_G_M=link_G_m$peak[link_G_m$score>0.2 & link_G_m$pvalue< 0.05]\n",
    "\n",
    "\n",
    "\n",
    "      open.peaks <- AccessiblePeaks(multiomics, idents = c(\"Female Fasciculata\"),assay = 'peaks') # find a set of peaks that are open in a subset of cells\n",
    "      \n",
    "      meta.feature <- GetAssayData(multiomics, assay = \"peaks\", slot = \"meta.features\")# match the overall GC content in the peak set\n",
    "      peaks.matched <- MatchRegionStats(\n",
    "        meta.feature = meta.feature[open.peaks, ],\n",
    "        query.feature = meta.feature[peak_LINK_F_F, ],\n",
    "        n = 50000\n",
    "      )\n",
    "\n",
    "\n",
    "\n",
    "enriched.motifs_male <- FindMotifs(\n",
    "  object = multiomics,\n",
    "  features = peak_LINK_F_M  ,           # take vector of peak\n",
    "   background = peaks.matched\n",
    ")\n",
    "\n",
    "enriched.motifs_female <- FindMotifs(\n",
    "  object = multiomics,\n",
    "  features = peak_LINK_F_F  ,           # take vector of peak\n",
    "   background = peaks.matched\n",
    ")\n",
    "\n",
    "\n",
    "\n",
    "View(enriched.motifs_female)\n",
    "View(enriched.motifs_male)\n",
    "\n",
    "enriched.motifs_male=enriched.motifs_male[enriched.motifs_male$fold.enrichment>2 &enriched.motifs_male$pvalue<0.05 ,]\n",
    "\n",
    "enriched.motifs_female=enriched.motifs_female[enriched.motifs_female$fold.enrichment>2 &enriched.motifs_female$pvalue<0.05 ,]\n",
    "\n",
    "female_up = enriched.motifs_female$motif[order(enriched.motifs_female$fold.enrichment, decreasing = TRUE)]\n",
    "female_up=head(female_up)\n",
    "\n",
    "male_up = enriched.motifs_male$motif[order(enriched.motifs_male$fold.enrichment, decreasing = TRUE)]\n",
    "male_up=head(male_up)\n",
    "\n",
    "DefaultAssay(multiomics) <- 'chromvar'\n",
    "\n",
    "Male_motif=c('MA0007.3','MA0113.3','MA0727.1','MA1545.1')\n",
    "Female_motif=c('MA1628.1','MA0102.4','MA0043.3')\n",
    "\n",
    "\n",
    "  View(enrich_motif_Fem_TZONE)\n",
    "\n",
    " FeaturePlot(\n",
    "  object = only_STERO,\n",
    "  features = 'MA0043.3',\n",
    "  min.cutoff = 'q10',\n",
    "  max.cutoff = 'q100',\n",
    "  pt.size = 2,reduction = 'wnn.umap'\n",
    ")\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "190db23a",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "9ad5e22f",
   "metadata": {},
   "source": [
    "# Plot motif heatmap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b2d5c728",
   "metadata": {
    "lines_to_next_cell": 0,
    "name": "Plot motif heatmap"
   },
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "View()\n",
    "\n",
    "\n",
    "Male_motif=c('MA0007.3','MA0113.3')\n",
    "Female_motif=c('MA1628.1','MA0102.4','MA0043.3')\n",
    "\n",
    "minmaxthis = function(x) { \n",
    "  (x - min(x))/(max(x) - min(x))\n",
    "}\n",
    "\n",
    "trajFA=as.data.frame(only_STERO@reductions[[\"wnn.umap\"]]@cell.embeddings)\n",
    "trajFA$index=rownames(trajFA)\n",
    "\n",
    "\n",
    "# plot figure \n",
    "TFsnames =c('MA1628.1','MA0102.4','MA0043.3')\n",
    "genenames=c('Agtr1b')\n",
    "\n",
    "plot_list = list() \n",
    "\n",
    "    for (x in genenames) {\n",
    "      print(x)\n",
    "      \n",
    "      DefaultAssay(only_STERO) = \"SCT_RUN_STERO\"\n",
    "      p10 = FeaturePlot(\n",
    "        object = only_STERO,\n",
    "        features = str_replace_all(x, fixed(\"_\"), \"-\"),\n",
    "        min.cutoff = 'q10',\n",
    "        max.cutoff = 'q95',\n",
    "        pt.size = 3)\n",
    "      \n",
    "      chromvardf = p10$data\n",
    "      chromvardf$index = rownames(chromvardf)\n",
    "      \n",
    "      mergedFAchromvar = merge(chromvardf, trajFA, by = \"index\")\n",
    "      \n",
    "      \n",
    "      TFactplot = ggplot(mergedFAchromvar, aes( x = wnnUMAP_1.x, y = wnnUMAP_2.x, \n",
    "                                                colour = minmaxthis(mergedFAchromvar[,5]))) + \n",
    "        geom_point(size = 1.5, alpha = 1) + \n",
    "        scale_color_gradient(low = \"gray93\",high = \"red1\",breaks = seq(0,1,0.2))+\n",
    "        xlab(\"WNN1\") + ylab(\"WNN2\") +\n",
    "        theme(panel.border=element_blank(), \n",
    "              axis.line = element_line(colour = 'black', size = 1), \n",
    "              axis.ticks = element_blank(),\n",
    "              axis.text = element_blank(), \n",
    "              panel.grid.major = element_blank(),\n",
    "              panel.grid.minor = element_blank(),\n",
    "              panel.background = element_blank(),\n",
    "              legend.key=element_blank(),\n",
    "              legend.title=element_blank(),\n",
    "              legend.text=element_text(size=12)) + \n",
    "        # guides(color = guide_legend(override.aes max.cutoff= list(size=5))) + \n",
    "        coord_capped_cart(bottom='right', left='none')  + \n",
    "      ggtitle(x)\n",
    "      TFactplot\n",
    "      \n",
    "      \n",
    "      print(TFactplot)\n",
    "  \n",
    "}\n",
    "\n",
    "\n",
    "\n",
    "allmarkers=FindMarkers(only_STERO,ident.1 = c(\"Male_Fasciculata\", \"Male_Glomerulosa\",\"Male_tZone\"),ident.2 = c(\"Female_Glomerulosa\", \"Female_Fasciculata\",\"Female_tZone\"),logfc.threshold = 0,assay = 'SCT_RUN_STERO')\n",
    "\n",
    "allmarkers=FindAllMarkers(only_STERO,logfc.threshold = 0,assay = 'SCT_RUN_STERO')\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b561a495",
   "metadata": {},
   "source": [
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b1c7d79d",
   "metadata": {},
   "source": [
    "https://hbctraining.github.io/scRNA-seq_online/lessons/seurat_subclustering.html\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "74bf3e96",
   "metadata": {
    "lines_to_next_cell": 0
   },
   "outputs": [],
   "source": [
    "#keep only female in the data set\n",
    "library(Seurat)\n",
    "gene1traj <- WhichCells(alldata.int, expression = orig.ident == c(\"WTA\", \"WTC1\"))\n",
    "subtraj <- subset(alldata.int, cells=  gene1traj )\n",
    "DefaultAssay(alldata.int) <- \"CCA2\"\n",
    "\n",
    "\n",
    "\n",
    "#scanpy python\n",
    "#Convert RData file in h5 file\n",
    "library(SeuratDisk)\n",
    "library(SeuratData)\n",
    "SaveH5Seurat(subtraj, filename = \"subtraj.h5Seurat\")\n",
    "\n",
    "only_STERO=SCTransform(only_STERO,variable.features.n =NULL,variable.features.rv.th =1.1,return.only.var.genes = FALSE,assay = 'RNA',new.assay.name = 'SCT_RUN_STERO',vars.to.regress = 'percent.mt')\n",
    "\n",
    "# commande linux, modules installation, run just the first time\n",
    "\n",
    "#pip install pandas --user pip install matplotlib --user pip install scanpy --user pip install igraph --user pip install fa2 --user pip install leidenalg --user pip install anndata --user\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "51a20e11",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "b285babc",
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "# monocle3 trajectory analysis atac"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1881e52a",
   "metadata": {
    "name": "monocle3 trajectory analysis atac"
   },
   "outputs": [],
   "source": [
    "library(monocle3)\n",
    "library(monocle)\n",
    "library(cicero)\n",
    "library(Seurat)\n",
    "library(Signac)\n",
    "library(Signac)\n",
    "library(Seurat)\n",
    "library(SeuratWrappers)\n",
    "library(monocle3)\n",
    "library(Matrix)\n",
    "library(ggplot2)\n",
    "library(patchwork)\n",
    "DefaultAssay(multiomics) <- \"peaks\"\n",
    "\n",
    "\n",
    "## choose datasets\n",
    "male <- multiomics[,  multiomics$identity %in% c('male')]\n",
    "female <- multiomics[, multiomics$identity %in% c('female')]\n",
    "\n",
    "all_STERO <- multiomics[,  multiomics$cell_types %in% c(\"Male Fasciculata Replicate 1\",\"Male Fasciculata Replicate 2\",\"Female Glomerulosa\",\"Female Fasciculate Replicate 1\" ,\"Male Glomerulosa\",\"Capsular Cells\",\"Immune Cells\",\"Fetal adrenocortical Cells \",\"X zone \",\"Female Fasiculata Replicate 2\")]\n",
    "\n",
    "\n",
    "male_STERO <- multiomics[,  multiomics$identity %in% c(\"male\")& multiomics$cell_types %in% c(\"Male Fasciculata Replicate 1\",\"Male Fasciculata Replicate 2\",\"Female Glomerulosa\",\"Female Fasciculate Replicate 1\" ,\"Male Glomerulosa\",\"Capsular Cells\",\"Immune Cells\",\"Fetal adrenocortical Cells \",\"X zone \",\"Female Fasiculata Replicate 2\")]\n",
    "\n",
    "female_STERO <- multiomics[, multiomics$identity %in% c(\"female\")& multiomics$cell_types %in% c(\"Male Fasciculata Replicate 1\",\"Male Fasciculata Replicate 2\",\"Female Glomerulosa\",\"Female Fasciculate Replicate 1\" ,\"Male Glomerulosa\",\"Capsular Cells\",\"Immune Cells\",\"Fetal adrenocortical Cells \",\"X zone \",\"Female Fasiculata Replicate 2\")]\n",
    "\n",
    "### take reduction \n",
    "\n",
    "female_STERO@reductions$UMAP=female_STERO@reductions[[\"wnn.umap\"]]\n",
    "male_STERO@reductions$UMAP=male_STERO@reductions[[\"wnn.umap\"]]\n",
    "all_STERO@reductions$UMAP=all_STERO@reductions[[\"wnn.umap\"]]\n",
    "\n",
    "\n",
    "\n",
    "# root for trajectory \n",
    "all_root <- all_STERO[, all_STERO$cell_types %in% c(\"Female Glomerulosa\",\"Male Glomerulosa\")]\n",
    "vector_root_all=colnames(all_root)\n",
    "\n",
    "\n",
    "female_glomerulosa_root <- female_STERO[, female_STERO$cell_types %in% c(\"Female Glomerulosa\",\"Male Glomerulosa\")]\n",
    "male_glomerulosa_root <- male_STERO[, male_STERO$cell_types %in% c(\"Male Glomerulosa\",\"Female Glomerulosa\")]\n",
    "vector_root_M=colnames(male_glomerulosa_root)\n",
    "vector_root_F=colnames(female_glomerulosa_root)\n",
    "\n",
    "\n",
    "# Building trajectories with Monocle 3\n",
    "\n",
    "\n",
    "\n",
    "all.cds <- as.cell_data_set(all_STERO)\n",
    "all.cds <- cluster_cells(cds = all.cds, reduction_method = \"UMAP\",resolution = 30,max.overlaprs=10)\n",
    "all.cds <- learn_graph(all.cds, use_partition = TRUE)\n",
    "\n",
    "female.cds <- as.cell_data_set(female_STERO)\n",
    "female.cds <- cluster_cells(cds = female.cds, reduction_method = \"UMAP\")\n",
    "female.cds <- learn_graph(female.cds, use_partition = TRUE)\n",
    "\n",
    "male.cds <- as.cell_data_set(male_STERO)\n",
    "male.cds <- cluster_cells(cds = male.cds, reduction_method = \"UMAP\")\n",
    "male.cds <- learn_graph(male.cds, use_partition = TRUE)\n",
    "\n",
    "# order cells\n",
    "all.cds <- order_cells(all.cds, reduction_method = \"UMAP\" ,root_cells =vector_root_all )\n",
    "\n",
    "\n",
    "male.cds <- order_cells(male.cds, reduction_method = \"UMAP\" ,root_cells =vector_root_M )\n",
    "female.cds <- order_cells(female.cds, reduction_method = \"UMAP\",root_cells = vector_root_F )\n",
    "\n",
    "\n",
    "\n",
    "x=plot_cells(\n",
    "  cds = male.cds,\n",
    "  color_cells_by = \"pseudotime\",\n",
    "  show_trajectory_graph = TRUE,cell_size = 2,trajectory_graph_segment_size = 2,graph_label_size = 2,label_cell_groups = F,label_branch_points = T,)+\n",
    "  plot_cells(\n",
    "  cds = female.cds,\n",
    "  color_cells_by = \"pseudotime\",\n",
    "  show_trajectory_graph = TRUE,cell_size = 2,trajectory_graph_segment_size = 2,graph_label_size = 2,label_cell_groups = F,label_branch_points = T)\n",
    "\n",
    "\n",
    "\n",
    "# # a helper function to identify the root principal points:\n",
    "get_earliest_principal_node <- function(cds){\n",
    "  cell_ids <- which(colData(cds)[, \"cell_types\"] %in% c(\"Female Glomerulosa\",\"Male Glomerulosa\"))\n",
    "  \n",
    "  closest_vertex <-\n",
    "  cds@principal_graph_aux[[\"UMAP\"]]$pr_graph_cell_proj_closest_vertex\n",
    "  closest_vertex <- as.matrix(closest_vertex[colnames(cds), ])\n",
    "  root_pr_nodes <-\n",
    "  igraph::V(principal_graph(cds)[[\"UMAP\"]])$name[as.numeric(names\n",
    "  (which.max(table(closest_vertex[cell_ids,]))))]\n",
    "  \n",
    "  root_pr_nodes\n",
    "}\n",
    "cds_male <- order_cells(male.cds, root_pr_nodes=get_earliest_principal_node(male.cds))\n",
    "cds_female <- order_cells(female.cds, root_pr_nodes=get_earliest_principal_node(female.cds))\n",
    "all.cds <- order_cells(all.cds, root_pr_nodes=get_earliest_principal_node(all.cds))\n",
    "\n",
    "\n",
    "# add  umap pseudotime value in seurat object \n",
    "\n",
    "\n",
    "multiomics <- AddMetaData(\n",
    "  object = multiomics,\n",
    "  metadata = all.cds@principal_graph_aux@listData$UMAP$pseudotime,\n",
    "  col.name = \"all_sexe_trajectory_umap\"\n",
    ")\n",
    "\n",
    "\n",
    "multiomics <- AddMetaData(\n",
    "  object = multiomics,\n",
    "  metadata = male.cds@principal_graph_aux@listData$UMAP$pseudotime,\n",
    "  col.name = \"male\"\n",
    ")\n",
    "\n",
    "FeaturePlot(multiomics, c(\"female\", \"male\"), pt.size = 1,reduction = 'wnn.umap',label = T) & scale_color_viridis_c()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6ad58d1b",
   "metadata": {
    "lines_to_next_cell": 2,
    "name": "Trajectory analysis with TF"
   },
   "outputs": [],
   "source": [
    "library(dplyr)\n",
    "\n",
    "\n",
    "multiomics@meta.data$index=rownames(multiomics@meta.data)\n",
    "TRAJECTORY_ALL_STERO = select(multiomics@meta.data, all_sexe_trajectory_umap, barcode)\n",
    "TRAJECTORY_ALL_STERO=na.omit(object =TRAJECTORY_ALL_STERO)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c7072d8d",
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "https://mojaveazure.github.io/seurat-disk/articles/convert-anndata.html\n",
    "\n",
    "# convert Seurat object to scanpy object "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "91897f79",
   "metadata": {
    "name": "convert seurat object to scanpy"
   },
   "outputs": [],
   "source": [
    "library(SeuratDisk)\n",
    "multiomics_nomotif=multiomics\n",
    "multiomics_nomotif@assays[[\"peaks\"]]@motifs =NULL # error if features in motif slot \n",
    "SaveH5Seurat(multiomics_nomotif, filename = \"multiomics_convert.h5Seurat\")\n",
    "Convert(\"multiomics_convert.h5Seurat\", dest = \"h5ad\")\n",
    "pbmc_ad <- ad$read_h5ad(\"/C:/Users/User/Downloads\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eea81c1f",
   "metadata": {
    "name": "run scanpy"
   },
   "outputs": [],
   "source": [
    "%%python\n",
    "import scanpy  # pip install in  R terminal\n",
    "import scanpy as sc\n",
    "import numpy as np\n",
    "\n",
    "import pandas as pd\n",
    "from matplotlib import rcParams\n",
    "import os\n",
    "\n",
    "\n",
    "import anndata2ri\n",
    "adata = sc.read_h5ad(\"subtraj.h5ad\")\n",
    "scanpy_peaks=scanpy.read_h5ad('C:/Users/User/Downloads/multiomics_convert.h5ad')\n",
    "\n",
    "\n",
    "\n",
    "# sc.tl.pca(adata, svd_solver='arpack')\n",
    "sc.pp.neighbors(adata, n_neighbors=4, n_pcs=20)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f0b3d78c",
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "https://smorabit.github.io/tutorials/8_velocyto/#:~:text=Briefly%2C%20RNA%20velocity%20analysis%20allows,give%20cell%20population%20of%20interest.\n",
    "\n",
    "# RNA velovity "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6c1a3d6b",
   "metadata": {
    "lines_to_next_cell": 2,
    "name": "RNA velocity analysis with scVelo"
   },
   "outputs": [],
   "source": [
    "\n",
    "library(Matrix)\n",
    "\n",
    "\n",
    "#Step -1: Convert data from Seurat to Python / anndata\n",
    "\n",
    "# save metadata table:\n",
    "multiomics$barcode <- colnames(multiomics)\n",
    "multiomics$UMAP_1 <- multiomics@reductions$wnn.umap@cell.embeddings[,1]\n",
    "multiomics$UMAP_2 <- multiomics@reductions$wnn.umap@cell.embeddings[,2]\n",
    "write.csv(multiomics@meta.data, file='metadata.csv', quote=F, row.names=F)\n",
    "# write expression counts matrix\n",
    "counts_matrix <- GetAssayData(multiomics, assay='peaks', slot='counts')\n",
    "writeMM(counts_matrix, file= ('counts.mtx'))\n",
    "\n",
    "# write dimesnionality reduction matrix, in this example case pca matrix\n",
    "  write.csv(multiomics@reductions$lsi@cell.embeddings, file='lsi.csv', quote=F, row.names=F)\n",
    "\n",
    "# write gene names\n",
    "write.table(\n",
    "  data.frame('peaks'=rownames(counts_matrix)),file='peaks_names.csv',\n",
    "  quote=F,row.names=F,col.names=F\n",
    ")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e5da0fef",
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "# PLOT FIG5 "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "280885f9",
   "metadata": {
    "lines_to_next_cell": 0,
    "name": "PLOT FIG5"
   },
   "outputs": [],
   "source": [
    "library(Seurat)\n",
    "library(tidyverse)\n",
    "library(BSgenome.Mmusculus.UCSC.mm10)\n",
    "library(EnsDb.Mmusculus.v79)\n",
    "library(Signac)\n",
    "library(cowplot)\n",
    "library(patchwork)\n",
    "library(forcats)\n",
    "    # feature matrix \n",
    "peaks=GetAssayData(object = multiomics, slot = \"counts\",assay = 'peaks')\n",
    "\n",
    "ext = c(3)\n",
    "\n",
    "# extract gene coordinates from Ensembl, and ensure name formatting is consistent with Seurat object \n",
    "gene.coords <- genes(EnsDb.Mmusculus.v79, filter = ~ gene_biotype == \"protein_coding\")\n",
    "ucsc.levels <- str_replace(string=paste(\"chr\",seqlevels(gene.coords),sep=\"\"), pattern=\"chrMT\", replacement=\"chrM\")\n",
    "seqlevels(gene.coords) <- ucsc.levels\n",
    "genebody.coords <- keepStandardChromosomes(gene.coords, pruning.mode = 'coarse')\n",
    "\n",
    "genebodyandpromoter.coords <- Extend(x = gene.coords, upstream = 1000*ext, downstream = 1000*ext)\n",
    "gene.key <- genebodyandpromoter.coords$gene_name\n",
    "names(gene.key) <- GRangesToString(grange = genebodyandpromoter.coords)\n",
    "\n",
    "rownames(peaks) <- gene.key[rownames(peaks)]\n",
    "\n",
    "peaks = peaks[unique(rownames(peaks)),]\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "  multiomics[[paste0(\"RNA_geneprom_\",ext,\"kb\")]] <- CreateAssayObject(counts = peaks)\n",
    "  # test = BinarizeCounts(object = test, assay = c(\"RNA\"))\n",
    "  \n",
    "  test <- NormalizeData(\n",
    "    object = test,\n",
    "    assay = paste0(\"RNA_geneprom_\",ext,\"kb\"),\n",
    "    normalization.method = 'LogNormalize',\n",
    "    scale.factor = median(test@meta.data[,paste0(\"nFeature_RNA_geneprom_\",ext,\"kb\")])\n",
    "  )\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "#top10 markers genes through all clusters \n",
    "all_markers_genes_peaks=FindAllMarkers(multiomics,assay = 'SCT')\n",
    "x=all_markers_genes[all_markers_genes$p_val_adj<0.05,]\n",
    "\n",
    "gen_marker_table <- function(x){\n",
    "all_markers_genes[all_markers_genes$cluster == x, ] %>%\n",
    "head(n=3)}\n",
    "# view\n",
    "vec=levels(multiomics)\n",
    "top10_markers <- map_dfr(vec, gen_marker_table)\n",
    "top10_markers=top10_markers[!duplicated(top10_markers$gene),]\n",
    "top_genes=as.vector(top10_markers$gene)\n",
    "\n",
    "\n",
    "dotx=DotPlot(multiomics,assay= 'SCT',features = top_genes)+ RotatedAxis() +  scale_colour_gradient2(low = \"darkred\", mid = \"white\",high = \"darkgreen\",midpoint = 0)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ca4f241a",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "12e374f8",
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "# Ploting motif along cluster  Fig6"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3cf5a4f6",
   "metadata": {
    "lines_to_next_cell": 0,
    "name": "PLOT FIG6"
   },
   "outputs": [],
   "source": [
    "# Main plot\n",
    "  minmaxthis = function(x) { \n",
    "    (x - min(x))/(max(x) - min(x))\n",
    "  }\n",
    "  \n",
    "  data.dot = dotx$data\n",
    "  \n",
    "  \n",
    "  scaled_data <- \n",
    "    data.dot %>%\n",
    "    group_by(features.plot) %>%\n",
    "    mutate(minmaxscale = minmaxthis(avg.exp))\n",
    "  \n",
    "  scaled_data$avg.exp2 = ( scaled_data$avg.exp - min(scaled_data$avg.exp) )    / \n",
    "    (max(scaled_data$avg.exp)- min(scaled_data$avg.exp))\n",
    "  \n",
    "  \n",
    "  pmain <- ggplot(scaled_data)  + \n",
    "    geom_point(aes(x = features.plot, y = id, color = minmaxscale, size = pct.exp), stat = \"identity\") + \n",
    "    scale_size(range = c(0, 4.25), breaks  = seq(0,100,20), labels  = paste0(seq(0,100,20),\"%\"))  + \n",
    "    scale_color_gradient(low = \"white\",high = \"#0f4d91\",breaks = seq(0,1,0.2))  +\n",
    "    guides(color = guide_colourbar(title = \"Accessibility\", barwidth = 1, barheight = 10),\n",
    "           size = guide_legend(title = \"Cells (%)\",reverse=TRUE)) + \n",
    "    xlab(\"\") + ylab(\"\") + \n",
    "    \n",
    "    theme(panel.border=element_blank(), \n",
    "          axis.line = element_line(colour = 'black', size = 1), \n",
    "          axis.ticks = element_blank(),\n",
    "          axis.text = element_text(size=12, colour = \"black\"), \n",
    "          axis.text.x = element_text(angle = 90, hjust = 1),\n",
    "          panel.grid.major = element_blank(),\n",
    "          panel.grid.minor = element_blank(),\n",
    "          panel.background = element_blank(),\n",
    "          legend.key=element_blank(),\n",
    "          # legend.title=element_blank(),\n",
    "          legend.text=element_text(size=12))\n",
    "  \n",
    "\n",
    "  \n",
    " \n",
    "  xdens <- ggplot() +\n",
    "    geom_bar(data = scaled_data, aes(x = features.plot, y = avg.exp2),stat = \"summary\",\n",
    "             alpha = 1, size = 0.2, fill = \"black\") + \n",
    "    scale_y_continuous(breaks = seq(0,1,0.5), labels = seq(0,1,0.5), limits = c(0,1)) + \n",
    "    \n",
    "    # theme_bw() + \n",
    "    theme(panel.border=element_blank(), \n",
    "          axis.line = element_line(colour = 'black', size = 1), \n",
    "          axis.ticks = element_blank(),\n",
    "          axis.line.x = element_blank(),\n",
    "          axis.text = element_text(size=12, colour = \"black\"), \n",
    "          panel.grid.major = element_blank(),\n",
    "          panel.grid.minor = element_blank(),\n",
    "          panel.background = element_blank(),\n",
    "          legend.key=element_blank(),\n",
    "          # legend.title=element_blank(),\n",
    "          legend.text=element_text(size=12),\n",
    "          axis.text.x = element_blank(),\n",
    "          axis.ticks.x = element_blank(),\n",
    "          axis.title.x  = element_blank()) + \n",
    "    ylab(\"Average \\n accessibility\")\n",
    "\n",
    "\n",
    "\n",
    "  dotx1 = xdens + pmain + plot_layout(ncol = 1, heights = c(1,3))\n",
    "\n",
    "\n",
    "  # Main plot\n",
    "\n",
    "  minmaxthis = function(x) { \n",
    "    (x - min(x))/(max(x) - min(x))\n",
    "  }\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f03c934d",
   "metadata": {},
   "source": [
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "74e3d913",
   "metadata": {
    "lines_to_next_cell": 0
   },
   "outputs": [],
   "source": [
    "dfATAC = only_STERO@meta.data\n",
    "dfATAC$index = rownames(dfATAC)\n",
    "\n",
    "trajFA=as.data.frame(only_STERO@reductions[[\"wnn.umap\"]]@cell.embeddings)\n",
    "trajFA$index=rownames(trajFA)\n",
    "\n",
    "\n",
    "mergeFAATAC = merge(trajFA, dfATAC, by = \"index\")\n",
    "\n",
    "# \n",
    "plot3C = ggplot(mergeFAATAC, aes( x = wnnUMAP_1, y = wnnUMAP_2, colour = new_ident)) + \n",
    "  geom_point(size = 1, alpha = 1) + \n",
    "  scale_colour_manual(values = c( '#3CB44B', '#FFE119', '#4363D8', '#F58231', '#911EB4', '#46F0F0','#E6194B')) +\n",
    "  xlab(\"FA1\") + ylab(\"FA2\") +\n",
    "  theme(panel.border=element_blank(), \n",
    "        axis.line = element_line(colour = 'black', size = 1), \n",
    "        axis.ticks = element_blank(),\n",
    "        axis.text = element_blank(), \n",
    "        panel.grid.major = element_blank(),\n",
    "        panel.grid.minor = element_blank(),\n",
    "        panel.background = element_blank(),\n",
    "        legend.key=element_blank(),\n",
    "        legend.title=element_blank(),\n",
    "        legend.text=element_text(size=12)) + \n",
    "  guides(color = guide_legend(override.aes = list(size=5))) + \n",
    "  coord_capped_cart(bottom='right', left='none') \n",
    "\n",
    "plot3D = ggplot(mergeFAATAC, aes( x = wnnUMAP_1, y = wnnUMAP_2, colour = minmaxthis(nCount_peaks))) + \n",
    "  geom_point(size = 1, alpha = 1) + \n",
    "   scale_color_viridis(breaks = seq(0,1,0.2)) + \n",
    "   # scale_color_gradient(low = \"gray93\",high = \"red1\",breaks = seq(0,1,0.2), limits = c(0.4,1)) +\n",
    "  xlab(\"FA1\") + ylab(\"FA2\") +\n",
    "  theme(panel.border=element_blank(), \n",
    "        axis.line = element_line(colour = 'black', size = 1), \n",
    "        axis.ticks = element_blank(),\n",
    "        axis.text = element_blank(), \n",
    "        panel.grid.major = element_blank(),\n",
    "        panel.grid.minor = element_blank(),\n",
    "        panel.background = element_blank(),\n",
    "        legend.key=element_blank(),\n",
    "        legend.title=element_blank(),\n",
    "        legend.text=element_text(size=12)) + \n",
    "  # guides(color = guide_legend(override.aes = list(size=5))) + \n",
    "  coord_capped_cart(bottom='right', left='none') \n",
    "\n",
    "  print(plot3C)\n",
    "  print(plot3D)\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "04385d89",
   "metadata": {},
   "source": [
    "\n"
   ]
  }
 ],
 "metadata": {
  "jupytext": {
   "cell_metadata_filter": "name,-all",
   "main_language": "R",
   "notebook_metadata_filter": "-all"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
